// @wix/business-manager-api/dist/src/appDefIds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.appDefIds = void 0;
// these are the currently known apps, add your own if needed
exports.appDefIds = {
    dashboard: 'a7597ab5-1ed4-458f-a5c1-5884d5e14281',
    autoCMS: '9f87f125-e7d5-4ed9-a16c-999d862e96ab',
    pointOfSale: '44562f1b-17ec-4907-b242-b76ac06e707a',
    challenges: '2936472a-a1ed-4ae5-9f71-614313a9f4e7',
    blogImporter: 'd025eced-50bd-4579-b91b-9de9ac81eb8e',
    metaSite: '22bef345-3c5b-4c18-b782-74d4085112ff',
    wixECommerce: '1380b703-ce81-ff05-f115-39571d94dfcd',
    wixStores: '215238eb-22a5-4c36-9e7b-e7c08025e04e',
    ecomPlatform: '1380b703-ce81-ff05-f115-39571d94dfcd',
    shoutout: '135c3d92-0fea-1f9d-2ba5-2a1dfb04297e',
    invoices: '13ee94c1-b635-8505-3391-97919052c16f',
    recurringInvoices: '35aec784-bbec-4e6e-abcb-d3d724af52cf',
    crmFinancialSettings: '8ea9df15-9ff6-4acf-bbb8-8d3a69ae5841',
    priceQuotes: '55cd9036-36bb-480b-8ddc-afda3cb2eb8d',
    engage: '141fbfae-511e-6817-c9f0-48993a7547d1',
    contacts: '74bff718-5977-47f2-9e5f-a9fd0047fd1f',
    cashier: '14bca956-e09f-f4d6-14d7-466cb3f09103',
    wixBookings: '13d21c63-b5ec-5912-8397-c3a5ddb27a97',
    restaurants: '13c1402c-27f2-d4ab-7463-ee7c89e07578',
    hotels: '135aad86-9125-6074-7346-29dc6a3c9bcf',
    restaurantsKit: '14583ff5-e781-063a-3bc4-6b79fb966992',
    restaurantsOrders: '13e8d036-5516-6104-b456-c8466db39542',
    restaurantsSeating: '1475ab65-206b-d79a-856d-fa10bdb479ea',
    restaurantsPhoneOrders: '89584bcb-7d89-46c9-89a3-8aa6cc3111c9',
    orders: '13e8d036-5516-6104-b456-c8466db39542',
    kit: '14583ff5-e781-063a-3bc4-6b79fb966992',
    wixContactForm: 'ContactForm',
    wixSubscriptionForm: '13bd3cab-9ae3-852f-a4a9-6087f6e6c50a',
    wixSiteMembers: 'SiteMembers',
    formBuilder123: '12aacf69-f3fb-5334-2847-e00a8f13c12f',
    getSubscribers: '1375baa8-8eca-5659-ce9d-455b2009250d',
    constantContact: '13472cb5-8c98-0884-ad1b-95ddfe1577d4',
    oldWixECommerce: 'ECommerce',
    triggers: '139ef4fa-c108-8f9a-c7be-d5f492a2c939',
    wixForms: '14ce1214-b278-a7e4-1373-00cebd1bef7c',
    waiverAppInstaller: 'fd8664de-f48e-450d-ba14-c2fec494d9c1',
    waiverApp: '1d9fdfdf-e744-4db1-a62d-b9ef26fd48bc',
    siteSettings: '0fbaa11c-94d4-451f-b62a-c024879c5b38',
    businessSettings: 'ccad1ffa-21aa-4e5d-9ec3-431ce56a79e9',
    gmail: '130513da-106b-c416-6418-8464c68e7228',
    wixMusic: '13bb5d67-1add-e770-a71f-001277e17c57',
    scheduler: '13d21c63-b5ec-5912-8397-c3a5ddb27a97',
    quotes: '13ee94c1-b635-8505-3391-97919052c16f',
    inbox: '141fbfae-511e-6817-c9f0-48993a7547d1',
    inboxKeyboards: 'e1c67caa-0593-46d9-bb4a-21fdb3a89409',
    seoWizard: '1480c568-5cbd-9392-5604-1148f5faffa0',
    oldSeo: '130e98ba-8b86-2a9b-4e5b-aee5345cb49a',
    events: '140603ad-af8d-84a5-2c80-a0f60cb47351',
    video: '14409595-f076-4753-8303-9a86f9f71469',
    videoMaker: '14e12b04-943e-fd32-456d-70b1820a2ff2',
    chat: '14517e1a-3ff0-af98-408e-2bd6953c36a2',
    faqChatBot: '21eaa91c-2228-4109-b129-fcc552ff1332',
    faq: '14c92d28-031e-7910-c9a8-a670011e062d',
    promote: '14b89688-9b25-5214-d1cb-a3fb9683618b',
    promoteSeo: '1480c568-5cbd-9392-5604-1148f5faffa0',
    engageEmail: '143b5694-c94b-2eba-62a5-fbb94bde27c4',
    photographers: '13ff8629-c1fc-e289-e81f-bc8c8968e9d6',
    artStore: '147ab90e-91c5-21b2-d6ca-444c28c8a23b',
    emptyGuid: 'ffffffff-ffff-ffff-ffff-ffffffffffff',
    socialBlog: '14bcded7-0066-7c35-14d7-466cb3f09103',
    blog: '61f33d50-3002-4882-ae86-d319c1a249ab',
    wixCodeApp: 'CloudSiteExtension',
    adminPages: '1514b01b-cb08-41b4-8582-0a88551ac769',
    insights: '14986290-586b-b5cb-cc68-cba25a16ff54',
    facebookAds: '143e2a90-2b87-5318-1dea-c8e35204db73',
    coupons: '14d7032a-0a65-5270-cca7-30f599708fed',
    wixCode: '675bbcef-18d8-41f5-800e-131ec9e08762',
    multilingual: '14d84998-ae09-1abf-c6fc-3f3cace5bf19',
    marketingIntegration: '150ae7ee-c74a-eecd-d3d7-2112895b988a',
    oldMagentoStore: '55a88716-958a-4b91-b666-6c1118abdee4',
    membership: '1522827f-c56c-a5c9-2ac9-00f9e6ae12d3',
    staffManagement: 'df40ed72-1924-408e-be53-3eb48700c4a5',
    subscriptions: '8725b255-2aa2-4a53-b76d-7d3c363aaeea',
    wixChat: '14517e1a-3ff0-af98-408e-2bd6953c36a2',
    platforms101Workshop: '366624a5-4c5e-48c4-8147-86638ec2d010',
    promoteHome: 'f123e8f1-4350-4c9b-b269-04adfadda977',
    ascendTasks: '146c0d71-352e-4464-9a03-2e868aabe7b9',
    ascend: 'ee21fe60-48c5-45e9-95f4-6ca8f9b1c9d9',
    ascendBizModule: '9b9535c7-62d9-4ff6-aae4-cd0ffb49e5c4',
    gdprUsersClient: 'f9f22411-f2b8-411f-9fd3-f0118d6ff88c',
    crmHome: 'b030d78e-30c7-4913-beb5-26d3b70e53aa',
    promoteVideoMakerHome: 'e3118e0a-b1c1-4e1d-b67d-ddf0cb92309b',
    videoMakerHome: '307ba931-689c-4b55-bb1d-6a382bad9222',
    workflow: 'ea2821fc-7d97-40a9-9f75-772f29178430',
    promotePaidAds: '9bead16f-1c73-4cda-b6c4-28cff46988db',
    wixPhotoAlbums: '64db1bfa-e66f-4861-8ad0-8c34f39e4335',
    wixExpertsDashboard: 'bdd8f657-64e6-421e-a5e3-cc8f52136c0d',
    wixPartnerProgram: '0b026edb-8e4f-4880-826b-0d916067ff63',
    revenueShareForPartnerProgram: '49e26bf5-1bc4-461f-bcfb-53945b669f00',
    ascendForPartnerProgram: 'c980e8d0-41a4-4c58-a85a-1095124638db',
    clientBillingForPartnerProgram: 'a6e484b2-8fba-48a7-aac5-f1f459f873bb',
    arena: '2edf484d-9c39-461b-8cbd-a3ba6ac9c575',
    promoteSeoPatterns: '13aa9735-aa50-4bdb-877c-0bb46804bd71',
    promoteSeoTools: 'd70b68e2-8d77-4e0c-9c00-c292d6e0025e',
    MobileTab: 'a322993b-2c74-426f-bbb8-444db73d0d1b',
    wixGroups: '148c2287-c669-d849-d153-463c7486a694',
    virtualNumbers: '94bc563b-675f-41ad-a2a6-5494f211c47b',
    forum: '14724f35-6794-cd1a-0244-25fd138f9242',
    secretsVault: '4b10fcce-732d-4be3-9d46-801d271acda9',
    zoomApp: 'c6d111d1-1a8d-49f0-a78c-4f1823fb6592',
    consentPolicyBanner: 'f105ba16-6b7a-4b52-a2e5-712bdc749f76',
    consentPolicySettingsApp: 'b62fa321-f61f-492d-a4f3-fa2681538ac0',
    siteRolesAndPermissions: '1ee3cefa-6fa0-4270-93c5-6412a4418aac',
    wixPaymentsBalance: '44616341-9ea2-4d3b-a091-a0e717a485c0',
    wixPaymentsTransactions: '74c726d0-7b48-4963-8c53-9c26a35696c0',
    wixPaymentsDashboard: '42b081e6-c404-434f-a958-975864968d52',
    cashierPaymentsDashboard: 'd01c8411-bcf8-4852-8c8e-faee6921e7f9',
    premiumManageSubscriptions: 'b83ea993-1e5d-49e3-8c29-32b3aedce5c1',
    premiumPartnersClientSubscriptions: '8eb38b13-5884-47a1-8e6c-a8d6a4c9bb0d',
    fedata: 'c8ec1742-0cff-42c7-b3c5-b52a22d24a3c',
    releaseManager: '8c41f3a6-47f7-40be-ad77-02a72cb2ddb3',
    logoBuilder: '6e71286b-8b8d-467a-8f55-4539481cb946',
    memberPermissions: 'eb377299-86b4-4a86-a1b5-774a2d1d374b',
    analytics: 'dd9167a2-71d4-4f60-b3a8-ead4863d930c',
    siteHistory: '741c78ba-0caa-40e6-bf1e-9a9e1b0d7c51',
    assignee: 'f8859728-6fe9-49e6-8c63-8c7abf593087',
    partnersTeam: '0587eb5b-7261-4429-a946-3ed91f2351a6',
    businessManagerSettings: '3cfd35f9-4087-4192-ada5-d10938150d26',
    badgeDefinitions: '89b7f172-365a-4467-a2c7-7ad8b224e927',
    editor: '1797f372-c9c6-4fb9-9699-5f8b2fcc233c',
    newReleasesWidget: '9a42dcb9-08f2-41d3-ae1d-417471cb7439',
    printShop: '138a4c17-3f53-420a-bb71-21c2f9e0a361',
    ping: '2e4dafe7-1c15-49b8-acf1-2ae651ca2faf',
    appMarket: '98a1ffc7-89be-4b7d-b2d4-a4d0f603cb26',
    etpaContainer: '31077522-9329-4244-91d1-97653a06ca46',
    financialAppCollection: 'c3ec9b5f-3b7b-4346-94ec-85b137536898',
    manageInstalledAppsContainer: '21a8622c-5b74-4a65-b8f5-c5b9a90068b4',
    marketplaceContainer: '16c3b514-cce9-433c-aa1e-6434606c6a7a',
    siteBranches: 'd97ee8b6-1b2b-451c-970e-4eef8b8aa7d7',
    wixRicosDemo: '3e1dc7e8-3444-45ba-a54b-884466c7da0e',
    liveVideo: '2608697e-067a-44e2-8250-8e390df5bf71',
    partnersPackage: '9149c32c-a204-4d50-bf31-e7630999cc9c',
    wixHotels: '88c4fa14-da2c-457c-83d2-33974d0df2e6',
    loyalty: '553c79f3-5625-4f38-b14b-ef7c0d1e87df',
    loyaltyBm: 'de1a454f-92ac-4c7a-a4c3-cf253313bfbd',
    loyaltyExampleIntegration: '26d10d24-a108-40ec-b757-5d614ad8e9ee',
    faqSite: '14769d50-cb0d-0fc8-e97c-bd56b874d581',
    moderation: '8c32640d-e73d-4a15-91e2-d9b7a89a98f6',
    fileShare: '1537b24e-29d1-6d8f-b8e1-d6860f2f70b9',
    promoteSeoRedirects: 'fd925ec6-23ff-4871-9883-955115ce1bc4',
    fileShareApp: '1533b427-ca7c-a246-0434-83d3d3737bf3',
    websiteChannel: '36cafb61-2a71-4331-9c69-681b55f88430',
    brandedAppChannel: 'eec3496e-44a8-45ac-9581-868a67345be8',
};
