// @wix/business-manager-api/dist/src/buildPageComponentsInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var wix_experiments_1 = require("@wix/wix-experiments");
var appDefIds_1 = require("./appDefIds");
var PageComponentId_1 = require("./PageComponentId");
var ModuleId_1 = require("./ModuleId");
//....................................................................................
//......SSSSSSSSS......TTTTTTTTTTTTTTTTTTT......OOOOOOOOOO........PPPPPPPPPPPP........
//....SSSSSSSSSSSSS...STTTTTTTTTTTTTTTTTTT....OOOOOOOOOOOOO......PPPPPPPPPPPPPPPP.....
//...SSSSSSSSSSSSSSS..STTTTTTTTTTTTTTTTTTT...OOOOOOOOOOOOOOOO....PPPPPPPPPPPPPPPPP....
//...SSSSSSSSSSSSSSSS.STTTTTTTTTTTTTTTTTTT..OOOOOOOOOOOOOOOOOO...PPPPPPPPPPPPPPPPPP...
//..SSSSSSSSSSSSSSSSSSSTTTTTTTTTTTTTTTTTTT.OOOOOOOOOOOOOOOOOOO...PPPPPPPPPPPPPPPPPP...
//..SSSSSSS....SSSSSSS.TTTTTTTTTTTTTTTTTTTOOOOOOOOO...OOOOOOOOO..PPPPPPPPPPPPPPPPPPP..
//..SSSSSS......SSSSSS.......TTTTTTT......OOOOOOO.......OOOOOOO..PPPPPPP.....PPPPPPP..
//..SSSSSS......SSSSSS.......TTTTTTT......OOOOOOO.......OOOOOOOO.PPPPPPP.....PPPPPPP..
//..SSSSSSSS.................TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPP.....PPPPPPP..
//..SSSSSSSSSSSS.............TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPP.....PPPPPPP..
//..SSSSSSSSSSSSSSS..........TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPPPPPPPPPPPPPP..
//...SSSSSSSSSSSSSSSS........TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPPPPPPPPPPPPPP..
//....SSSSSSSSSSSSSSSS.......TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPPPPPPPPPPPPP...
//......SSSSSSSSSSSSSSS......TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPPPPPPPPPPPP....
//..........SSSSSSSSSSS......TTTTTTT.....TOOOOOO.........OOOOOOO.PPPPPPPPPPPPPPPP.....
//..SSSSS.......SSSSSSS......TTTTTTT......OOOOOO.........OOOOOOO.PPPPPPPPPPPPPP.......
//.SSSSSSS.......SSSSSS......TTTTTTT......OOOOOOO.......OOOOOOOO.PPPPPPP..............
//.SSSSSSS......SSSSSSS......TTTTTTT......OOOOOOOO.....OOOOOOOO..PPPPPPP..............
//.SSSSSSSS....SSSSSSSS......TTTTTTT.......OOOOOOOO...OOOOOOOOO..PPPPPPP..............
//..SSSSSSSSSSSSSSSSSS.......TTTTTTT.......OOOOOOOOOOOOOOOOOOO...PPPPPPP..............
//...SSSSSSSSSSSSSSSSS.......TTTTTTT........OOOOOOOOOOOOOOOOO....PPPPPPP..............
//...SSSSSSSSSSSSSSSS........TTTTTTT.........OOOOOOOOOOOOOOO.....PPPPPPP..............
//.....SSSSSSSSSSSSS.........TTTTTTT..........OOOOOOOOOOOOO......PPPPPPP..............
//.......SSSSSSSSS..............................OOOOOOOOO.............................
//....................................................................................
// THIS FILE IS DEPRECATED PLEASE DO NOT ADD CODE TO IT
function buildPageComponentsInfo(experiments, isDebug) {
    if (experiments === void 0) { experiments = {}; }
    if (isDebug === void 0) { isDebug = false; }
    var _a, _b;
    var wixExperiments = new wix_experiments_1.default({ experiments: experiments });
    var homePageComponent = {
        route: 'home',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Home,
        pageComponentName: 'mini-home-module',
        moduleId: ModuleId_1.ModuleId.Home,
        isEnabled: true
    };
    var storesModule = {
        route: 'store',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Stores,
        pageComponentName: 'ecom',
        moduleId: ModuleId_1.ModuleId.Stores,
        isEnabled: true,
        appDefId: appDefIds_1.appDefIds.wixECommerce
    };
    var settingsPageComponent = {
        route: 'settings',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Settings,
        pageComponentName: 'settings-lazy-module',
        moduleId: ModuleId_1.ModuleId.Settings,
        isEnabled: true
    };
    var siteSettingsPageComponent = {
        route: 'manage-website',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.SiteSettings,
        pageComponentName: 'site-settings-lazy-module',
        moduleId: ModuleId_1.ModuleId.SiteSettings,
        isEnabled: true
    };
    var contactsPageComponent = {
        route: 'contacts',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Contacts,
        pageComponentName: 'contacts-page-component',
        moduleId: ModuleId_1.ModuleId.Contacts,
        isEnabled: true
    };
    var engagePageComponent = {
        route: 'inbox',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Engage,
        pageComponentName: 'engage-crazy-chat-lazy',
        moduleId: ModuleId_1.ModuleId.Engage,
        isEnabled: true,
        appDefId: appDefIds_1.appDefIds.engage
    };
    var shoutoutPageComponent = {
        route: 'shoutout',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Shoutout,
        pageComponentName: 'shoutout',
        moduleId: ModuleId_1.ModuleId.Shoutout,
        isEnabled: true,
        appDefId: appDefIds_1.appDefIds.shoutout
    };
    var invoicesPageComponent = {
        route: 'quotes',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Invoices,
        pageComponentName: 'invoices',
        moduleId: ModuleId_1.ModuleId.Invoices,
        isEnabled: true,
        appDefId: appDefIds_1.appDefIds.invoices
    };
    var cashierPageComponent = {
        route: 'payments',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Cashier,
        pageComponentName: 'cashier-merchant-settings',
        moduleId: ModuleId_1.ModuleId.Cashier,
        isEnabled: true,
        appDefId: appDefIds_1.appDefIds.cashier
    };
    var exampleReactPageComponent = {
        route: 'demo-react-lazy',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Example_React,
        pageComponentName: 'demo-react-lazy',
        moduleId: ModuleId_1.ModuleId.Examples,
        isEnabled: isDebug
    };
    var exampleAngularPageComponent = {
        route: 'demo-angular-lazy',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.Example_Angular,
        pageComponentName: 'demo-angular-lazy',
        moduleId: ModuleId_1.ModuleId.Examples,
        isEnabled: isDebug
    };
    var promoteSeoPageComponent = {
        route: 'promote-seo',
        isMain: true,
        pageComponentId: PageComponentId_1.PageComponentId.PromoteSeo,
        pageComponentName: 'PromoteSeoLazyComponent',
        isEnabled: wixExperiments.enabled('specs.wos.displayPromoteSeo'),
        moduleId: ModuleId_1.ModuleId.PromoteSeo,
        appDefId: appDefIds_1.appDefIds.promoteSeo
    };
    var couponsPageComponent = {
        route: 'coupons',
        isMain: true,
        isEnabled: wixExperiments.enabled('specs.wos.CouponsPlatform'),
        pageComponentId: PageComponentId_1.PageComponentId.Coupons,
        pageComponentName: 'coupons.app',
        moduleId: ModuleId_1.ModuleId.Coupons,
        appDefId: appDefIds_1.appDefIds.coupons
    };
    var restaurantsPageComponent = {
        route: 'restaurants',
        isMain: true,
        isEnabled: true,
        pageComponentId: PageComponentId_1.PageComponentId.Restaurants,
        pageComponentName: 'restaurants',
        moduleId: ModuleId_1.ModuleId.Restaurants,
        appDefId: appDefIds_1.appDefIds.restaurants
    };
    var triggersPageComponent = {
        route: 'triggers',
        isMain: true,
        isEnabled: wixExperiments.enabled('specs.wos.displayTriggers'),
        pageComponentId: PageComponentId_1.PageComponentId.Triggers,
        pageComponentName: 'triggers',
        moduleId: ModuleId_1.ModuleId.Triggers,
        appDefId: appDefIds_1.appDefIds.triggers
    };
    var membersPermissionsPageComponent = {
        route: 'member-permissions',
        isMain: true,
        isEnabled: true,
        pageComponentId: PageComponentId_1.PageComponentId.MemberPermissions,
        pageComponentName: 'member-permissions',
        moduleId: ModuleId_1.ModuleId.MemberPermissions,
        appDefId: appDefIds_1.appDefIds.wixSiteMembers
    };
    var pages = (_a = {},
        _a[PageComponentId_1.PageComponentId.Stores] = storesModule,
        _a[PageComponentId_1.PageComponentId.SiteSettings] = siteSettingsPageComponent,
        _a[PageComponentId_1.PageComponentId.Contacts] = contactsPageComponent,
        _a[PageComponentId_1.PageComponentId.Shoutout] = shoutoutPageComponent,
        _a[PageComponentId_1.PageComponentId.Invoices] = invoicesPageComponent,
        _a[PageComponentId_1.PageComponentId.Cashier] = cashierPageComponent,
        _a[PageComponentId_1.PageComponentId.PromoteSeo] = promoteSeoPageComponent,
        _a[PageComponentId_1.PageComponentId.Example_React] = exampleReactPageComponent,
        _a[PageComponentId_1.PageComponentId.Example_Angular] = exampleAngularPageComponent,
        _a[PageComponentId_1.PageComponentId.Coupons] = couponsPageComponent,
        _a[PageComponentId_1.PageComponentId.Restaurants] = restaurantsPageComponent,
        _a[PageComponentId_1.PageComponentId.Triggers] = triggersPageComponent,
        _a[PageComponentId_1.PageComponentId.MemberPermissions] = membersPermissionsPageComponent,
        _a);
    if (wixExperiments.enabled('specs.wos.DeprecateBuildPageComponentsInfo')) {
        return pages;
    }
    return tslib_1.__assign({}, pages, (_b = {}, _b[PageComponentId_1.PageComponentId.Home] = homePageComponent, _b[PageComponentId_1.PageComponentId.Settings] = settingsPageComponent, _b[PageComponentId_1.PageComponentId.Engage] = engagePageComponent, _b));
}
exports.buildPageComponentsInfo = buildPageComponentsInfo;
