// @wix/my-account-header/dist/es/src/client/components/Header/HeaderLayout/HeaderLayout.js
import { __extends } from "tslib";
import * as React from 'react';
import classnames from 'classnames';
import Logo from '../../Logo/Logo';
import * as css from './HeaderLayout.scss';
import { dataHooks } from '../../../constants/dataHooks';
export var Position;
(function (Position) {
    Position["LEFT"] = "leftSide";
    Position["RIGHT"] = "rightSide";
})(Position || (Position = {}));
var Section = /** @class */ (function (_super) {
    __extends(Section, _super);
    function Section() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Section.prototype.render = function () {
        var sideClass = this.props.position === Position.LEFT ? css.left : css.right;
        var className = classnames([css.section, sideClass, css.sitesDropDown]);
        return (React.createElement("div", { className: className }, this.props.children));
    };
    return Section;
}(React.Component));
export { Section };
var HeaderLayout = /** @class */ (function (_super) {
    __extends(HeaderLayout, _super);
    function HeaderLayout() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HeaderLayout.prototype.render = function () {
        return (React.createElement("div", { "data-hook": dataHooks.Header, className: css.header },
            React.createElement(Logo, null),
            this.props.children));
    };
    return HeaderLayout;
}(React.Component));
export { HeaderLayout };
export var SectionItem = function (_a) {
    var children = _a.children;
    return (React.createElement("span", { className: css.sectionItemComponent }, children));
};
