// @wix/analytics-ng-common/dist/es/src/enums.js
export var PredefinedTimeframe;
(function (PredefinedTimeframe) {
    PredefinedTimeframe["LAST_7_DAYS"] = "LAST_7_DAYS";
    PredefinedTimeframe["LAST_14_DAYS"] = "LAST_14_DAYS";
    PredefinedTimeframe["LAST_30_DAYS"] = "LAST_30_DAYS";
    PredefinedTimeframe["LAST_90_DAYS"] = "LAST_90_DAYS";
    PredefinedTimeframe["THIS_WEEK"] = "THIS_WEEK";
    PredefinedTimeframe["THIS_MONTH"] = "THIS_MONTH";
})(PredefinedTimeframe || (PredefinedTimeframe = {}));
export var MeasureType;
(function (MeasureType) {
    MeasureType["prime"] = "prime";
    MeasureType["money"] = "money";
    MeasureType["time"] = "time";
    MeasureType["percent"] = "percent";
})(MeasureType || (MeasureType = {}));
export var ChartTypes;
(function (ChartTypes) {
    ChartTypes["SUMMARY_BOX"] = "SUMMARY_BOX";
    ChartTypes["SUMMARY_BOX_VELO"] = "SUMMARY_BOX_VELO";
    ChartTypes["SUMMARY_BOX_NEXT"] = "SUMMARY_BOX_NEXT";
})(ChartTypes || (ChartTypes = {}));
export var ServiceProfileFields;
(function (ServiceProfileFields) {
    ServiceProfileFields["hasFBCampignStarted"] = "hasFBCampignStarted";
    ServiceProfileFields["hasSetUpAbandonedCardEmailAutomation"] = "hasSetUpAbandonedCardEmailAutomation";
    ServiceProfileFields["hasFinishedSEOWizard"] = "hasFinishedSEOWizard";
    ServiceProfileFields["hasInstalledStores"] = "hasInstalledStores";
})(ServiceProfileFields || (ServiceProfileFields = {}));
export var StatusFields;
(function (StatusFields) {
    StatusFields["HAS_FREE_PRICING_PLANS"] = "hasFreePricingPlans";
})(StatusFields || (StatusFields = {}));
export var Actions;
(function (Actions) {
    Actions["HIDE_REPORT"] = "HIDE_REPORT";
})(Actions || (Actions = {}));
