// @wix/analytics-ng-common/dist/es/src/reports.js
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
import { appDefIds } from '@wix/business-manager-api';
import { Actions, ChartTypes, MeasureType, ServiceProfileFields, StatusFields } from './enums';
export var reports = [
    {
        reportId: 'traffic_api.sessions_count',
        headerText: 'widget.dashboard.stat.sessions.title',
        theme: 'traffic',
        group: 'traffic',
        reportLink: '0b0118c7-98d3-4ec8-8b42-9d45bc145138',
        measureLabel: 'widget.dashboard.stat.sessions.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.sessions.description',
        smartActions: [
            {
                conditions: [(_a = {}, _a[ServiceProfileFields.hasFinishedSEOWizard] = true, _a)],
                footerSmartActionTitle: 'widget.dashboard.stat.sessions.footer.email-marketing.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.shoutout,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.sessions.footer.email-marketing.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.sessions.footer.email-marketing.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.sessions.footer.email-marketing.tooltip.cta',
                biActionName: 'Create an Email Campaign',
            },
            {
                conditions: [(_b = {}, _b[ServiceProfileFields.hasFinishedSEOWizard] = false, _b)],
                footerSmartActionTitle: 'widget.dashboard.stat.sessions.footer.seo-wiz.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteSeo,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.sessions.footer.seo-wiz.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.sessions.footer.seo-wiz.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.sessions.footer.seo-wiz.tooltip.cta',
                biActionName: 'Go to SEO Wiz',
            },
        ],
        queryData: {
            query: {
                dimensions: ['traffic_api.created_date'],
                measures: ['traffic_api.sessions_count'],
                filters: {},
                sorts: ['traffic_api.created_date'],
                view: 'traffic_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.sessions_count',
        },
        isDashboardReport: true
    },
    {
        reportId: 'traffic_api.visitors_count',
        headerText: 'widget.dashboard.stat.visitors.title',
        theme: 'traffic',
        group: 'traffic',
        reportLink: '0b0118c7-98d3-4ec8-8b42-9d45bc145138',
        measureLabel: 'widget.dashboard.stat.visitors.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.visitors.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.visitors.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.visitors.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.visitors.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.visitors.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['traffic_api.created_date'],
                measures: ['traffic_api.visitors_count'],
                filters: {},
                sorts: ['traffic_api.created_date'],
                view: 'traffic_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.visitors_count',
        },
        isDashboardReport: true
    },
    {
        reportId: 'traffic_api.view_count',
        headerText: 'widget.dashboard.stat.views.title',
        theme: 'traffic',
        group: 'traffic',
        reportLink: '0b0118c7-98d3-4ec8-8b42-9d45bc145138',
        measureLabel: 'widget.dashboard.stat.views.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.views.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.views.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.views.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.views.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.views.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['traffic_api.created_date'],
                measures: ['traffic_api.view_count'],
                filters: {},
                sorts: ['traffic_api.created_date'],
                view: 'traffic_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.view_count',
        },
        isDashboardReport: true
    },
    {
        reportId: 'traffic_api.bounce_rate',
        headerText: 'widget.dashboard.stat.bounce-rate.title',
        theme: 'traffic',
        group: 'traffic',
        reportLink: '0b0118c7-98d3-4ec8-8b42-9d45bc145138',
        measureLabel: 'widget.dashboard.stat.bounce-rate.measure-label',
        measureType: MeasureType.percent,
        description: 'widget.dashboard.stat.bounce-rate.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.bounce-rate.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.bounce-rate.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.bounce-rate.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.bounce-rate.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['traffic_api.created_date'],
                measures: ['traffic_api.bounce_rate'],
                filters: {},
                sorts: ['traffic_api.created_date'],
                view: 'traffic_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.bounce_rate',
        },
        isDashboardReport: true,
        invertedTrend: true,
    },
    {
        reportId: 'traffic_api.site_avg_time',
        headerText: 'widget.dashboard.stat.average-time.title',
        theme: 'traffic',
        group: 'traffic',
        reportLink: '0b0118c7-98d3-4ec8-8b42-9d45bc145138',
        measureLabel: 'widget.dashboard.stat.average-time.measure-label',
        measureType: MeasureType.time,
        description: 'widget.dashboard.stat.average-time.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.average-time.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.average-time.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.average-time.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.average-time.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['traffic_api.created_date'],
                measures: ['traffic_api.site_avg_time'],
                filters: {},
                sorts: ['traffic_api.created_date'],
                view: 'traffic_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.site_avg_time',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.amount',
        headerText: 'widget.dashboard.stat.total-sales.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.total-sales.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.total-sales.description',
        smartActions: [
            {
                conditions: [
                    (_c = {}, _c[ServiceProfileFields.hasInstalledStores] = true, _c),
                    (_d = {}, _d[ServiceProfileFields.hasFBCampignStarted] = true, _d),
                    (_e = {}, _e[ServiceProfileFields.hasSetUpAbandonedCardEmailAutomation] = true, _e),
                ],
                footerSmartActionTitle: 'widget.dashboard.stat.total-sales.footer.coupons.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.coupons,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.total-sales.footer.coupons.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.total-sales.footer.coupons.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.total-sales.footer.coupons.tooltip.cta',
                biActionName: 'Create a Coupon',
            },
            {
                conditions: [
                    (_f = {}, _f[ServiceProfileFields.hasInstalledStores] = true, _f),
                    (_g = {}, _g[ServiceProfileFields.hasFBCampignStarted] = true, _g),
                    (_h = {},
                        _h[ServiceProfileFields.hasSetUpAbandonedCardEmailAutomation] = false,
                        _h),
                ],
                footerSmartActionTitle: 'widget.dashboard.stat.total-sales.footer.abandoned-cart.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.triggers,
                    contextData: { appState: '?recipe=abandoned_cart&referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.total-sales.footer.abandoned-cart.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.total-sales.footer.abandoned-cart.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.total-sales.footer.abandoned-cart.tooltip.cta',
                biActionName: 'Go to Automations',
            },
            {
                conditions: [
                    (_j = {}, _j[ServiceProfileFields.hasInstalledStores] = true, _j),
                    (_k = {}, _k[ServiceProfileFields.hasFBCampignStarted] = false, _k),
                ],
                footerSmartActionTitle: 'widget.dashboard.stat.total-sales.footer.facebook-ads.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promotePaidAds,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.total-sales.footer.facebook-ads.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.total-sales.footer.facebook-ads.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.total-sales.footer.facebook-ads.tooltip.cta',
                biActionName: 'Create a Facebook Ad',
            },
            {
                conditions: [(_l = {}, _l[ServiceProfileFields.hasInstalledStores] = false, _l)],
                footerSmartActionTitle: 'widget.dashboard.stat.total-sales.footer.marketing.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.total-sales.footer.marketing.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.total-sales.footer.marketing.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.total-sales.footer.marketing.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.amount'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.amount',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.orders',
        headerText: 'widget.dashboard.stat.total-orders.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.total-orders.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.total-orders.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.total-orders.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.total-orders.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.total-orders.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.total-orders.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.orders'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.orders',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.revenue',
        headerText: 'widget.dashboard.stat.revenue.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.revenue.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.revenue.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.revenue.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.revenue.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.revenue.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.revenue.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.revenue'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.revenue',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.items_price_total',
        headerText: 'widget.dashboard.stat.items-total-price.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.items-total-price.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.items-total-price.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.items-total-price.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.items-total-price.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.items-total-price.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.items-total-price.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.items_price_total'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.items_price_total',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.refunds',
        headerText: 'widget.dashboard.stat.refunds.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.refunds.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.refunds.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.refunds.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.refunds.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.refunds.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.refunds.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.refunds'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.refunds',
        },
        isDashboardReport: true,
        invertedTrend: true
    },
    {
        reportId: 'transactions_api.net_revenue',
        headerText: 'widget.dashboard.stat.net-revenue.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.net-revenue.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.net-revenue.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.net-revenue.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.net-revenue.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.net-revenue.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.net-revenue.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.net_revenue'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.net_revenue',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.quantity',
        headerText: 'widget.dashboard.stat.quantity.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.quantity.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.quantity.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.quantity.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.quantity.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.quantity.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.quantity.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.quantity'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.quantity',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.average_order_value',
        headerText: 'widget.dashboard.stat.avg-order-value.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '6c7a6f39-5691-4883-980f-4c5c343c6ec3',
        measureLabel: 'widget.dashboard.stat.avg-order-value.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.avg-order-value.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.avg-order-value.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.avg-order-value.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.avg-order-value.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.avg-order-value.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.average_order_value'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.average_order_value',
        },
        isDashboardReport: true
    },
    {
        reportId: 'people_api.active_leads',
        headerText: 'widget.dashboard.stat.active-leads.title',
        theme: 'people',
        group: 'people',
        reportLink: '4deca0dd-ef78-4b93-90c3-8ded012faaf7',
        measureLabel: 'widget.dashboard.stat.active-leads.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.active-leads.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.active-leads.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.crmHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.active-leads.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.active-leads.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.active-leads.footer.tooltip.cta',
                biActionName: 'See More Tips',
            },
        ],
        queryData: {
            query: {
                dimensions: ['people_api.created_date'],
                measures: ['people_api.active_leads'],
                filters: {},
                sorts: ['people_api.created_date'],
                view: 'people_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.active_leads',
        },
        isDashboardReport: true
    },
    {
        reportId: 'transactions_api.customers_count',
        headerText: 'widget.dashboard.stat.customers-count.title',
        theme: 'people',
        group: 'people',
        reportLink: '4deca0dd-ef78-4b93-90c3-8ded012faaf7',
        measureLabel: 'widget.dashboard.stat.customers-count.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.customers-count.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.customers-count.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.crmHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.customers-count.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.customers-count.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.customers-count.footer.tooltip.cta',
                biActionName: 'Go to Customer Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['transactions_api.created_date'],
                measures: ['transactions_api.customers_count'],
                filters: {},
                sorts: ['transactions_api.created_date'],
                view: 'transactions_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.customers_count',
        },
        isDashboardReport: true
    },
    {
        reportId: 'people_api.contacts_count',
        headerText: 'widget.dashboard.stat.contacts-count.title',
        theme: 'people',
        group: 'people',
        reportLink: '4deca0dd-ef78-4b93-90c3-8ded012faaf7',
        measureLabel: 'widget.dashboard.stat.contacts-count.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.contacts-count.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.contacts-count.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.crmHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.contacts-count.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.contacts-count.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.contacts-count.footer.tooltip.cta',
                biActionName: 'Take Me There',
            },
        ],
        queryData: {
            query: {
                dimensions: ['people_api.created_date'],
                measures: ['people_api.members'],
                filters: {},
                sorts: ['people_api.created_date'],
                view: 'people_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.members',
        },
        isDashboardReport: true
    },
    {
        reportId: 'blog_api.view_count',
        headerText: 'widget.dashboard.stat.post-views.title',
        theme: 'traffic',
        group: 'blog',
        reportLink: '477c482b-be24-4f8e-ab01-b4eeabe8c4f5',
        measureLabel: 'widget.dashboard.stat.post-views.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.post-views.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.post-views.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.post-views.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.post-views.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.post-views.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['blog_api.created_date'],
                measures: ['blog_api.post_view_count'],
                filters: {},
                sorts: ['blog_api.created_date'],
                view: 'blog_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.post_view_count',
        },
        isDashboardReport: true
    },
    {
        reportId: 'bookings_api.bookings',
        headerText: 'widget.dashboard.stat.bookings.title',
        theme: 'people',
        group: 'bookings',
        reportLink: '788ac97c-f71e-4c4f-b5c0-8bdcaaf956fd',
        measureLabel: 'widget.dashboard.stat.bookings.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.bookings.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.bookings.footer.cta',
                footerNavigateConfig: {
                    pageComponentId: appDefIds.promoteHome,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                footerTooltipTitle: 'widget.dashboard.stat.bookings.footer.tooltip.title',
                footerTooltipText: 'widget.dashboard.stat.bookings.footer.tooltip.text',
                footerTooltipCta: 'widget.dashboard.stat.bookings.footer.tooltip.cta',
                biActionName: 'See Marketing Tools',
            },
        ],
        queryData: {
            query: {
                dimensions: ['bookings_api.created_date'],
                measures: ['bookings_api.confirmed_bookings'],
                filters: {},
                sorts: ['bookings_api.created_date'],
                view: 'bookings_api',
            },
        },
        unifiedApi: {
            dataView: 'unified_api',
            measure: 'unified_api.confirmed_bookings',
        },
        isDashboardReport: true
    },
    {
        reportId: 'pricing_plans_api.plans_purchased',
        headerText: 'widget.dashboard.stat.pricing-plans-purchased.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '',
        measureLabel: 'widget.dashboard.stat.pricing-plans-purchased.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.pricing-plans-purchased.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.pricing-plans-purchased.cta',
                ctaNavigateConfig: {
                    pageComponentId: appDefIds.subscriptions,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                addTimePeriod: false,
                biActionName: '',
            }
        ],
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.pp_orders',
        },
        isDashboardReport: false
    },
    {
        reportId: 'pricing_plans_api.plans_sales',
        headerText: 'widget.dashboard.stat.pricing-plans-sales.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '',
        measureLabel: 'widget.dashboard.stat.pricing-plans-sales.measure-label',
        measureType: MeasureType.money,
        description: 'widget.dashboard.stat.pricing-plans-sales.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.pricing-plans-sales.cta',
                ctaNavigateConfig: {
                    pageComponentId: 'analytics-ng-bm',
                    contextData: { appState: "3147631d-e0ad-4858-9e85-cfd3ef24990d_advanced?choose_measure=revenue&comparison_period=previousPeriod&date_granularity=day&source_app=Paid%20Plans" },
                },
                addTimePeriod: true,
                biActionName: '',
            }
        ],
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.pp_amount',
        },
        isDashboardReport: false
    },
    {
        reportId: 'pricing_plans_api.free_plans',
        headerText: 'widget.dashboard.stat.pricing-plans-free.title',
        theme: 'sales',
        group: 'sales',
        reportLink: '',
        measureLabel: 'widget.dashboard.stat.pricing-plans-free.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.pricing-plans-free.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.pricing-plans-free.cta',
                ctaNavigateConfig: {
                    pageComponentId: appDefIds.subscriptions,
                    contextData: { appState: '?referralInfo=analytics_BM_widget' },
                },
                addTimePeriod: false,
                biActionName: '',
            }
        ],
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.pp_free_orders',
        },
        isDashboardReport: false
    },
    {
        reportId: 'velo.request_count',
        headerText: 'widget.dashboard.stat.velo-requests.title',
        theme: 'traffic',
        group: 'velo',
        reportLink: '',
        measureLabel: 'widget.dashboard.stat.velo-requests.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.velo-requests.description',
        smartActions: [],
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.request_count',
        },
        isDashboardReport: false
    },
    {
        reportId: 'velo.rpm_count',
        headerText: 'widget.dashboard.stat.velo-throttled-requests.title',
        theme: 'traffic',
        group: 'velo',
        reportLink: '',
        measureLabel: 'widget.dashboard.stat.velo-throttled-requests.measure-label',
        measureType: MeasureType.prime,
        description: 'widget.dashboard.stat.velo-throttled-requests.description',
        smartActions: [
            {
                conditions: [],
                footerSmartActionTitle: 'widget.dashboard.stat.velo-throttled-requests.cta',
                ctaNavigateConfig: 'https://support.wix.com/article/velo-about-backend-limitations',
                addTimePeriod: false,
                biActionName: 'velo-throttled-requests-cta',
            },
        ],
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.rpm_count',
        },
        isDashboardReport: false,
        invertedTrend: true,
        onlyZeroIsGood: true,
    },
];
export var requirements = (_m = {},
    _m[StatusFields.HAS_FREE_PRICING_PLANS] = {
        unifiedApi: {
            dataView: 'unified_api_embed',
            measure: 'unified_api.pp_free_orders',
        },
        cache: true
    },
    _m);
export var embeddedWidgets = {
    'pricing_plans': {
        requirements: [StatusFields.HAS_FREE_PRICING_PLANS],
        reports: ['pricing_plans_api.plans_sales', 'pricing_plans_api.plans_purchased', 'pricing_plans_api.free_plans'],
        reportKey: 'pricing_plans',
        actions: [{
                conditions: (_o = {},
                    _o[StatusFields.HAS_FREE_PRICING_PLANS] = false,
                    _o),
                action: Actions.HIDE_REPORT,
                payload: 'pricing_plans_api.free_plans'
            }],
        title: 'widget.dashboard.pricing-plans',
        visualisationType: ChartTypes.SUMMARY_BOX,
    },
    'velo': {
        requirements: [],
        reports: ['velo.request_count', 'velo.rpm_count'],
        reportKey: 'velo',
        actions: [],
        title: 'widget.dashboard.velo',
        visualisationType: ChartTypes.SUMMARY_BOX_VELO,
    }
};
export var PeopleApiName = 'people_api';
export var SalesApiName = 'transactions_api';
export var TrafficApiName = 'traffic_api';
