// @wix/ambassador-shoutout-editor/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionType = exports.Type = exports.TranslationState = exports.TemplateType = exports.TemplateState = exports.StructNullValue = exports.SortOrder = exports.RuleStatus = exports.RestrictionAction = exports.RejectionReasonEnum = exports.PreconditionType = exports.PlaceholderType = exports.Operator = exports.MessageVisibilityStatus = exports.MessageDistributionStatus = exports.ComplianceStatus = exports.CampaignVisibilityStatusEnum = exports.CampaignTypeEnum = exports.CampaignTranslationStatusEnum = exports.CampaignStatusEnum = exports.CampaignEditorTypeEnum = exports.CampaignDistributionStatusEnum = exports.AccountStatusEnum = void 0;
var AccountStatusEnum;
(function (AccountStatusEnum) {
    AccountStatusEnum["UNKNOWN"] = "UNKNOWN";
    AccountStatusEnum["SUSPENDED"] = "SUSPENDED";
    AccountStatusEnum["WARNED"] = "WARNED";
    AccountStatusEnum["BANNED"] = "BANNED";
    AccountStatusEnum["ACTIVE"] = "ACTIVE";
})(AccountStatusEnum = exports.AccountStatusEnum || (exports.AccountStatusEnum = {}));
var CampaignDistributionStatusEnum;
(function (CampaignDistributionStatusEnum) {
    CampaignDistributionStatusEnum["SAMPLING"] = "SAMPLING";
    CampaignDistributionStatusEnum["NOT_STARTED"] = "NOT_STARTED";
    CampaignDistributionStatusEnum["TERMINATED"] = "TERMINATED";
    CampaignDistributionStatusEnum["UNKNOWN"] = "UNKNOWN";
    CampaignDistributionStatusEnum["SCHEDULED"] = "SCHEDULED";
    CampaignDistributionStatusEnum["IN_DETECTION"] = "IN_DETECTION";
    CampaignDistributionStatusEnum["IN_MODERATION"] = "IN_MODERATION";
    CampaignDistributionStatusEnum["SENDING"] = "SENDING";
    CampaignDistributionStatusEnum["PAUSED"] = "PAUSED";
    CampaignDistributionStatusEnum["REJECTED"] = "REJECTED";
    CampaignDistributionStatusEnum["DISTRIBUTED"] = "DISTRIBUTED";
})(CampaignDistributionStatusEnum = exports.CampaignDistributionStatusEnum || (exports.CampaignDistributionStatusEnum = {}));
var CampaignEditorTypeEnum;
(function (CampaignEditorTypeEnum) {
    CampaignEditorTypeEnum["MOBILE"] = "MOBILE";
    CampaignEditorTypeEnum["UNKNOWN"] = "UNKNOWN";
    CampaignEditorTypeEnum["WEB"] = "WEB";
})(CampaignEditorTypeEnum = exports.CampaignEditorTypeEnum || (exports.CampaignEditorTypeEnum = {}));
var CampaignStatusEnum;
(function (CampaignStatusEnum) {
    CampaignStatusEnum["DELETED"] = "DELETED";
    CampaignStatusEnum["ACTIVE"] = "ACTIVE";
    CampaignStatusEnum["ARCHIVED"] = "ARCHIVED";
    CampaignStatusEnum["UNKNOWN"] = "UNKNOWN";
})(CampaignStatusEnum = exports.CampaignStatusEnum || (exports.CampaignStatusEnum = {}));
var CampaignTranslationStatusEnum;
(function (CampaignTranslationStatusEnum) {
    CampaignTranslationStatusEnum["DRAFT"] = "DRAFT";
    CampaignTranslationStatusEnum["PUBLISHED"] = "PUBLISHED";
})(CampaignTranslationStatusEnum = exports.CampaignTranslationStatusEnum || (exports.CampaignTranslationStatusEnum = {}));
var CampaignTypeEnum;
(function (CampaignTypeEnum) {
    CampaignTypeEnum["EMAIL_MARKETING"] = "EMAIL_MARKETING";
    CampaignTypeEnum["AUTOMATION"] = "AUTOMATION";
    CampaignTypeEnum["UNKNOWN"] = "UNKNOWN";
    CampaignTypeEnum["INVITATION"] = "INVITATION";
})(CampaignTypeEnum = exports.CampaignTypeEnum || (exports.CampaignTypeEnum = {}));
var CampaignVisibilityStatusEnum;
(function (CampaignVisibilityStatusEnum) {
    CampaignVisibilityStatusEnum["TEMPLATE"] = "TEMPLATE";
    CampaignVisibilityStatusEnum["UNKNOWN"] = "UNKNOWN";
    CampaignVisibilityStatusEnum["DRAFT"] = "DRAFT";
    CampaignVisibilityStatusEnum["PUBLISHED"] = "PUBLISHED";
})(CampaignVisibilityStatusEnum = exports.CampaignVisibilityStatusEnum || (exports.CampaignVisibilityStatusEnum = {}));
var ComplianceStatus;
(function (ComplianceStatus) {
    ComplianceStatus["Active"] = "Active";
    ComplianceStatus["Warned"] = "Warned";
    ComplianceStatus["Suspended"] = "Suspended";
    ComplianceStatus["Banned"] = "Banned";
    ComplianceStatus["LimitedMultiSendables"] = "LimitedMultiSendables";
    ComplianceStatus["AutoliftSuspended"] = "AutoliftSuspended";
})(ComplianceStatus = exports.ComplianceStatus || (exports.ComplianceStatus = {}));
var MessageDistributionStatus;
(function (MessageDistributionStatus) {
    MessageDistributionStatus["NotStarted"] = "NotStarted";
    MessageDistributionStatus["Scheduled"] = "Scheduled";
    MessageDistributionStatus["InDetection"] = "InDetection";
    MessageDistributionStatus["InModeration"] = "InModeration";
    MessageDistributionStatus["Sampling"] = "Sampling";
    MessageDistributionStatus["Sending"] = "Sending";
    MessageDistributionStatus["Paused"] = "Paused";
    MessageDistributionStatus["Rejected"] = "Rejected";
    MessageDistributionStatus["Terminated"] = "Terminated";
    MessageDistributionStatus["Distributed"] = "Distributed";
})(MessageDistributionStatus = exports.MessageDistributionStatus || (exports.MessageDistributionStatus = {}));
var MessageVisibilityStatus;
(function (MessageVisibilityStatus) {
    MessageVisibilityStatus["Draft"] = "Draft";
    MessageVisibilityStatus["Published"] = "Published";
    MessageVisibilityStatus["Template"] = "Template";
})(MessageVisibilityStatus = exports.MessageVisibilityStatus || (exports.MessageVisibilityStatus = {}));
var Operator;
(function (Operator) {
    Operator["UNDEFINED"] = "UNDEFINED";
    Operator["IN"] = "IN";
})(Operator = exports.Operator || (exports.Operator = {}));
var PlaceholderType;
(function (PlaceholderType) {
    PlaceholderType["Text"] = "Text";
    PlaceholderType["Html"] = "Html";
    PlaceholderType["LiveSiteLink"] = "LiveSiteLink";
    PlaceholderType["BackofficeLink"] = "BackofficeLink";
    PlaceholderType["Money"] = "Money";
    PlaceholderType["DateTime"] = "DateTime";
    PlaceholderType["DateTimeWithTimezone"] = "DateTimeWithTimezone";
    PlaceholderType["Enum"] = "Enum";
})(PlaceholderType = exports.PlaceholderType || (exports.PlaceholderType = {}));
var PreconditionType;
(function (PreconditionType) {
    PreconditionType["UNKNOWN_PRECONDITION_TYPE"] = "UNKNOWN_PRECONDITION_TYPE";
    PreconditionType["PETRI"] = "PETRI";
    PreconditionType["STATE"] = "STATE";
    PreconditionType["APP_DEPENDENCY"] = "APP_DEPENDENCY";
    PreconditionType["DEALER_OFFERING"] = "DEALER_OFFERING";
})(PreconditionType = exports.PreconditionType || (exports.PreconditionType = {}));
var RejectionReasonEnum;
(function (RejectionReasonEnum) {
    RejectionReasonEnum["UNKNOWN"] = "UNKNOWN";
    RejectionReasonEnum["LOW_ENGAGEMENT"] = "LOW_ENGAGEMENT";
    RejectionReasonEnum["SENDER_DETAILS"] = "SENDER_DETAILS";
    RejectionReasonEnum["OTHER"] = "OTHER";
    RejectionReasonEnum["SPAM_TRAPS"] = "SPAM_TRAPS";
})(RejectionReasonEnum = exports.RejectionReasonEnum || (exports.RejectionReasonEnum = {}));
var RestrictionAction;
(function (RestrictionAction) {
    RestrictionAction["Warn"] = "Warn";
    RestrictionAction["Suspend"] = "Suspend";
    RestrictionAction["Ban"] = "Ban";
    RestrictionAction["LimitMultiSendables"] = "LimitMultiSendables";
    RestrictionAction["AutoliftSuspend"] = "AutoliftSuspend";
    RestrictionAction["AgreeToTerms"] = "AgreeToTerms";
    RestrictionAction["OpenTicket"] = "OpenTicket";
    RestrictionAction["FillForm"] = "FillForm";
    RestrictionAction["AnswerQuestion"] = "AnswerQuestion";
    RestrictionAction["FillAutoliftForm"] = "FillAutoliftForm";
    RestrictionAction["LiftWarning"] = "LiftWarning";
    RestrictionAction["LiftSuspension"] = "LiftSuspension";
    RestrictionAction["LiftBan"] = "LiftBan";
    RestrictionAction["LiftLimitForMultiSendables"] = "LiftLimitForMultiSendables";
    RestrictionAction["LiftAutoliftSuspension"] = "LiftAutoliftSuspension";
})(RestrictionAction = exports.RestrictionAction || (exports.RestrictionAction = {}));
var RuleStatus;
(function (RuleStatus) {
    RuleStatus["UnknownStatus"] = "UnknownStatus";
    RuleStatus["Active"] = "Active";
    RuleStatus["Suspended"] = "Suspended";
})(RuleStatus = exports.RuleStatus || (exports.RuleStatus = {}));
var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder = exports.SortOrder || (exports.SortOrder = {}));
var StructNullValue;
(function (StructNullValue) {
    StructNullValue["NULL_VALUE"] = "NULL_VALUE";
})(StructNullValue = exports.StructNullValue || (exports.StructNullValue = {}));
var TemplateState;
(function (TemplateState) {
    TemplateState["UNKNOWN_TEMPLATE_STATUS"] = "UNKNOWN_TEMPLATE_STATUS";
    TemplateState["INTERNAL"] = "INTERNAL";
    TemplateState["PUBLIC"] = "PUBLIC";
    TemplateState["ARCHIVED"] = "ARCHIVED";
    TemplateState["LIVE"] = "LIVE";
})(TemplateState = exports.TemplateState || (exports.TemplateState = {}));
var TemplateType;
(function (TemplateType) {
    TemplateType["UNKNOWN_TEMPLATE_TYPE"] = "UNKNOWN_TEMPLATE_TYPE";
    TemplateType["PREINSTALLED"] = "PREINSTALLED";
    TemplateType["RECIPE"] = "RECIPE";
})(TemplateType = exports.TemplateType || (exports.TemplateType = {}));
var TranslationState;
(function (TranslationState) {
    TranslationState["UNKNOWN_TRANSLATION_STATE"] = "UNKNOWN_TRANSLATION_STATE";
    TranslationState["NOT_REQUESTED"] = "NOT_REQUESTED";
    TranslationState["PENDING"] = "PENDING";
    TranslationState["PARTIAL"] = "PARTIAL";
    TranslationState["TRANSLATED"] = "TRANSLATED";
})(TranslationState = exports.TranslationState || (exports.TranslationState = {}));
var Type;
(function (Type) {
    Type["HTML"] = "HTML";
    Type["ENUM"] = "ENUM";
    Type["LIVE_SITE_LINK"] = "LIVE_SITE_LINK";
    Type["DATETIME"] = "DATETIME";
    Type["PLAIN_TEXT"] = "PLAIN_TEXT";
    Type["UNKNOWN"] = "UNKNOWN";
    Type["MONEY"] = "MONEY";
    Type["BACKOFFICE_LINK"] = "BACKOFFICE_LINK";
    Type["DATETIME_WITH_TIMEZONE"] = "DATETIME_WITH_TIMEZONE";
})(Type = exports.Type || (exports.Type = {}));
var VersionType;
(function (VersionType) {
    VersionType["UNKNOWN_VERSION_TYPE"] = "UNKNOWN_VERSION_TYPE";
    VersionType["MINOR"] = "MINOR";
    VersionType["MAJOR"] = "MAJOR";
})(VersionType = exports.VersionType || (exports.VersionType = {}));
