// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/MigrateToEditorX/getMigrateSiteToEditorXSiteAction.js
import * as React from 'react';
import { DialogNames, ActionType } from '../../../../types';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { EditorType } from '@wix/business-manager-api';
import { MigrateToEditorXSiteAction } from './MigrateToEdtiroXSiteAction';
import EditorXSmall from 'wix-ui-icons-common/EditorXSmall';
export var getMigrateToEditorXSiteAction = function (quickActionsData, experiments, dealerData) {
    var editorType = quickActionsData.editorType, isOwner = quickActionsData.isOwner;
    var isAllowedEditor = editorType === EditorType.EDITOR;
    var enableMigrateToEditorX = experiments['specs.wos1.EnableMigrateToEditorX'] === 'true';
    var redesignExperiment = experiments['specs.wossm.ImproveUXofSiteActions'] === 'true';
    var shouldShowMigrateToEditorX = enableMigrateToEditorX && isAllowedEditor && isOwner && dealerData.userIsRegisteredToEditorX;
    return {
        type: ActionType.dialog,
        dialog: {
            dialogName: DialogNames.migrateToEditorX,
            component: MigrateToEditorXSiteAction,
            biName: 'quick_actions.Rebuild_site_on_editor_x'
        },
        actionName: 'rebuild-site-on-editorx',
        textKey: redesignExperiment ? 'quickActions.rebuild-on-editorx' : 'quickActions.rebuild-site-on-editorx',
        //TODO: replace icon once supplied by UX
        icon: redesignExperiment ? React.createElement(EditorXSmall, null) : React.createElement("div", null),
        isPermitted: PermissionMethods.always,
        visible: shouldShowMigrateToEditorX,
    };
};
