// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/EditSite/getEditSiteAction.js
import * as React from 'react';
import * as uuid from 'uuid/v4';
import EditSmall from 'wix-ui-icons-common/EditSmall';
import { TOrigin, ActionType } from '../../../../types';
import { PERMISSIONS } from '../../../../consts/permissions';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { BIEvents, BISources } from '../../../../consts/bi';
import { sendSiteActionsBI } from '../../../../services/BIService';
import { EditorType } from '@wix/business-manager-api';
function onEditSiteClick(params) {
    var origin = params.origin, _a = params.quickActionsData, isAdi = _a.isAdi, msid = _a.metaSiteId;
    var esi = uuid();
    sendSiteActionsBI({
        evid: BIEvents.quickActionClicked,
        name: 'quick_actions.edit_site',
        msid: msid,
        origin: origin
    });
    sendSiteActionsBI({
        src: BISources.myAccount,
        evid: BIEvents.openEditor,
        msid: msid,
        hosting: 'MA',
        place: 'quick actions',
        ds_origin: isAdi ? 'ADI' : 'editor',
        origin: origin
    });
    return window.open("/editor/" + msid + "?editorSessionId=" + esi + "&referralInfo=" + origin);
}
export var getEditSiteAction = function (quickActionsData, origin) {
    return {
        type: ActionType.click,
        actionName: 'edit-site',
        textKey: 'quickActions.edit_site',
        icon: React.createElement(EditSmall, null),
        isPermitted: PermissionMethods.withPermissions(PERMISSIONS.EDIT_SITE),
        visible: isVisible(quickActionsData, origin),
        onActionClick: onEditSiteClick,
    };
};
var isVisible = function (quickActionsData, origin) {
    var editorType = quickActionsData.editorType, isPro = quickActionsData.isPro;
    var isEditorX = editorType === EditorType.EDITORX;
    var isDashboardOrigin = origin === TOrigin.Dashboard;
    return isDashboardOrigin ? isEditorX : isPro;
};
