// @wix/chat-sdk/es/src/modules/unseen-chatroom-count/api.unseen-chatroom-count.js
import { getUnseenChatroomCount } from './selector.unseen-chatroom-count';
import _isEqual from 'lodash/isEqual';
import ConnectivityService from '../../services/connectivity.service';
var UnseenChatroomCountApi = /** @class */ (function () {
    function UnseenChatroomCountApi(_a) {
        var store = _a.store, actions = _a.actions;
        this.store = store;
        this.actions = actions;
    }
    UnseenChatroomCountApi.prototype.subscribe = function (callback, filter) {
        var _this = this;
        var prevRes;
        var storeCallback = function () {
            var state = _this.store.getState();
            var res = getUnseenChatroomCount(state, filter);
            if (res && !_isEqual(prevRes, res)) {
                prevRes = res;
                callback(res);
            }
        };
        storeCallback();
        ConnectivityService.waitForOnline.then(function () { return _this.store.dispatch(_this.actions.updateUnseenChatrooms(filter)); });
        return {
            unsubscribe: this.store.subscribe(storeCallback),
        };
    };
    UnseenChatroomCountApi.prototype.markChatroomsAsSeen = function (filter) {
        return this.store.dispatch(this.actions.markChatroomsAsSeen(filter));
    };
    return UnseenChatroomCountApi;
}());
export { UnseenChatroomCountApi };
