// @wix/paywall-universal/dist/esm/banner/templates/default.js
export default {
    blocks: [
        {
            key: 'agp7s',
            text: '{{want_more}}',
            type: 'header-three',
            depth: 0,
            inlineStyleRanges: [{ offset: 0, length: 13, style: 'BOLD' }],
            entityRanges: [],
            data: { textAlignment: 'center' },
        },
        {
            key: 'cv2g0',
            text: '{{text}}\n',
            type: 'unstyled',
            depth: 0,
            inlineStyleRanges: [],
            entityRanges: [],
            data: { textAlignment: 'center' },
        },
        {
            key: 'c3gvh',
            text: ' ',
            type: 'atomic',
            depth: 0,
            inlineStyleRanges: [],
            entityRanges: [{ offset: 0, length: 1, key: 0 }],
            data: {},
        },
    ],
    entityMap: {
        '0': {
            type: 'wix-draft-plugin-action-button',
            mutability: 'IMMUTABLE',
            data: {
                config: { alignment: 'center', size: 'small', width: 'fit-content' },
                button: {
                    settings: {
                        buttonText: '{{subscribe}}',
                        url: '',
                        rel: false,
                        target: true,
                    },
                    design: {
                        activeButton: 0,
                        borderRadius: 0,
                        borderWidth: 0,
                        padding: 12,
                        background: '#0261FF',
                        color: '#FFFFFF',
                        borderColor: '#0261FF',
                    },
                },
            },
        },
    },
    VERSION: '7.7.1',
};
