// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/GoToFolder/getGoToFolderSiteAction.js
import * as React from 'react';
import { ActionType } from '../../../../types';
import FolderFindSmall from 'wix-ui-icons-common/FolderFindSmall';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { sendOpenDialogBI } from '../../LazyActionsComponent/component';
var GoToFolderBiName = 'quick_actions.go_to_folder';
function onGoToFolderClick(_a) {
    var origin = _a.origin, onDialogActionComplete = _a.onDialogActionComplete, _b = _a.quickActionsData, metaSiteId = _b.metaSiteId, foldersInfo = _b.foldersInfo;
    sendOpenDialogBI(GoToFolderBiName, metaSiteId, origin);
    var dialogResult = { actionName: 'go-to-folder', metaSiteId: metaSiteId, result: foldersInfo };
    onDialogActionComplete(dialogResult);
}
export var getGoToFolderSiteAction = function (quickActionsData) {
    var _a;
    var isInFilter = Boolean((_a = quickActionsData.foldersInfo) === null || _a === void 0 ? void 0 : _a.filter);
    return {
        type: ActionType.click,
        actionName: 'go-to-folder',
        textKey: 'quickActions.go-to-folder',
        icon: React.createElement(FolderFindSmall, null),
        isPermitted: PermissionMethods.always,
        visible: isInFilter,
        onActionClick: onGoToFolderClick,
    };
};
