// @wix/my-account-header/dist/es/src/client/components/Header/EditorX/EditorXHeader.js
import { __extends } from "tslib";
import * as React from 'react';
import * as css from './EditorXHeader.scss';
import { ModuleRegistry } from 'react-module-container';
import Help from '../../Help/Help';
import { withTranslation } from '@wix/wix-i18n-config';
import { ON_CLOSE_IN_EDITOR } from '@wix/business-manager-api';
import { withHeaderProvider } from '../../HeaderProvider/HeaderProvider';
import { Text } from 'wix-style-react';
import { ThemeProvider } from 'wix-style-react';
import { theme } from 'wix-style-react/themes/editorX';
import { dataHooks } from '../../../constants/dataHooks';
var EditorXHeaderComponentClass = /** @class */ (function (_super) {
    __extends(EditorXHeaderComponentClass, _super);
    function EditorXHeaderComponentClass(props) {
        var _this = _super.call(this, props) || this;
        _this.notifyAndClose = function (event) { return _this.notifyAndCloseIfNotPrevented(event); };
        return _this;
    }
    EditorXHeaderComponentClass.prototype.notifyAndCloseIfNotPrevented = function (event) {
        ModuleRegistry.notifyListeners(ON_CLOSE_IN_EDITOR, { event: event });
        if (!event.defaultPrevented) {
            this.closeBizMgr();
        }
    };
    /**
     * MEMO: if there will be more postMessages to send to the Editor - extract to a SDK
     */
    EditorXHeaderComponentClass.prototype.closeBizMgr = function () {
        parent.postMessage('{"intent":"TPA2", "action":"CLOSE", "type":"closeWindow", "namespace": "Wix", "data": {}}', '*');
    };
    EditorXHeaderComponentClass.prototype.render = function () {
        return (React.createElement(ThemeProvider, { theme: theme() },
            React.createElement("header", { className: css.header, "data-hook": dataHooks.Header },
                React.createElement(Text, { className: css.title, weight: "thin", size: "medium", light: true, dataHook: "header-title" }, this.props.t('header.editor-title')),
                React.createElement("div", { className: css.actions },
                    React.createElement("div", { className: css.helpButton },
                        React.createElement(Help, { iconClassName: css.helpIconEditor, locale: this.props.locale, useIcon: true })),
                    React.createElement("a", { "data-hook": "close-from-editor", onClick: this.notifyAndClose, className: css.button },
                        React.createElement("div", { className: css.closeButtonContent }))))));
    };
    return EditorXHeaderComponentClass;
}(React.Component));
export var EditorXHeaderComponent = withTranslation()(EditorXHeaderComponentClass);
export var EditorXHeader = withHeaderProvider(EditorXHeaderComponent, false);
export default EditorXHeader;
