// @wix/sites-list-api/dist/es/src/client/client.js
import { __assign } from "tslib";
import axios from 'axios';
import { wixAxiosInstanceConfig } from '@wix/wix-axios-config';
var SitesListApi = /** @class */ (function () {
    function SitesListApi(baseURL) {
        if (baseURL === void 0) { baseURL = '/'; }
        this.axios = wixAxiosInstanceConfig(axios, { baseURL: baseURL });
    }
    SitesListApi.prototype.searchSites = function (params, config) {
        return this.axios.get('/account/sites/api/sites/search', __assign(__assign({}, config), { params: params }));
    };
    SitesListApi.prototype.getSite = function (msid, config) {
        return this.axios.get("/_api/sites-list/sites/" + msid, __assign({}, config));
    };
    SitesListApi.prototype.getSiteRoles = function (msid, config) {
        return this.axios.get("/_api/sites-list/sites/" + msid + "/roles", __assign({}, config));
    };
    SitesListApi.prototype.setRecentlyViewedSites = function (msid, config) {
        return this.axios.post("/_api/sites-list/sites/" + msid + "/setRecentlyViewed", {}, __assign({}, config));
    };
    return SitesListApi;
}());
export { SitesListApi };
