// @wix/my-account-header/dist/es/src/client/components/HelpProDropdown/HelpProDropdown.container.js
import { connect } from 'react-redux';
import { getLanguage, getUserId } from '../../selectors/currentUserSelectors';
import { getDashboredWidgetStaticsVersion } from '../../selectors/topologySelectors';
import HelpProDropdown from './HelpProDropdown';
import { getCurrentSiteId } from '../../selectors/currentSiteSelectors';
import { getCurrentPage } from '../../selectors/currentPageSelectors';
import { getDebugStatus } from '../../selectors/debugSelectors';
import { isAccountLevelAboveExplorer } from '../../selectors/expertsSelectors';
import { isEditorXBrand } from '../../selectors/isEditorXSelector';
import { experimentsSelectors } from '@wix/wix-experiments-redux';
var mapStateToProps = function (state) {
    return ({
        userId: getUserId(state),
        locale: getLanguage(state),
        dashboardAnswersWidgetStaticsVersion: getDashboredWidgetStaticsVersion(state),
        msid: getCurrentSiteId(state),
        origin: getCurrentPage(state),
        isDebug: getDebugStatus(state),
        showPioritySupportFooter: isAccountLevelAboveExplorer(state),
        isEditorX: isEditorXBrand(state),
        showNewAnswersWidget: experimentsSelectors(state).enabled('specs.wos1.smartAnswersWidget'),
    });
};
export default connect(mapStateToProps)(HelpProDropdown);
