// data-capsule/dist/src/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COOKIE_CONSENT_DISALLOWED = exports.LOCAL_STORAGE_UNSUPPORTED = exports.SERVER_ERROR = exports.NOT_FOUND = exports.MESSAGE_MAX_TIMEOUT = exports.CONNECTION_MAX_TIMEOUT = exports.STORAGE_PREFIX = exports.KEY_SEPARATOR = exports.PREFIX_SEPARATOR = exports.toError = void 0;
var errors = {
    NOT_FOUND: new Error('Key was not found in capsule'),
    SERVER_ERROR: new Error('Failed to perform operarion on server'),
    LOCAL_STORAGE_UNSUPPORTED: new Error('LocalStorage is not supported'),
    COOKIE_CONSENT_DISALLOWED: new Error('The item cannot be set because the user has not approved the category it belongs to'),
};
function toError(str) {
    return (Object.keys(errors)
        .map(function (key) { return errors[key]; })
        .find(function (err) { return err.message === str; }) || new Error(str));
}
exports.toError = toError;
exports.PREFIX_SEPARATOR = '|';
exports.KEY_SEPARATOR = '#';
exports.STORAGE_PREFIX = 'capsule';
exports.CONNECTION_MAX_TIMEOUT = 2000;
exports.MESSAGE_MAX_TIMEOUT = 8000;
exports.NOT_FOUND = errors.NOT_FOUND;
exports.SERVER_ERROR = errors.SERVER_ERROR;
exports.LOCAL_STORAGE_UNSUPPORTED = errors.LOCAL_STORAGE_UNSUPPORTED;
exports.COOKIE_CONSENT_DISALLOWED = errors.COOKIE_CONSENT_DISALLOWED;
