// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/Preview/getPreviewSiteAction.js
import * as React from 'react';
import { ActionType } from '../../../../types';
import Visible from 'wix-ui-icons-common/Visible';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { PERMISSIONS } from '../../../../consts/permissions';
import { sendSiteActionsBI } from '../../../../services/BIService';
import { BIEvents } from '../../../../consts/bi';
function onPreviewSiteClicked(params) {
    var origin = params.origin, _a = params.quickActionsData, metaSiteId = _a.metaSiteId, htmlWebId = _a.htmlWebId;
    sendSiteActionsBI({
        evid: BIEvents.quickActionClicked,
        name: 'quick_actions.preview',
        msid: metaSiteId,
        origin: origin
    });
    window.open("//editor.wix.com/html/editor/web/renderer/external_preview/document/" + htmlWebId + "?metaSiteId=" + metaSiteId);
}
export var getPreviewSiteAction = function (quickActionsData) {
    var isNotPublished = !quickActionsData.published;
    return {
        type: ActionType.click,
        actionName: 'preview',
        textKey: 'quickActions.preview',
        icon: React.createElement(Visible, null),
        isPermitted: PermissionMethods.withPermissions(PERMISSIONS.PREVIEW_SITE),
        visible: isNotPublished,
        onActionClick: onPreviewSiteClicked
    };
};
