// @wix/ambassador-payment-services-web/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2TransactionType = exports.V2OrderItemOrderItemCategory = exports.V1TransactionStatus = exports.Type = exports.TransactionTransactionType = exports.TransactionTransactionStatus = exports.TransactionStatusStatus = exports.SubscriptionStatus = exports.SubscriptionFrequency = exports.RefundStatus = exports.RedirectTarget = exports.RecommendationType = exports.Reason = exports.ProviderAccountStatus = exports.PaymentMethodType = exports.PaymentMethodStatus = exports.OrderStatusStatus = exports.OrderSnapshotStatus = exports.OrderOrderType = exports.OrderOrderItemOrderItemCategory = exports.OrderDescriptionOrderType = exports.MethodType = exports.MerchantV2TransactionStatus = exports.MerchantCapability = exports.LanguageCode = exports.Frequency = exports.FlowType = exports.Domain = exports.CreditCardNetwork = exports.Country = exports.CheckoutType = exports.ChargeIntent = void 0;
var ChargeIntent;
(function (ChargeIntent) {
    ChargeIntent["UNDEFINED"] = "UNDEFINED";
    ChargeIntent["AUTHORIZATION"] = "AUTHORIZATION";
    ChargeIntent["RECURRING"] = "RECURRING";
    ChargeIntent["SALE"] = "SALE";
})(ChargeIntent = exports.ChargeIntent || (exports.ChargeIntent = {}));
var CheckoutType;
(function (CheckoutType) {
    CheckoutType["EXPRESS"] = "EXPRESS";
    CheckoutType["REGULAR"] = "REGULAR";
    CheckoutType["UNKNOWN"] = "UNKNOWN";
})(CheckoutType = exports.CheckoutType || (exports.CheckoutType = {}));
var Country;
(function (Country) {
    Country["AD"] = "AD";
    Country["AE"] = "AE";
    Country["AF"] = "AF";
    Country["AG"] = "AG";
    Country["AI"] = "AI";
    Country["AL"] = "AL";
    Country["AM"] = "AM";
    Country["AN"] = "AN";
    Country["AO"] = "AO";
    Country["AQ"] = "AQ";
    Country["AR"] = "AR";
    Country["AS"] = "AS";
    Country["AT"] = "AT";
    Country["AU"] = "AU";
    Country["AW"] = "AW";
    Country["AX"] = "AX";
    Country["AZ"] = "AZ";
    Country["BA"] = "BA";
    Country["BB"] = "BB";
    Country["BD"] = "BD";
    Country["BE"] = "BE";
    Country["BF"] = "BF";
    Country["BG"] = "BG";
    Country["BH"] = "BH";
    Country["BI"] = "BI";
    Country["BJ"] = "BJ";
    Country["BL"] = "BL";
    Country["BM"] = "BM";
    Country["BN"] = "BN";
    Country["BO"] = "BO";
    Country["BQ"] = "BQ";
    Country["BR"] = "BR";
    Country["BS"] = "BS";
    Country["BT"] = "BT";
    Country["BV"] = "BV";
    Country["BW"] = "BW";
    Country["BY"] = "BY";
    Country["BZ"] = "BZ";
    Country["CA"] = "CA";
    Country["CC"] = "CC";
    Country["CD"] = "CD";
    Country["CF"] = "CF";
    Country["CG"] = "CG";
    Country["CH"] = "CH";
    Country["CI"] = "CI";
    Country["CK"] = "CK";
    Country["CL"] = "CL";
    Country["CM"] = "CM";
    Country["CN"] = "CN";
    Country["CO"] = "CO";
    Country["CR"] = "CR";
    Country["CU"] = "CU";
    Country["CV"] = "CV";
    Country["CW"] = "CW";
    Country["CX"] = "CX";
    Country["CY"] = "CY";
    Country["CZ"] = "CZ";
    Country["DE"] = "DE";
    Country["DJ"] = "DJ";
    Country["DK"] = "DK";
    Country["DM"] = "DM";
    Country["DO"] = "DO";
    Country["DZ"] = "DZ";
    Country["EC"] = "EC";
    Country["EE"] = "EE";
    Country["EG"] = "EG";
    Country["EH"] = "EH";
    Country["ER"] = "ER";
    Country["ES"] = "ES";
    Country["ET"] = "ET";
    Country["FI"] = "FI";
    Country["FJ"] = "FJ";
    Country["FK"] = "FK";
    Country["FM"] = "FM";
    Country["FO"] = "FO";
    Country["FR"] = "FR";
    Country["GA"] = "GA";
    Country["GB"] = "GB";
    Country["GD"] = "GD";
    Country["GE"] = "GE";
    Country["GF"] = "GF";
    Country["GG"] = "GG";
    Country["GH"] = "GH";
    Country["GI"] = "GI";
    Country["GL"] = "GL";
    Country["GM"] = "GM";
    Country["GN"] = "GN";
    Country["GP"] = "GP";
    Country["GQ"] = "GQ";
    Country["GR"] = "GR";
    Country["GS"] = "GS";
    Country["GT"] = "GT";
    Country["GU"] = "GU";
    Country["GW"] = "GW";
    Country["GY"] = "GY";
    Country["HK"] = "HK";
    Country["HM"] = "HM";
    Country["HN"] = "HN";
    Country["HR"] = "HR";
    Country["HT"] = "HT";
    Country["HU"] = "HU";
    Country["ID"] = "ID";
    Country["IE"] = "IE";
    Country["IL"] = "IL";
    Country["IM"] = "IM";
    Country["IN"] = "IN";
    Country["IO"] = "IO";
    Country["IQ"] = "IQ";
    Country["IR"] = "IR";
    Country["IS"] = "IS";
    Country["IT"] = "IT";
    Country["JE"] = "JE";
    Country["JM"] = "JM";
    Country["JO"] = "JO";
    Country["JP"] = "JP";
    Country["KE"] = "KE";
    Country["KG"] = "KG";
    Country["KH"] = "KH";
    Country["KI"] = "KI";
    Country["KM"] = "KM";
    Country["KN"] = "KN";
    Country["KP"] = "KP";
    Country["KR"] = "KR";
    Country["KW"] = "KW";
    Country["KY"] = "KY";
    Country["KZ"] = "KZ";
    Country["LA"] = "LA";
    Country["LB"] = "LB";
    Country["LC"] = "LC";
    Country["LI"] = "LI";
    Country["LK"] = "LK";
    Country["LR"] = "LR";
    Country["LS"] = "LS";
    Country["LT"] = "LT";
    Country["LU"] = "LU";
    Country["LV"] = "LV";
    Country["LY"] = "LY";
    Country["MA"] = "MA";
    Country["MC"] = "MC";
    Country["MD"] = "MD";
    Country["ME"] = "ME";
    Country["MF"] = "MF";
    Country["MG"] = "MG";
    Country["MH"] = "MH";
    Country["MK"] = "MK";
    Country["ML"] = "ML";
    Country["MM"] = "MM";
    Country["MN"] = "MN";
    Country["MO"] = "MO";
    Country["MP"] = "MP";
    Country["MQ"] = "MQ";
    Country["MR"] = "MR";
    Country["MS"] = "MS";
    Country["MT"] = "MT";
    Country["MU"] = "MU";
    Country["MV"] = "MV";
    Country["MW"] = "MW";
    Country["MX"] = "MX";
    Country["MY"] = "MY";
    Country["MZ"] = "MZ";
    Country["NA"] = "NA";
    Country["NC"] = "NC";
    Country["NE"] = "NE";
    Country["NF"] = "NF";
    Country["NG"] = "NG";
    Country["NI"] = "NI";
    Country["NL"] = "NL";
    Country["NO"] = "NO";
    Country["NP"] = "NP";
    Country["NR"] = "NR";
    Country["NU"] = "NU";
    Country["NZ"] = "NZ";
    Country["OM"] = "OM";
    Country["PA"] = "PA";
    Country["PE"] = "PE";
    Country["PF"] = "PF";
    Country["PG"] = "PG";
    Country["PH"] = "PH";
    Country["PK"] = "PK";
    Country["PL"] = "PL";
    Country["PM"] = "PM";
    Country["PN"] = "PN";
    Country["PR"] = "PR";
    Country["PS"] = "PS";
    Country["PT"] = "PT";
    Country["PW"] = "PW";
    Country["PY"] = "PY";
    Country["QA"] = "QA";
    Country["RE"] = "RE";
    Country["RO"] = "RO";
    Country["RS"] = "RS";
    Country["RU"] = "RU";
    Country["RW"] = "RW";
    Country["SA"] = "SA";
    Country["SB"] = "SB";
    Country["SC"] = "SC";
    Country["SD"] = "SD";
    Country["SE"] = "SE";
    Country["SG"] = "SG";
    Country["SH"] = "SH";
    Country["SI"] = "SI";
    Country["SJ"] = "SJ";
    Country["SK"] = "SK";
    Country["SL"] = "SL";
    Country["SM"] = "SM";
    Country["SN"] = "SN";
    Country["SO"] = "SO";
    Country["SR"] = "SR";
    Country["SS"] = "SS";
    Country["ST"] = "ST";
    Country["SV"] = "SV";
    Country["SX"] = "SX";
    Country["SY"] = "SY";
    Country["SZ"] = "SZ";
    Country["TC"] = "TC";
    Country["TD"] = "TD";
    Country["TF"] = "TF";
    Country["TG"] = "TG";
    Country["TH"] = "TH";
    Country["TJ"] = "TJ";
    Country["TK"] = "TK";
    Country["TL"] = "TL";
    Country["TM"] = "TM";
    Country["TN"] = "TN";
    Country["TO"] = "TO";
    Country["TR"] = "TR";
    Country["TT"] = "TT";
    Country["TV"] = "TV";
    Country["TW"] = "TW";
    Country["TZ"] = "TZ";
    Country["UA"] = "UA";
    Country["UG"] = "UG";
    Country["UM"] = "UM";
    Country["US"] = "US";
    Country["UY"] = "UY";
    Country["UZ"] = "UZ";
    Country["VA"] = "VA";
    Country["VC"] = "VC";
    Country["VE"] = "VE";
    Country["VG"] = "VG";
    Country["VI"] = "VI";
    Country["VN"] = "VN";
    Country["VU"] = "VU";
    Country["WF"] = "WF";
    Country["WS"] = "WS";
    Country["YE"] = "YE";
    Country["YT"] = "YT";
    Country["ZA"] = "ZA";
    Country["ZM"] = "ZM";
    Country["ZW"] = "ZW";
})(Country = exports.Country || (exports.Country = {}));
var CreditCardNetwork;
(function (CreditCardNetwork) {
    CreditCardNetwork["VPAY"] = "VPAY";
    CreditCardNetwork["PRIVATE_LABEL"] = "PRIVATE_LABEL";
    CreditCardNetwork["CARTES_BANCAIRES"] = "CARTES_BANCAIRES";
    CreditCardNetwork["ELO"] = "ELO";
    CreditCardNetwork["INTERAC"] = "INTERAC";
    CreditCardNetwork["MADA"] = "MADA";
    CreditCardNetwork["UNDEFINED_NETWORK"] = "UNDEFINED_NETWORK";
    CreditCardNetwork["ELECTRON"] = "ELECTRON";
    CreditCardNetwork["MASTER_CARD"] = "MASTER_CARD";
    CreditCardNetwork["JCB"] = "JCB";
    CreditCardNetwork["VISA"] = "VISA";
    CreditCardNetwork["AMEX"] = "AMEX";
    CreditCardNetwork["MAESTRO"] = "MAESTRO";
    CreditCardNetwork["DISCOVER"] = "DISCOVER";
    CreditCardNetwork["CHINA_UNION_PAY"] = "CHINA_UNION_PAY";
    CreditCardNetwork["EFTPOS"] = "EFTPOS";
})(CreditCardNetwork = exports.CreditCardNetwork || (exports.CreditCardNetwork = {}));
var Domain;
(function (Domain) {
    Domain["PAYMENT_METHOD"] = "PAYMENT_METHOD";
    Domain["CREDIT_CARD"] = "CREDIT_CARD";
    Domain["SAVED_CREDIT_CARD"] = "SAVED_CREDIT_CARD";
    Domain["ORDER"] = "ORDER";
    Domain["UNDEFINED"] = "UNDEFINED";
})(Domain = exports.Domain || (exports.Domain = {}));
var FlowType;
(function (FlowType) {
    FlowType["ASYNC"] = "ASYNC";
    FlowType["NA"] = "NA";
    FlowType["SYNC"] = "SYNC";
})(FlowType = exports.FlowType || (exports.FlowType = {}));
var Frequency;
(function (Frequency) {
    Frequency["DAY"] = "DAY";
    Frequency["WEEK"] = "WEEK";
    Frequency["MONTH"] = "MONTH";
    Frequency["UNDEFINED"] = "UNDEFINED";
    Frequency["YEAR"] = "YEAR";
})(Frequency = exports.Frequency || (exports.Frequency = {}));
var LanguageCode;
(function (LanguageCode) {
    LanguageCode["aa"] = "aa";
    LanguageCode["ab"] = "ab";
    LanguageCode["af"] = "af";
    LanguageCode["ak"] = "ak";
    LanguageCode["sq"] = "sq";
    LanguageCode["am"] = "am";
    LanguageCode["ar"] = "ar";
    LanguageCode["an"] = "an";
    LanguageCode["hy"] = "hy";
    LanguageCode["as"] = "as";
    LanguageCode["av"] = "av";
    LanguageCode["ae"] = "ae";
    LanguageCode["ay"] = "ay";
    LanguageCode["az"] = "az";
    LanguageCode["ba"] = "ba";
    LanguageCode["bm"] = "bm";
    LanguageCode["eu"] = "eu";
    LanguageCode["be"] = "be";
    LanguageCode["bn"] = "bn";
    LanguageCode["bh"] = "bh";
    LanguageCode["bi"] = "bi";
    LanguageCode["bs"] = "bs";
    LanguageCode["br"] = "br";
    LanguageCode["bg"] = "bg";
    LanguageCode["my"] = "my";
    LanguageCode["ca"] = "ca";
    LanguageCode["ch"] = "ch";
    LanguageCode["ce"] = "ce";
    LanguageCode["zh"] = "zh";
    LanguageCode["cu"] = "cu";
    LanguageCode["cv"] = "cv";
    LanguageCode["kw"] = "kw";
    LanguageCode["co"] = "co";
    LanguageCode["cr"] = "cr";
    LanguageCode["cs"] = "cs";
    LanguageCode["da"] = "da";
    LanguageCode["dv"] = "dv";
    LanguageCode["nl"] = "nl";
    LanguageCode["dz"] = "dz";
    LanguageCode["en"] = "en";
    LanguageCode["eo"] = "eo";
    LanguageCode["et"] = "et";
    LanguageCode["ee"] = "ee";
    LanguageCode["fo"] = "fo";
    LanguageCode["fj"] = "fj";
    LanguageCode["fi"] = "fi";
    LanguageCode["fr"] = "fr";
    LanguageCode["fy"] = "fy";
    LanguageCode["ff"] = "ff";
    LanguageCode["ka"] = "ka";
    LanguageCode["de"] = "de";
    LanguageCode["gd"] = "gd";
    LanguageCode["ga"] = "ga";
    LanguageCode["gl"] = "gl";
    LanguageCode["gv"] = "gv";
    LanguageCode["el"] = "el";
    LanguageCode["gn"] = "gn";
    LanguageCode["gu"] = "gu";
    LanguageCode["ht"] = "ht";
    LanguageCode["ha"] = "ha";
    LanguageCode["he"] = "he";
    LanguageCode["hz"] = "hz";
    LanguageCode["hi"] = "hi";
    LanguageCode["ho"] = "ho";
    LanguageCode["hr"] = "hr";
    LanguageCode["hu"] = "hu";
    LanguageCode["ig"] = "ig";
    LanguageCode["is"] = "is";
    LanguageCode["io"] = "io";
    LanguageCode["ii"] = "ii";
    LanguageCode["iu"] = "iu";
    LanguageCode["ie"] = "ie";
    LanguageCode["ia"] = "ia";
    LanguageCode["id"] = "id";
    LanguageCode["ik"] = "ik";
    LanguageCode["it"] = "it";
    LanguageCode["jv"] = "jv";
    LanguageCode["ja"] = "ja";
    LanguageCode["kl"] = "kl";
    LanguageCode["kn"] = "kn";
    LanguageCode["ks"] = "ks";
    LanguageCode["kr"] = "kr";
    LanguageCode["kk"] = "kk";
    LanguageCode["km"] = "km";
    LanguageCode["ki"] = "ki";
    LanguageCode["rw"] = "rw";
    LanguageCode["ky"] = "ky";
    LanguageCode["kv"] = "kv";
    LanguageCode["kg"] = "kg";
    LanguageCode["ko"] = "ko";
    LanguageCode["kj"] = "kj";
    LanguageCode["ku"] = "ku";
    LanguageCode["lo"] = "lo";
    LanguageCode["la"] = "la";
    LanguageCode["lv"] = "lv";
    LanguageCode["li"] = "li";
    LanguageCode["ln"] = "ln";
    LanguageCode["lt"] = "lt";
    LanguageCode["lb"] = "lb";
    LanguageCode["ld"] = "ld";
    LanguageCode["lu"] = "lu";
    LanguageCode["lg"] = "lg";
    LanguageCode["mk"] = "mk";
    LanguageCode["mh"] = "mh";
    LanguageCode["ml"] = "ml";
    LanguageCode["mi"] = "mi";
    LanguageCode["mr"] = "mr";
    LanguageCode["ms"] = "ms";
    LanguageCode["mg"] = "mg";
    LanguageCode["mt"] = "mt";
    LanguageCode["mn"] = "mn";
    LanguageCode["na"] = "na";
    LanguageCode["nv"] = "nv";
    LanguageCode["nr"] = "nr";
    LanguageCode["nd"] = "nd";
    LanguageCode["ng"] = "ng";
    LanguageCode["ne"] = "ne";
    LanguageCode["nn"] = "nn";
    LanguageCode["nb"] = "nb";
    LanguageCode["no"] = "no";
    LanguageCode["ny"] = "ny";
    LanguageCode["oc"] = "oc";
    LanguageCode["oj"] = "oj";
    LanguageCode["or"] = "or";
    LanguageCode["om"] = "om";
    LanguageCode["os"] = "os";
    LanguageCode["pa"] = "pa";
    LanguageCode["fa"] = "fa";
    LanguageCode["pi"] = "pi";
    LanguageCode["pl"] = "pl";
    LanguageCode["pt"] = "pt";
    LanguageCode["ps"] = "ps";
    LanguageCode["qu"] = "qu";
    LanguageCode["rm"] = "rm";
    LanguageCode["ro"] = "ro";
    LanguageCode["rn"] = "rn";
    LanguageCode["ru"] = "ru";
    LanguageCode["sg"] = "sg";
    LanguageCode["sa"] = "sa";
    LanguageCode["si"] = "si";
    LanguageCode["sk"] = "sk";
    LanguageCode["sl"] = "sl";
    LanguageCode["se"] = "se";
    LanguageCode["sm"] = "sm";
    LanguageCode["sn"] = "sn";
    LanguageCode["sd"] = "sd";
    LanguageCode["so"] = "so";
    LanguageCode["st"] = "st";
    LanguageCode["es"] = "es";
    LanguageCode["sc"] = "sc";
    LanguageCode["sr"] = "sr";
    LanguageCode["ss"] = "ss";
    LanguageCode["su"] = "su";
    LanguageCode["sw"] = "sw";
    LanguageCode["sv"] = "sv";
    LanguageCode["ty"] = "ty";
    LanguageCode["ta"] = "ta";
    LanguageCode["tt"] = "tt";
    LanguageCode["te"] = "te";
    LanguageCode["tg"] = "tg";
    LanguageCode["tl"] = "tl";
    LanguageCode["th"] = "th";
    LanguageCode["bo"] = "bo";
    LanguageCode["ti"] = "ti";
    LanguageCode["to"] = "to";
    LanguageCode["tn"] = "tn";
    LanguageCode["ts"] = "ts";
    LanguageCode["tk"] = "tk";
    LanguageCode["tr"] = "tr";
    LanguageCode["tw"] = "tw";
    LanguageCode["ug"] = "ug";
    LanguageCode["uk"] = "uk";
    LanguageCode["ur"] = "ur";
    LanguageCode["uz"] = "uz";
    LanguageCode["ve"] = "ve";
    LanguageCode["vi"] = "vi";
    LanguageCode["vo"] = "vo";
    LanguageCode["cy"] = "cy";
    LanguageCode["wa"] = "wa";
    LanguageCode["wo"] = "wo";
    LanguageCode["xh"] = "xh";
    LanguageCode["yi"] = "yi";
    LanguageCode["yo"] = "yo";
    LanguageCode["za"] = "za";
    LanguageCode["zu"] = "zu";
})(LanguageCode = exports.LanguageCode || (exports.LanguageCode = {}));
var MerchantCapability;
(function (MerchantCapability) {
    MerchantCapability["SUPPORTS_EMV"] = "SUPPORTS_EMV";
    MerchantCapability["SUPPORTS_CREDIT"] = "SUPPORTS_CREDIT";
    MerchantCapability["UNDEFINED_CAPABILITY"] = "UNDEFINED_CAPABILITY";
    MerchantCapability["SUPPORTS_3DS"] = "SUPPORTS_3DS";
    MerchantCapability["SUPPORTS_DEBIT"] = "SUPPORTS_DEBIT";
})(MerchantCapability = exports.MerchantCapability || (exports.MerchantCapability = {}));
var MerchantV2TransactionStatus;
(function (MerchantV2TransactionStatus) {
    MerchantV2TransactionStatus["REFUND"] = "REFUND";
    MerchantV2TransactionStatus["PENDING"] = "PENDING";
    MerchantV2TransactionStatus["INITIALIZED"] = "INITIALIZED";
    MerchantV2TransactionStatus["CHARGE_BACK"] = "CHARGE_BACK";
    MerchantV2TransactionStatus["EXPIRED"] = "EXPIRED";
    MerchantV2TransactionStatus["IN_PROCESS"] = "IN_PROCESS";
    MerchantV2TransactionStatus["TPA_CANCELED"] = "TPA_CANCELED";
    MerchantV2TransactionStatus["PENDING_BUYER"] = "PENDING_BUYER";
    MerchantV2TransactionStatus["APPROVED"] = "APPROVED";
    MerchantV2TransactionStatus["DECLINED"] = "DECLINED";
    MerchantV2TransactionStatus["OFFLINE"] = "OFFLINE";
    MerchantV2TransactionStatus["TIMEOUT"] = "TIMEOUT";
    MerchantV2TransactionStatus["FAILED"] = "FAILED";
    MerchantV2TransactionStatus["BUYER_CANCELED"] = "BUYER_CANCELED";
    MerchantV2TransactionStatus["PENDING_MERCHANT"] = "PENDING_MERCHANT";
    MerchantV2TransactionStatus["DISPUTE"] = "DISPUTE";
    MerchantV2TransactionStatus["COMPLETED_FUNDS_HELD"] = "COMPLETED_FUNDS_HELD";
    MerchantV2TransactionStatus["UNDEFINED"] = "UNDEFINED";
    MerchantV2TransactionStatus["VOID"] = "VOID";
    MerchantV2TransactionStatus["PARTIAL_REFUND"] = "PARTIAL_REFUND";
})(MerchantV2TransactionStatus = exports.MerchantV2TransactionStatus || (exports.MerchantV2TransactionStatus = {}));
var MethodType;
(function (MethodType) {
    MethodType["FACE_TO_FACE"] = "FACE_TO_FACE";
    MethodType["ONLINE"] = "ONLINE";
    MethodType["UNDEFINED"] = "UNDEFINED";
})(MethodType = exports.MethodType || (exports.MethodType = {}));
var OrderDescriptionOrderType;
(function (OrderDescriptionOrderType) {
    OrderDescriptionOrderType["ONE_TIME"] = "ONE_TIME";
    OrderDescriptionOrderType["SUBSCRIPTION"] = "SUBSCRIPTION";
    OrderDescriptionOrderType["UNDEFINED"] = "UNDEFINED";
})(OrderDescriptionOrderType = exports.OrderDescriptionOrderType || (exports.OrderDescriptionOrderType = {}));
var OrderOrderItemOrderItemCategory;
(function (OrderOrderItemOrderItemCategory) {
    OrderOrderItemOrderItemCategory["DIGITAL"] = "DIGITAL";
    OrderOrderItemOrderItemCategory["PHYSICAL"] = "PHYSICAL";
    OrderOrderItemOrderItemCategory["UNDEFINED"] = "UNDEFINED";
})(OrderOrderItemOrderItemCategory = exports.OrderOrderItemOrderItemCategory || (exports.OrderOrderItemOrderItemCategory = {}));
var OrderOrderType;
(function (OrderOrderType) {
    OrderOrderType["OneTime"] = "OneTime";
    OrderOrderType["Recurring"] = "Recurring";
})(OrderOrderType = exports.OrderOrderType || (exports.OrderOrderType = {}));
var OrderSnapshotStatus;
(function (OrderSnapshotStatus) {
    OrderSnapshotStatus["SubscriptionPending"] = "SubscriptionPending";
    OrderSnapshotStatus["SubscriptionActive"] = "SubscriptionActive";
    OrderSnapshotStatus["NotApproved"] = "NotApproved";
    OrderSnapshotStatus["Approved"] = "Approved";
    OrderSnapshotStatus["Expired"] = "Expired";
    OrderSnapshotStatus["SubscriptionCancel"] = "SubscriptionCancel";
    OrderSnapshotStatus["SubscriptionFinished"] = "SubscriptionFinished";
    OrderSnapshotStatus["InProcess"] = "InProcess";
    OrderSnapshotStatus["SubscriptionSuspended"] = "SubscriptionSuspended";
    OrderSnapshotStatus["Initialized"] = "Initialized";
})(OrderSnapshotStatus = exports.OrderSnapshotStatus || (exports.OrderSnapshotStatus = {}));
var OrderStatusStatus;
(function (OrderStatusStatus) {
    OrderStatusStatus["EXPIRED"] = "EXPIRED";
    OrderStatusStatus["NOT_APPROVED"] = "NOT_APPROVED";
    OrderStatusStatus["SUBSCRIPTION_ACTIVE"] = "SUBSCRIPTION_ACTIVE";
    OrderStatusStatus["APPROVED"] = "APPROVED";
    OrderStatusStatus["SUBSCRIPTION_CANCELLED"] = "SUBSCRIPTION_CANCELLED";
    OrderStatusStatus["SUBSCRIPTION_FINISHED"] = "SUBSCRIPTION_FINISHED";
    OrderStatusStatus["SUBSCRIPTION_PENDING"] = "SUBSCRIPTION_PENDING";
    OrderStatusStatus["UNDEFINED"] = "UNDEFINED";
    OrderStatusStatus["INITIALIZED"] = "INITIALIZED";
    OrderStatusStatus["SUBSCRIPTION_SUSPENDED"] = "SUBSCRIPTION_SUSPENDED";
    OrderStatusStatus["IN_PROCESS"] = "IN_PROCESS";
})(OrderStatusStatus = exports.OrderStatusStatus || (exports.OrderStatusStatus = {}));
var PaymentMethodStatus;
(function (PaymentMethodStatus) {
    PaymentMethodStatus["CONNECTED"] = "CONNECTED";
    PaymentMethodStatus["SELECTED"] = "SELECTED";
})(PaymentMethodStatus = exports.PaymentMethodStatus || (exports.PaymentMethodStatus = {}));
var PaymentMethodType;
(function (PaymentMethodType) {
    PaymentMethodType["OFFLINE_METHOD"] = "OFFLINE_METHOD";
    PaymentMethodType["ONLINE_METHOD"] = "ONLINE_METHOD";
})(PaymentMethodType = exports.PaymentMethodType || (exports.PaymentMethodType = {}));
var ProviderAccountStatus;
(function (ProviderAccountStatus) {
    ProviderAccountStatus["FUNDS_HELD"] = "FUNDS_HELD";
    ProviderAccountStatus["MUST_BE_CLAIMED"] = "MUST_BE_CLAIMED";
    ProviderAccountStatus["NOT_CLAIMED"] = "NOT_CLAIMED";
    ProviderAccountStatus["CLAIMED"] = "CLAIMED";
    ProviderAccountStatus["BLOCKED"] = "BLOCKED";
    ProviderAccountStatus["UNKNOWN"] = "UNKNOWN";
    ProviderAccountStatus["SUSPENDED"] = "SUSPENDED";
})(ProviderAccountStatus = exports.ProviderAccountStatus || (exports.ProviderAccountStatus = {}));
var Reason;
(function (Reason) {
    Reason["PENDING_REFUND"] = "PENDING_REFUND";
    Reason["PROVIDER_IS_DOWN"] = "PROVIDER_IS_DOWN";
    Reason["FORBIDDEN"] = "FORBIDDEN";
    Reason["NOT_SUPPORTED"] = "NOT_SUPPORTED";
    Reason["EXPIRED"] = "EXPIRED";
    Reason["ALREADY_REFUNDED"] = "ALREADY_REFUNDED";
})(Reason = exports.Reason || (exports.Reason = {}));
var RecommendationType;
(function (RecommendationType) {
    RecommendationType["TOBACCO_ALCOHOL"] = "TOBACCO_ALCOHOL";
    RecommendationType["OTHER"] = "OTHER";
    RecommendationType["MEDICAL"] = "MEDICAL";
    RecommendationType["RESTRICTED_GOODS"] = "RESTRICTED_GOODS";
    RecommendationType["UNKNOWN"] = "UNKNOWN";
    RecommendationType["FINANCIAL"] = "FINANCIAL";
    RecommendationType["BINARY_OPTIONS_CRYPTOCURRENCIES"] = "BINARY_OPTIONS_CRYPTOCURRENCIES";
    RecommendationType["CBD"] = "CBD";
    RecommendationType["GAMBLING_LOTTERIES_SKILL_GAMES"] = "GAMBLING_LOTTERIES_SKILL_GAMES";
    RecommendationType["TRAVEL_AGENCIES"] = "TRAVEL_AGENCIES";
    RecommendationType["MARKETPLACES"] = "MARKETPLACES";
    RecommendationType["FIREARMS_WEAPONS"] = "FIREARMS_WEAPONS";
    RecommendationType["ADULT"] = "ADULT";
})(RecommendationType = exports.RecommendationType || (exports.RecommendationType = {}));
var RedirectTarget;
(function (RedirectTarget) {
    RedirectTarget["EMBEDDED"] = "EMBEDDED";
    RedirectTarget["NEW_WINDOW"] = "NEW_WINDOW";
    RedirectTarget["SAME_WINDOW"] = "SAME_WINDOW";
})(RedirectTarget = exports.RedirectTarget || (exports.RedirectTarget = {}));
var RefundStatus;
(function (RefundStatus) {
    RefundStatus["FAILED"] = "FAILED";
    RefundStatus["PENDING"] = "PENDING";
    RefundStatus["SUCCEEDED"] = "SUCCEEDED";
})(RefundStatus = exports.RefundStatus || (exports.RefundStatus = {}));
var SubscriptionFrequency;
(function (SubscriptionFrequency) {
    SubscriptionFrequency["MONTH"] = "MONTH";
    SubscriptionFrequency["WEEK"] = "WEEK";
    SubscriptionFrequency["UNDEFINED"] = "UNDEFINED";
    SubscriptionFrequency["DAY"] = "DAY";
    SubscriptionFrequency["YEAR"] = "YEAR";
})(SubscriptionFrequency = exports.SubscriptionFrequency || (exports.SubscriptionFrequency = {}));
var SubscriptionStatus;
(function (SubscriptionStatus) {
    SubscriptionStatus["SUBSCRIPTION_SUSPENDED"] = "SUBSCRIPTION_SUSPENDED";
    SubscriptionStatus["SUBSCRIPTION_FINISHED"] = "SUBSCRIPTION_FINISHED";
    SubscriptionStatus["SUBSCRIPTION_PENDING"] = "SUBSCRIPTION_PENDING";
    SubscriptionStatus["SUBSCRIPTION_CANCELLED"] = "SUBSCRIPTION_CANCELLED";
    SubscriptionStatus["SUBSCRIPTION_ACTIVE"] = "SUBSCRIPTION_ACTIVE";
    SubscriptionStatus["SUBSCRIPTION_UNDEFINED"] = "SUBSCRIPTION_UNDEFINED";
})(SubscriptionStatus = exports.SubscriptionStatus || (exports.SubscriptionStatus = {}));
var TransactionStatusStatus;
(function (TransactionStatusStatus) {
    TransactionStatusStatus["REFUND"] = "REFUND";
    TransactionStatusStatus["IN_PROCESS"] = "IN_PROCESS";
    TransactionStatusStatus["INITIALIZED"] = "INITIALIZED";
    TransactionStatusStatus["FAILED"] = "FAILED";
    TransactionStatusStatus["APPROVED"] = "APPROVED";
    TransactionStatusStatus["TIMEOUT"] = "TIMEOUT";
    TransactionStatusStatus["PENDING_BUYER"] = "PENDING_BUYER";
    TransactionStatusStatus["OFFLINE"] = "OFFLINE";
    TransactionStatusStatus["PENDING_MERCHANT"] = "PENDING_MERCHANT";
    TransactionStatusStatus["CHARGE_BACK"] = "CHARGE_BACK";
    TransactionStatusStatus["INITIALIZATION"] = "INITIALIZATION";
    TransactionStatusStatus["BUYER_CANCELED"] = "BUYER_CANCELED";
    TransactionStatusStatus["UNDEFINED"] = "UNDEFINED";
    TransactionStatusStatus["DECLINED"] = "DECLINED";
})(TransactionStatusStatus = exports.TransactionStatusStatus || (exports.TransactionStatusStatus = {}));
var TransactionTransactionStatus;
(function (TransactionTransactionStatus) {
    TransactionTransactionStatus["Initialized"] = "Initialized";
    TransactionTransactionStatus["Failure"] = "Failure";
    TransactionTransactionStatus["Initialization"] = "Initialization";
    TransactionTransactionStatus["Offline"] = "Offline";
    TransactionTransactionStatus["CanceledReversal"] = "CanceledReversal";
    TransactionTransactionStatus["Void"] = "Void";
    TransactionTransactionStatus["PendingMerchant"] = "PendingMerchant";
    TransactionTransactionStatus["GatewayFailure"] = "GatewayFailure";
    TransactionTransactionStatus["PendingBuyer"] = "PendingBuyer";
    TransactionTransactionStatus["InProcess"] = "InProcess";
    TransactionTransactionStatus["PartialRefund"] = "PartialRefund";
    TransactionTransactionStatus["Unknown"] = "Unknown";
    TransactionTransactionStatus["Pending"] = "Pending";
    TransactionTransactionStatus["BuyerCanceled"] = "BuyerCanceled";
    TransactionTransactionStatus["Dispute"] = "Dispute";
    TransactionTransactionStatus["Approved"] = "Approved";
    TransactionTransactionStatus["Refund"] = "Refund";
    TransactionTransactionStatus["BackofficeTransactionStatusUndefined"] = "BackofficeTransactionStatusUndefined";
    TransactionTransactionStatus["ChargeBack"] = "ChargeBack";
    TransactionTransactionStatus["Declined"] = "Declined";
    TransactionTransactionStatus["Timeout"] = "Timeout";
})(TransactionTransactionStatus = exports.TransactionTransactionStatus || (exports.TransactionTransactionStatus = {}));
var TransactionTransactionType;
(function (TransactionTransactionType) {
    TransactionTransactionType["Sale"] = "Sale";
    TransactionTransactionType["BackofficeTransactionTypeUndefined"] = "BackofficeTransactionTypeUndefined";
    TransactionTransactionType["InitializeRecurring"] = "InitializeRecurring";
    TransactionTransactionType["Recurring"] = "Recurring";
    TransactionTransactionType["Authorization"] = "Authorization";
})(TransactionTransactionType = exports.TransactionTransactionType || (exports.TransactionTransactionType = {}));
var Type;
(function (Type) {
    Type["FULL"] = "FULL";
    Type["PARTIAL"] = "PARTIAL";
})(Type = exports.Type || (exports.Type = {}));
var V1TransactionStatus;
(function (V1TransactionStatus) {
    V1TransactionStatus["CHARGEBACK"] = "CHARGEBACK";
    V1TransactionStatus["CANCELED"] = "CANCELED";
    V1TransactionStatus["UNKNOWN"] = "UNKNOWN";
    V1TransactionStatus["PENDING"] = "PENDING";
    V1TransactionStatus["REFUNDED"] = "REFUNDED";
    V1TransactionStatus["APPROVED"] = "APPROVED";
    V1TransactionStatus["CHARGED_BACK"] = "CHARGED_BACK";
    V1TransactionStatus["DECLINED"] = "DECLINED";
})(V1TransactionStatus = exports.V1TransactionStatus || (exports.V1TransactionStatus = {}));
var V2OrderItemOrderItemCategory;
(function (V2OrderItemOrderItemCategory) {
    V2OrderItemOrderItemCategory["DIGITAL"] = "DIGITAL";
    V2OrderItemOrderItemCategory["PHYSICAL"] = "PHYSICAL";
    V2OrderItemOrderItemCategory["UNDEFINED"] = "UNDEFINED";
})(V2OrderItemOrderItemCategory = exports.V2OrderItemOrderItemCategory || (exports.V2OrderItemOrderItemCategory = {}));
var V2TransactionType;
(function (V2TransactionType) {
    V2TransactionType["INITIALIZE_RECURRING"] = "INITIALIZE_RECURRING";
    V2TransactionType["RECURRING"] = "RECURRING";
    V2TransactionType["SALE"] = "SALE";
})(V2TransactionType = exports.V2TransactionType || (exports.V2TransactionType = {}));
