// @wix/ambassador-dealer-offers-serving-service/http.js
/* eslint-disable */
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DealerOffersServingService = void 0;
var http_1 = require("@wix/ambassador/runtime/http");
var DealerOffersServingService = function (baseUrl, httpClientConfig) {
    var httpClient = new http_1.AmbassadorHTTPClient(baseUrl, httpClientConfig);
    return {
        DealerOffersDebug: function () {
            return function (headers) {
                var service = {
                    debugOffer: function (debugOfferRequest) {
                        return httpClient.invoke('POST', '/v1/dealer/debug/offers/{offer_id}/debug', debugOfferRequest, headers);
                    },
                };
                service.debugOffer.__http_method = 'POST';
                service.debugOffer.__http_url = '/v1/dealer/debug/offers/{offer_id}/debug';
                return service;
            };
        },
        DealerOffersServing: function () {
            return function (headers) {
                var service = {
                    listOffers: function (listOffersRequest) {
                        return httpClient.invoke('POST', '/v1/dealer/serving/offers/list', listOffersRequest, headers);
                    },
                    listOffersBulk: function (listOffersBulkRequest) {
                        return httpClient.invoke('POST', '/v1/dealer/serving/offers/bulk', listOffersBulkRequest, headers);
                    },
                };
                service.listOffers.__http_method = 'POST';
                service.listOffers.__http_url = '/v1/dealer/serving/offers/list';
                service.listOffersBulk.__http_method = 'POST';
                service.listOffersBulk.__http_url = '/v1/dealer/serving/offers/bulk';
                return service;
            };
        },
    };
};
exports.DealerOffersServingService = DealerOffersServingService;
exports.DealerOffersServingService.__transport = 'http';
__exportStar(require("./types"), exports);
