// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/Transfer/getTransferSiteActions.js
import * as React from 'react';
import { DialogNames, ActionType } from '../../../../types';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { TransferSiteAction } from './TransferSiteAction';
import SiteMoveSmall from 'wix-ui-icons-common/SiteMoveSmall';
import { navigateToRolesAndPermissions } from '../../../../services/navigationService';
function openRolesAndPermissions(params) {
    var origin = params.origin, metaSiteId = params.quickActionsData.metaSiteId;
    navigateToRolesAndPermissions(metaSiteId, origin);
}
export var getTransferSiteActions = function (quickActionsData) {
    var showManagedTransfer = PermissionMethods.isInTransferAndHasTransferSitePermission(quickActionsData);
    var openRolesAndPermissionsIfManagedTransfer = function (params) {
        var shouldOpenDialog = true;
        if (showManagedTransfer) {
            openRolesAndPermissions(params);
            shouldOpenDialog = false;
        }
        return shouldOpenDialog;
    };
    return [
        {
            type: ActionType.click,
            actionName: 'manage-transfer',
            textKey: 'quickActions.manage_transfer',
            icon: React.createElement(SiteMoveSmall, null),
            visible: showManagedTransfer,
            isPermitted: function (data) { return !PermissionMethods.isSiteBlocked(data); },
            onActionClick: openRolesAndPermissions,
        },
        {
            type: ActionType.dialog,
            dialog: {
                dialogName: DialogNames.transfer,
                component: TransferSiteAction,
                biName: 'quick_actions.transfer_site',
                onInitialLoad: openRolesAndPermissionsIfManagedTransfer
            },
            actionName: 'transfer',
            textKey: 'quickActions.transfer',
            icon: React.createElement(SiteMoveSmall, null),
            visible: !showManagedTransfer,
            isPermitted: function (data) {
                return PermissionMethods.htmlSiteAndNotInTransferAndHasTransferSitePermission(data) &&
                    !PermissionMethods.isSiteBlocked(data);
            },
        }
    ];
};
