// @wix/ambassador-app-settings-service/http.js
/* eslint-disable */
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var http_1 = require("@wix/ambassador/runtime/http");
exports.AppSettingsService = function (baseUrl, httpClientConfig) {
    var httpClient = new http_1.AmbassadorHTTPClient(baseUrl, httpClientConfig);
    return {
        AppSettingsService: function () {
            return function (headers) {
                var service = {
                    get: function (getRequest) {
                        return httpClient.invoke('GET', 'v2/settings', getRequest, headers);
                    },
                    getComponentSettings: function (getComponentSettingsRequest) {
                        return httpClient.invoke('GET', '/v1/settings/components/{component_id}', getComponentSettingsRequest, headers);
                    },
                    getGlobalSettings: function (getGlobalSettingsRequest) {
                        return httpClient.invoke('GET', '/v1/settings/instances', getGlobalSettingsRequest, headers);
                    },
                    set: function (setRequest) {
                        return httpClient.invoke('PUT', 'v2/settings/{scope}/{externalId}', setRequest, headers);
                    },
                    setComponentSettings: function (setComponentSettingsRequest) {
                        return httpClient.invoke('PATCH', '/v1/settings/components/{component_id}', setComponentSettingsRequest, headers);
                    },
                    setGlobalSettings: function (setGlobalSettingsRequest) {
                        return httpClient.invoke('PATCH', '/v1/settings/instances', setGlobalSettingsRequest, headers);
                    },
                    update: function (updateRequest) {
                        return httpClient.invoke('PATCH', 'v2/settings/{scope}/{externalId}', updateRequest, headers);
                    }
                };
                service.get.__http_method = 'GET';
                service.get.__http_url = 'v2/settings';
                service.getComponentSettings.__http_method = 'GET';
                service.getComponentSettings.__http_url = '/v1/settings/components/{component_id}';
                service.getGlobalSettings.__http_method = 'GET';
                service.getGlobalSettings.__http_url = '/v1/settings/instances';
                service.set.__http_method = 'PUT';
                service.set.__http_url = 'v2/settings/{scope}/{externalId}';
                service.setComponentSettings.__http_method = 'PATCH';
                service.setComponentSettings.__http_url = '/v1/settings/components/{component_id}';
                service.setGlobalSettings.__http_method = 'PATCH';
                service.setGlobalSettings.__http_url = '/v1/settings/instances';
                service.update.__http_method = 'PATCH';
                service.update.__http_url = 'v2/settings/{scope}/{externalId}';
                return service;
            };
        }
    };
};
exports.AppSettingsService.__transport = 'http';
__export(require("./types"));
