// @wix/my-account-header/dist/es/src/client/components/Header/Editor/EditorHeader.js
import { __extends } from "tslib";
import * as React from 'react';
import * as css from './EditorHeader.scss';
import { ModuleRegistry } from 'react-module-container';
import Help from '../../Help/Help';
import { withTranslation } from '@wix/wix-i18n-config';
import { ON_CLOSE_IN_EDITOR } from '@wix/business-manager-api';
import { withHeaderProvider } from '../../HeaderProvider/HeaderProvider';
import { Text } from 'wix-style-react';
import { dataHooks } from '../../../constants/dataHooks';
var EditorHeaderComponentClass = /** @class */ (function (_super) {
    __extends(EditorHeaderComponentClass, _super);
    function EditorHeaderComponentClass(props) {
        var _this = _super.call(this, props) || this;
        _this.notifyAndClose = function (event) { return _this.notifyAndCloseIfNotPrevented(event); };
        return _this;
    }
    EditorHeaderComponentClass.prototype.notifyAndCloseIfNotPrevented = function (event) {
        ModuleRegistry.notifyListeners(ON_CLOSE_IN_EDITOR, { event: event });
        if (!event.defaultPrevented) {
            this.closeBizMgr();
        }
    };
    /**
     * MEMO: if there will be more postMessages to send to the Editor - extract to a SDK
     */
    EditorHeaderComponentClass.prototype.closeBizMgr = function () {
        parent.postMessage('{"intent":"TPA2", "action":"CLOSE", "type":"closeWindow", "namespace": "Wix", "data": {}}', '*');
    };
    EditorHeaderComponentClass.prototype.render = function () {
        return (React.createElement("header", { className: css.header, "data-hook": dataHooks.Header },
            React.createElement(Text, { weight: "thin", size: "medium", light: true, dataHook: "header-title" }, this.props.t('header.editor-title')),
            React.createElement("div", { className: css.actions },
                React.createElement("div", { className: css.helpButton },
                    React.createElement(Help, { iconClassName: css.helpIconEditor, locale: this.props.locale, useIcon: true })),
                React.createElement("a", { "data-hook": "close-from-editor", onClick: this.notifyAndClose, className: css.button },
                    React.createElement("div", { className: css.closeButtonContent })))));
    };
    return EditorHeaderComponentClass;
}(React.Component));
export var EditorHeaderComponent = withTranslation()(EditorHeaderComponentClass);
export var EditorHeader = withHeaderProvider(EditorHeaderComponent, false);
export default EditorHeader;
