// @wix/answers-widget-sdk/dist/types/event_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventType = void 0;
var EventType;
(function (EventType) {
    EventType["FIRST_OPEN"] = "FIRST_OPEN";
    EventType["HOME_VIEW"] = "HOME_VIEW";
    EventType["CATEGORY_VIEW"] = "CATEGORY_VIEW";
    EventType["FEATURED_ARTICLE_CLICK"] = "FEATURED_ARTICLE_CLICK";
    EventType["HOME_CATEGORY_CLICK"] = "HOME_CATEGORY_CLICK";
    EventType["CATEGORY_LIST_CLICK"] = "CATEGORY_LIST_CLICK";
    EventType["ARTICLE_LIST_CLICK"] = "ARTICLE_LIST_CLICK";
    EventType["ARTICLE_VIEW"] = "ARTICLE_VIEW";
    EventType["SEARCH_IGNORED"] = "SEARCH_IGNORED";
    EventType["SEARCH_NO_RESULT"] = "SEARCH_NO_RESULT";
    EventType["SEARCH_RESULT_CLICK"] = "SEARCH_RESULT_CLICK";
    EventType["SEARCH"] = "SEARCH";
    EventType["SEARCH_RESULT"] = "SEARCH_RESULT";
    EventType["CHAT_VIEW"] = "CHAT_VIEW";
    EventType["CHAT_CLOSED"] = "CHAT_CLOSED";
    EventType["CHAT_STARTED"] = "CHAT_STARTED";
    EventType["CHAT_STARTED_SUCCESS"] = "CHAT_STARTED_SUCCESS";
    EventType["CHAT_STARTED_FAIL"] = "CHAT_STARTED_FAIL";
    EventType["SUBSCRIBE_USER_TIMELINE"] = "SUBSCRIBE_USER_TIMELINE";
    EventType["SUBSCRIBE_USER_TIMELINE_SUCCESS"] = "SUBSCRIBE_USER_TIMELINE_SUCCESS";
    EventType["SUBSCRIBE_USER_TIMELINE_FAIL"] = "SUBSCRIBE_USER_TIMELINE_FAIL";
    EventType["SEND_CHAT_MESSAGE"] = "SEND_CHAT_MESSAGE";
    EventType["SEND_CHAT_MESSAGE_SUCCESS"] = "SEND_CHAT_MESSAGE_SUCCESS";
    EventType["SEND_CHAT_MESSAGE_FAIL"] = "SEND_CHAT_MESSAGE_FAIL";
    EventType["RECEIVE_CHAT_MESSAGE"] = "RECEIVE_CHAT_MESSAGE";
    EventType["PUBSUB_EVENT"] = "PUBSUB_EVENT";
    EventType["CHAT_UNLOAD"] = "CHAT_UNLOAD";
    EventType["CHAT_HIDDEN"] = "CHAT_HIDDEN";
    EventType["CHAT_VISIBLE"] = "CHAT_VISIBLE";
    EventType["CHAT_TIMELINE_UPDATED"] = "CHAT_TIMELINE_UPDATED";
    EventType["CHAT_ONLINE"] = "CHAT_ONLINE";
    EventType["SUBMIT_CHAT_FORM"] = "SUBMIT_CHAT_FORM";
    EventType["GET_AUTH_TOKEN"] = "GET_AUTH_TOKEN";
    EventType["GET_AUTH_TOKEN_SUCCESS"] = "GET_AUTH_TOKEN_SUCCESS";
    EventType["GET_AUTH_TOKEN_FAIL"] = "GET_AUTH_TOKEN_FAIL";
    EventType["CALLBACK_VIEW"] = "CALLBACK_VIEW";
    EventType["CALLBACK_SUBMIT"] = "CALLBACK_SUBMIT";
    EventType["CALLBACK_SUBMIT_SUCCESS"] = "CALLBACK_SUBMIT_SUCCESS";
    EventType["CALLBACK_SUBMIT_FAIL"] = "CALLBACK_SUBMIT_FAIL";
    EventType["CALLBACK_VERIFICATION_SUBMIT"] = "CALLBACK_VERIFICATION_SUBMIT";
    EventType["CALLBACK_VERIFICATION_SUBMIT_SUCCESS"] = "CALLBACK_VERIFICATION_SUBMIT_SUCCESS";
    EventType["CALLBACK_VERIFICATION_SUBMIT_FAIL"] = "CALLBACK_VERIFICATION_SUBMIT_FAIL";
    EventType["RESEND_VERIFICATION"] = "RESEND_VERIFICATION";
    EventType["RESEND_VERIFICATION_SUCCESS"] = "RESEND_VERIFICATION_SUCCESS";
    EventType["RESEND_VERIFICATION_FAIL"] = "RESEND_VERIFICATION_FAIL";
    EventType["CHANGE_VERIFICATION_MODE"] = "CHANGE_VERIFICATION_MODE";
    EventType["TICKET_FORM_VIEW"] = "TICKET_FORM_VIEW";
    EventType["TICKET_FORM_SUBMIT"] = "TICKET_FORM_SUBMIT";
    EventType["TICKET_FORM_SUBMIT_SUCCESS"] = "TICKET_FORM_SUBMIT_SUCCESS";
    EventType["TICKET_FORM_SUBMIT_FAIL"] = "TICKET_FORM_SUBMIT_FAIL";
    EventType["TICKET_FORM_CAPTCHA_RESOLVED"] = "TICKET_FORM_CAPTCHA_RESOLVED";
    EventType["TICKET_FORM_CUSTOM_FIELD_CHANGED"] = "TICKET_FORM_CUSTOM_FIELD_CHANGED";
    EventType["CUSTOM_EVENT"] = "CUSTOM_EVENT";
    EventType["NAVIGATION"] = "NAVIGATION";
    EventType["FEATURED_ARTICLES_UPDATED"] = "FEATURED_ARTICLES_UPDATED";
})(EventType = exports.EventType || (exports.EventType = {}));
