// @wix/answers-widget-sdk/dist/types/message_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WidgetMessageType = void 0;
var WidgetMessageType;
(function (WidgetMessageType) {
    WidgetMessageType["IDENTIFY_USER"] = "identify-user";
    WidgetMessageType["SEND_CHAT_MESSAGE"] = "send-chat-message";
    WidgetMessageType["SET_CUSTOM_FIELDS"] = "set-custom-fields";
    WidgetMessageType["ON_CHAT_STARTED"] = "on-chat-started";
    WidgetMessageType["CHAT_MESSAGE_RECEIVED"] = "chat-message-received";
    WidgetMessageType["CHAT_MESSAGE_SENT"] = "chat-message-sent";
    WidgetMessageType["SET_CUSTOM_KB_HEADER"] = "set-custom-kb-header";
    WidgetMessageType["SET_CUSTOM_HOME_PAGE_HEADER"] = "set-custom-home-page";
    WidgetMessageType["AUTHENTICATOR_ADDED"] = "authenticator-added";
    WidgetMessageType["VIEW_LOADED"] = "view-loaded";
    WidgetMessageType["VIEW_VISIBLE"] = "view-visible";
    WidgetMessageType["AUTHENTICATION_REQUEST"] = "authentication-request";
    WidgetMessageType["AUTHENTICATION_RESPONSE"] = "authentication-response";
    WidgetMessageType["ACTIVE_CHAT_COUNT"] = "active-chat-count";
    WidgetMessageType["SET_OVERRIDE_TEXTS"] = "set-override-texts";
    WidgetMessageType["LOCATION_CHANGED"] = "location-changed";
    WidgetMessageType["GO_TO_CATEGORY"] = "go-to-category";
    WidgetMessageType["GO_TO_ARTICLE"] = "go-to-article";
    WidgetMessageType["WINDOW_FOCUS"] = "window-focus";
    WidgetMessageType["WIDGET_OPEN"] = "widget-open";
    WidgetMessageType["UPDATE_DRAG_AREA"] = "update-drag-area";
    WidgetMessageType["COLLAPSE_WIDGET"] = "collapse-widget";
    WidgetMessageType["SET_LABELS"] = "set-labels";
    WidgetMessageType["DISPLAY_IMAGE"] = "display-image";
    WidgetMessageType["CHATBOT_INITIATED"] = "chatbot-initiated";
    WidgetMessageType["CHATBOT_USER_INPUT"] = "chatbot-user-input";
    WidgetMessageType["CHATBOT_NEXT_STEP"] = "chatbot-next-step";
    WidgetMessageType["GO_TO_CONTACT"] = "go-to-contact";
    WidgetMessageType["GO_TO_CHAT"] = "go-to-chat";
    WidgetMessageType["GO_TO_CHAT_LIST"] = "go-to-chat-list";
    WidgetMessageType["GO_TO_CALLBACK"] = "go-to-callback";
    WidgetMessageType["FILL_CONTACT_FORM"] = "fill-contact-form";
    WidgetMessageType["FEATURED_ARTICLE_SEARCH"] = "featured-article-search";
    WidgetMessageType["SET_FEATURED_ARTICLES"] = "set-featured-articles";
    WidgetMessageType["WIDGET_EVENT"] = "widget-event";
    WidgetMessageType["CONTACT_OPTIONS_REQUEST"] = "contact-options-request";
    WidgetMessageType["CONTACT_OPTIONS_RESPONSE"] = "contact-options-response";
    WidgetMessageType["PREVENT_REOPENING_CHAT"] = "prevent-reopening-chat";
    WidgetMessageType["SET_CUSTOM_FIELDS_CONFIGURATION"] = "set-custom-fields-configuration";
})(WidgetMessageType = exports.WidgetMessageType || (exports.WidgetMessageType = {}));
