// @wix/dashboard-smart-answers-widget/dist/esm/common/event-handlers.js
import { EventType, } from '@wix/answers-widget-sdk';
import { logArticleClicked, logFeaturedArticleClicked, logFeaturedArticlesViewd, logSearchCompleted, logSearchResultClicked, logSerachNoResult, } from '../services/bi-logger/connect-client-bi-logger';
import { logClickOnVisitHelpCenter } from '../services/bi-logger/my-account-bi-logger';
var handleArticleView = function (logger, _a) {
    var _b = _a.data, articleId = _b.articleId, _c = _b.article, title = _c.title, url = _c.url, articleLocale = _b.articleLocale;
    return logArticleClicked(logger, {
        item_id: articleId,
        item_name: title,
        lang: articleLocale,
        item_url: url,
    });
};
var handleFeaturedArticleClicked = function (logger, _a) {
    var _b = _a.data, position = _b.position, _c = _b.article, id = _c.id, locale = _c.locale;
    return logFeaturedArticleClicked(logger, {
        item_id: id,
        lang: locale,
        feature_article_position: position,
    });
};
var handleSearchResultClick = function (logger, _a) {
    var _b = _a.data, term = _b.term, _c = _b.article, title = _c.title, url = _c.url, locale = _c.locale, position = _b.position;
    return logSearchResultClicked(logger, {
        question: term,
        answer: title,
        item_url: url,
        order: position.toString(),
        lang: locale,
    });
};
var handleSearchResult = function (logger, _a) {
    var _b, _c;
    var _d = _a.data, term = _d.term, articles = _d.articles;
    return logSearchCompleted(logger, {
        question: term,
        first_answer: (_b = articles[0]) === null || _b === void 0 ? void 0 : _b.title,
        lang: (_c = articles[0]) === null || _c === void 0 ? void 0 : _c.locale,
    });
};
var handleSearchNoResult = function (logger, _a) {
    var term = _a.data.term;
    return logSerachNoResult(logger, {
        question: term,
    });
};
var handleFeauredArticlesUpdate = function (logger, _a) {
    var articles = _a.data.articles;
    return logFeaturedArticlesViewd(logger, {
        featured_articles_item_ids: articles.map(function (article) { return article.id; }).join(),
        featured_articles_lang: articles.map(function (article) { return article.locale; }).join(),
    });
};
export var sendBIEvents = function (event, connectBiLogger, myAccountBiLogger) {
    switch (event.type) {
        case EventType.ARTICLE_VIEW:
            handleArticleView(connectBiLogger, event);
            break;
        case EventType.FEATURED_ARTICLE_CLICK:
            handleFeaturedArticleClicked(connectBiLogger, event);
            break;
        case EventType.SEARCH_RESULT_CLICK:
            handleSearchResultClick(connectBiLogger, event);
            break;
        case EventType.SEARCH_RESULT:
            handleSearchResult(connectBiLogger, event);
            break;
        case EventType.SEARCH_NO_RESULT:
            handleSearchNoResult(connectBiLogger, event);
            break;
        case EventType.CUSTOM_EVENT:
            logClickOnVisitHelpCenter(myAccountBiLogger, {});
            break;
        case EventType.FEATURED_ARTICLES_UPDATED:
            handleFeauredArticlesUpdate(connectBiLogger, event);
            break;
        default:
            break;
    }
};
