// @wix/my-account-header/dist/es/src/client/components/ExperimentsProvider/ExperimentsProvider.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import Experiments from '@wix/wix-experiments';
export var withExperimentsProvider = function (_a) {
    var scope = _a.scope;
    return function (WrappedComponent) {
        return /** @class */ (function (_super) {
            __extends(class_1, _super);
            function class_1(props) {
                var _this = _super.call(this, props) || this;
                _this.mounted = false;
                _this.mounted = true;
                if (props.experiments) {
                    _this.state = {
                        loading: false,
                        experiments: props.experiments
                    };
                }
                else {
                    _this.state = {
                        loading: true,
                        experiments: {}
                    };
                    var wixExperiments_1 = new Experiments({ requestContext: {} });
                    wixExperiments_1.load(scope).then(function () {
                        if (!_this.mounted) {
                            return;
                        }
                        var experiments = wixExperiments_1.all();
                        var isEmpty = Object.keys(experiments).length === 0;
                        _this.setState({
                            experiments: isEmpty ? {} : experiments,
                            loading: false
                        });
                    });
                }
                return _this;
            }
            class_1.prototype.componentWillUnmount = function () {
                this.mounted = false;
            };
            class_1.prototype.render = function () {
                if (this.state.loading) {
                    return null;
                }
                return React.createElement(WrappedComponent, __assign({}, this.props, { experiments: this.state.experiments }));
            };
            return class_1;
        }(React.Component));
    };
};
