// @wix/ambassador-dealer-offers-serving-service/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserStatusCriteriaEnum = exports.UserRoleCriteriaEnum = exports.Status = exports.ResponseTypeEnum = exports.PremiumLandingTabCriteriaEnum = exports.PremiumAccountCriteriaEnum = exports.Plan = exports.OperatorValueTypeEnum = exports.Operator = exports.MailboxCriteriaEnum = exports.FreeSiteCriteriaEnum = exports.EligibleForSaleCriteriaEnum = exports.CustomParamOperator = exports.CustomFilterDataSource = exports.CriteriaType = exports.CountType = exports.ConnectedDomainCriteriaEnum = exports.ConditionTypeEnum = exports.ComparatorTypeEnum = exports.Brand = void 0;
var Brand;
(function (Brand) {
    Brand["UNKNOWN"] = "UNKNOWN";
    Brand["WIX"] = "WIX";
    Brand["EDITORX"] = "EDITORX";
})(Brand = exports.Brand || (exports.Brand = {}));
var ComparatorTypeEnum;
(function (ComparatorTypeEnum) {
    ComparatorTypeEnum["UNKNOWN"] = "UNKNOWN";
    ComparatorTypeEnum["EQUAL"] = "EQUAL";
    ComparatorTypeEnum["CONTAINED_IN"] = "CONTAINED_IN";
    ComparatorTypeEnum["CONTAINS_SOME"] = "CONTAINS_SOME";
    ComparatorTypeEnum["CONTAINS_NONE"] = "CONTAINS_NONE";
    ComparatorTypeEnum["NOT_CONTAINS"] = "NOT_CONTAINS";
    ComparatorTypeEnum["NOT_CONTAINED_IN"] = "NOT_CONTAINED_IN";
    ComparatorTypeEnum["BE_NOTHING"] = "BE_NOTHING";
    ComparatorTypeEnum["EXISTS"] = "EXISTS";
    ComparatorTypeEnum["MORE_THAN_DAYS"] = "MORE_THAN_DAYS";
    ComparatorTypeEnum["WITHIN_DAYS"] = "WITHIN_DAYS";
    ComparatorTypeEnum["LESS"] = "LESS";
    ComparatorTypeEnum["LESS_EQUAL"] = "LESS_EQUAL";
    ComparatorTypeEnum["GREATER"] = "GREATER";
    ComparatorTypeEnum["GREATER_EQUAL"] = "GREATER_EQUAL";
    ComparatorTypeEnum["SIZE_LESS_EQUAL"] = "SIZE_LESS_EQUAL";
    ComparatorTypeEnum["SIZE_GREATER_EQUAL"] = "SIZE_GREATER_EQUAL";
    ComparatorTypeEnum["NOT_EQUAL"] = "NOT_EQUAL";
})(ComparatorTypeEnum = exports.ComparatorTypeEnum || (exports.ComparatorTypeEnum = {}));
var ConditionTypeEnum;
(function (ConditionTypeEnum) {
    ConditionTypeEnum["UNKNOWN"] = "UNKNOWN";
    ConditionTypeEnum["EQUAL"] = "EQUAL";
    ConditionTypeEnum["GREATER_THAN"] = "GREATER_THAN";
    ConditionTypeEnum["SMALLER_THAN"] = "SMALLER_THAN";
})(ConditionTypeEnum = exports.ConditionTypeEnum || (exports.ConditionTypeEnum = {}));
var ConnectedDomainCriteriaEnum;
(function (ConnectedDomainCriteriaEnum) {
    ConnectedDomainCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    ConnectedDomainCriteriaEnum["CONNECTED"] = "CONNECTED";
    ConnectedDomainCriteriaEnum["NOT_CONNECTED"] = "NOT_CONNECTED";
})(ConnectedDomainCriteriaEnum = exports.ConnectedDomainCriteriaEnum || (exports.ConnectedDomainCriteriaEnum = {}));
var CountType;
(function (CountType) {
    CountType["UNKNOWN"] = "UNKNOWN";
    CountType["TOTAL"] = "TOTAL";
    CountType["DAYS"] = "DAYS";
})(CountType = exports.CountType || (exports.CountType = {}));
var CriteriaType;
(function (CriteriaType) {
    CriteriaType["UNKNOWN"] = "UNKNOWN";
    CriteriaType["COMPOUND"] = "COMPOUND";
    CriteriaType["CUSTOM"] = "CUSTOM";
    CriteriaType["GEO"] = "GEO";
    CriteriaType["LANGUAGE"] = "LANGUAGE";
    CriteriaType["USER_ROLE"] = "USER_ROLE";
    CriteriaType["USER_STATUS"] = "USER_STATUS";
    CriteriaType["USER_COLORS"] = "USER_COLORS";
    CriteriaType["USER_GUIDS"] = "USER_GUIDS";
    CriteriaType["TEMPLATE_GUIDS"] = "TEMPLATE_GUIDS";
    CriteriaType["POST_REG"] = "POST_REG";
    CriteriaType["CONNECTED_DOMAIN"] = "CONNECTED_DOMAIN";
    CriteriaType["PREMIUM_PLAN"] = "PREMIUM_PLAN";
    CriteriaType["EXCLUDE_IN_SALE"] = "EXCLUDE_IN_SALE";
    CriteriaType["MAILBOX"] = "MAILBOX";
    CriteriaType["CONNECTED_OFFERING_GROUP"] = "CONNECTED_OFFERING_GROUP";
    CriteriaType["INSTALLED_APPS"] = "INSTALLED_APPS";
    CriteriaType["PRODUCT_GROUPS"] = "PRODUCT_GROUPS";
    CriteriaType["DOMAIN_LENGTH"] = "DOMAIN_LENGTH";
    CriteriaType["MAILBOX_SALE"] = "MAILBOX_SALE";
    CriteriaType["VOUCHERS"] = "VOUCHERS";
    CriteriaType["FREE_SITE"] = "FREE_SITE";
    CriteriaType["BRAND"] = "BRAND";
    CriteriaType["EXPERIMENT"] = "EXPERIMENT";
    CriteriaType["CUSTOM_PARAM"] = "CUSTOM_PARAM";
    CriteriaType["AUDIENCE"] = "AUDIENCE";
    CriteriaType["PREMIUM_LANDING_TAB"] = "PREMIUM_LANDING_TAB";
    CriteriaType["PREMIUM_ACCOUNT"] = "PREMIUM_ACCOUNT";
    CriteriaType["ELIGIBLE_FOR_SALE"] = "ELIGIBLE_FOR_SALE";
})(CriteriaType = exports.CriteriaType || (exports.CriteriaType = {}));
var CustomFilterDataSource;
(function (CustomFilterDataSource) {
    CustomFilterDataSource["UNKNOWN"] = "UNKNOWN";
    CustomFilterDataSource["USER_PROFILE"] = "USER_PROFILE";
    CustomFilterDataSource["SITE_PROFILE"] = "SITE_PROFILE";
})(CustomFilterDataSource = exports.CustomFilterDataSource || (exports.CustomFilterDataSource = {}));
var CustomParamOperator;
(function (CustomParamOperator) {
    CustomParamOperator["UNKNOWN"] = "UNKNOWN";
    CustomParamOperator["EQUALS"] = "EQUALS";
    CustomParamOperator["NOT_EQUALS"] = "NOT_EQUALS";
    CustomParamOperator["CONTAINS_ALL"] = "CONTAINS_ALL";
    CustomParamOperator["NOT_CONTAINS_ALL"] = "NOT_CONTAINS_ALL";
    CustomParamOperator["CONTAINS_ANY"] = "CONTAINS_ANY";
    CustomParamOperator["NOT_CONTAINS_ANY"] = "NOT_CONTAINS_ANY";
    CustomParamOperator["EXISTS"] = "EXISTS";
    CustomParamOperator["NOT_EXISTS"] = "NOT_EXISTS";
})(CustomParamOperator = exports.CustomParamOperator || (exports.CustomParamOperator = {}));
var EligibleForSaleCriteriaEnum;
(function (EligibleForSaleCriteriaEnum) {
    EligibleForSaleCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    EligibleForSaleCriteriaEnum["NONE"] = "NONE";
    EligibleForSaleCriteriaEnum["SEASONAL"] = "SEASONAL";
    EligibleForSaleCriteriaEnum["PERSONAL"] = "PERSONAL";
    EligibleForSaleCriteriaEnum["CUSTOM"] = "CUSTOM";
})(EligibleForSaleCriteriaEnum = exports.EligibleForSaleCriteriaEnum || (exports.EligibleForSaleCriteriaEnum = {}));
var FreeSiteCriteriaEnum;
(function (FreeSiteCriteriaEnum) {
    FreeSiteCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    FreeSiteCriteriaEnum["HAS_FREE"] = "HAS_FREE";
    FreeSiteCriteriaEnum["NO_FREE"] = "NO_FREE";
})(FreeSiteCriteriaEnum = exports.FreeSiteCriteriaEnum || (exports.FreeSiteCriteriaEnum = {}));
var MailboxCriteriaEnum;
(function (MailboxCriteriaEnum) {
    MailboxCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    MailboxCriteriaEnum["HAS_MAILBOX"] = "HAS_MAILBOX";
    MailboxCriteriaEnum["NO_MAILBOX"] = "NO_MAILBOX";
})(MailboxCriteriaEnum = exports.MailboxCriteriaEnum || (exports.MailboxCriteriaEnum = {}));
var Operator;
(function (Operator) {
    Operator["UNKNOWN"] = "UNKNOWN";
    Operator["OR"] = "OR";
    Operator["AND"] = "AND";
})(Operator = exports.Operator || (exports.Operator = {}));
var OperatorValueTypeEnum;
(function (OperatorValueTypeEnum) {
    OperatorValueTypeEnum["UNKNOWN"] = "UNKNOWN";
    OperatorValueTypeEnum["NUMBER"] = "NUMBER";
    OperatorValueTypeEnum["STRING"] = "STRING";
    OperatorValueTypeEnum["LIST"] = "LIST";
    OperatorValueTypeEnum["TIMESTAMP"] = "TIMESTAMP";
    OperatorValueTypeEnum["BOOLEAN"] = "BOOLEAN";
})(OperatorValueTypeEnum = exports.OperatorValueTypeEnum || (exports.OperatorValueTypeEnum = {}));
var Plan;
(function (Plan) {
    Plan["UNKNOWN"] = "UNKNOWN";
    Plan["PREMIUM"] = "PREMIUM";
    Plan["NONE"] = "NONE";
})(Plan = exports.Plan || (exports.Plan = {}));
var PremiumAccountCriteriaEnum;
(function (PremiumAccountCriteriaEnum) {
    PremiumAccountCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    PremiumAccountCriteriaEnum["IS_PREMIUM"] = "IS_PREMIUM";
    PremiumAccountCriteriaEnum["IS_NOT_PREMIUM"] = "IS_NOT_PREMIUM";
})(PremiumAccountCriteriaEnum = exports.PremiumAccountCriteriaEnum || (exports.PremiumAccountCriteriaEnum = {}));
var PremiumLandingTabCriteriaEnum;
(function (PremiumLandingTabCriteriaEnum) {
    PremiumLandingTabCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    PremiumLandingTabCriteriaEnum["WEBSITE"] = "WEBSITE";
    PremiumLandingTabCriteriaEnum["BUSINESS"] = "BUSINESS";
})(PremiumLandingTabCriteriaEnum = exports.PremiumLandingTabCriteriaEnum || (exports.PremiumLandingTabCriteriaEnum = {}));
var ResponseTypeEnum;
(function (ResponseTypeEnum) {
    ResponseTypeEnum["UNKNOWN"] = "UNKNOWN";
    ResponseTypeEnum["STRING"] = "STRING";
    ResponseTypeEnum["NUMBER"] = "NUMBER";
})(ResponseTypeEnum = exports.ResponseTypeEnum || (exports.ResponseTypeEnum = {}));
var Status;
(function (Status) {
    Status["UNKNOWN"] = "UNKNOWN";
    Status["ALL"] = "ALL";
    Status["WIX_EMPLOYEE"] = "WIX_EMPLOYEE";
})(Status = exports.Status || (exports.Status = {}));
var UserRoleCriteriaEnum;
(function (UserRoleCriteriaEnum) {
    UserRoleCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    UserRoleCriteriaEnum["OWNER"] = "OWNER";
    UserRoleCriteriaEnum["NOT_OWNER"] = "NOT_OWNER";
})(UserRoleCriteriaEnum = exports.UserRoleCriteriaEnum || (exports.UserRoleCriteriaEnum = {}));
var UserStatusCriteriaEnum;
(function (UserStatusCriteriaEnum) {
    UserStatusCriteriaEnum["UNKNOWN"] = "UNKNOWN";
    UserStatusCriteriaEnum["NEW"] = "NEW";
})(UserStatusCriteriaEnum = exports.UserStatusCriteriaEnum || (exports.UserStatusCriteriaEnum = {}));
