// @wix/my-account-header/dist/es/src/client/components/MySitesDropdown/MySitesDropdown.data-hooks.js
export var MySitesDropdownDataHooks;
(function (MySitesDropdownDataHooks) {
    MySitesDropdownDataHooks["Trigger"] = "my-sites-dropdown-trigger";
    MySitesDropdownDataHooks["CurrentSiteTrigger"] = "current-site-my-sites-dropdown-trigger";
    MySitesDropdownDataHooks["Dropdown"] = "my-sites-dropdown";
    MySitesDropdownDataHooks["CreateNewSiteButton"] = "create-new-site-button";
    MySitesDropdownDataHooks["Site"] = "my-sites-dropdown-a-site";
    MySitesDropdownDataHooks["SiteItemDiv"] = "my-sites-dropdown-a-site-div";
    MySitesDropdownDataHooks["SiteName"] = "my-sites-dropdown-site-name";
    MySitesDropdownDataHooks["PremiumBadge"] = "my-sites-dropdown-site-premium-badge";
    MySitesDropdownDataHooks["Image"] = "my-sites-dropdown-site-image";
    MySitesDropdownDataHooks["GoToAllSites"] = "my-sites-dropdown-go-to-all-sites";
    MySitesDropdownDataHooks["CurrentSiteMarker"] = "my-sites-dropdown-current-site-marker";
    MySitesDropdownDataHooks["SiteDivider"] = "my-sites-dropdown-site-divider";
    MySitesDropdownDataHooks["noSitesState"] = "my-sites-dropdown-no-sites-content";
    MySitesDropdownDataHooks["Header"] = "my-sites-dropdown-header";
    MySitesDropdownDataHooks["SitesList"] = "my-sites-dropdown-sites-list";
    MySitesDropdownDataHooks["NoSiteHeading"] = "my-sites-dropdown-error-heading";
    MySitesDropdownDataHooks["NoSitesText"] = "my-sites-dropdown-error-first-text";
    MySitesDropdownDataHooks["Loader"] = "my-sites-dropdown-loader";
    MySitesDropdownDataHooks["MySitesTrigger"] = "my-sites-text-trigger";
    MySitesDropdownDataHooks["SiteNameLoader"] = "my-sites-site-name-loader";
})(MySitesDropdownDataHooks || (MySitesDropdownDataHooks = {}));
