// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/CreateNewSite/getCreateNewSiteAction.js
import * as React from 'react';
import { TOrigin, ActionType } from '../../../../types';
import SiteAddSmall from 'wix-ui-icons-common/SiteAddSmall';
import { PERMISSIONS } from '../../../../consts/permissions';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { BIEvents } from '../../../../consts/bi';
import { sendSiteActionsBI } from '../../../../services/BIService';
import { createSiteUrl, newCreateSiteUrl } from '../../../../consts/urls';
function onCreateNewSiteClicked(params, experiments) {
    var origin = params.origin, msid = params.quickActionsData.metaSiteId;
    sendSiteActionsBI({
        evid: BIEvents.quickActionClicked,
        name: 'quick_actions.create_site',
        msid: msid,
        origin: origin
    });
    sendSiteActionsBI({
        evid: BIEvents.createNewSiteClicked,
        origin: 'quick_action'
    });
    var url = experiments['specs.funnel.EnableNewCreateSiteUrl'] === 'true' ? newCreateSiteUrl : createSiteUrl;
    window.open(url);
}
export var getCreateNewSiteAction = function (origin, experiments) {
    var showCreateNewSite = origin === TOrigin.Dashboard;
    return {
        type: ActionType.click,
        actionName: 'create-new-site',
        textKey: 'quickActions.create-new-site',
        icon: React.createElement(SiteAddSmall, null),
        isPermitted: PermissionMethods.withPermissions(PERMISSIONS.CREATE_SITE),
        visible: showCreateNewSite,
        onActionClick: function (params) { return onCreateNewSiteClicked(params, experiments); }
    };
};
