// yoshi-flow-bm-runtime/build/esm/flow-api/createModuleFlowAPI.js
import { BrowserClient } from '@sentry/browser';
import { once } from 'lodash';
import { assertEnabledApi } from '../util/assert-enabled-api';
export function createModuleFlowAPI({ module, getFlowAPIInstances, bmModule, optionalDeps, useEssentials, }) {
    const { config } = module;
    const { sentry: sentryConfig, appDefId } = config;
    const { HttpClient, loadOptionalFlowAPIDeps } = optionalDeps;
    const sentry = (sentryConfig === null || sentryConfig === void 0 ? void 0 : sentryConfig.DSN) ? useEssentials
        ? bmModule._essentials.createSentryLogger({
            dsn: sentryConfig.DSN,
        })
        : new BrowserClient({ dsn: sentryConfig.DSN })
        : undefined;
    const httpClient = HttpClient ? new HttpClient(appDefId) : undefined;
    const getModuleFedopsLogger = once(async () => {
        const { createFedopsLogger } = await loadOptionalFlowAPIDeps();
        assertEnabledApi(createFedopsLogger, 'getModuleFedopsLogger');
        return createFedopsLogger(module.moduleId, Object.assign({}, module.config.fedops));
    });
    const getI18n = once(async () => {
        const { i18n } = await getFlowAPIInstances();
        if (!i18n.isInitialized) {
            await i18n.init();
        }
        return i18n;
    });
    const getBILogger = async () => {
        const { biLogger } = await getFlowAPIInstances();
        return biLogger;
    };
    const getExperiments = once(async () => {
        var _a;
        const { experiments } = await getFlowAPIInstances();
        const scopes = (_a = config.experiments) === null || _a === void 0 ? void 0 : _a.scopes;
        if (scopes != null) {
            await Promise.all(scopes.map((scope) => experiments === null || experiments === void 0 ? void 0 : experiments.load(scope)));
        }
        return experiments;
    });
    return {
        getBILogger,
        getI18n,
        getExperiments,
        getModuleFedopsLogger,
        module: bmModule,
        get moduleParams() {
            return bmModule.moduleParams;
        },
        sentry: sentry,
        get httpClient() {
            assertEnabledApi(httpClient, 'httpClient');
            return httpClient;
        },
        moduleInfo: module.config,
    };
}
