// @wix/ambassador-payment-settings-web/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1RestrictedGoodsFlowStatus = exports.V1RestrictedGoodsFlowCategory = exports.V1PartnerFlowApprovedPartnerServices = exports.V1CbdFlowStatus = exports.RequiredAction = exports.ProviderAccountStatus = exports.PaymentMethodStatus = exports.FieldType = exports.ApiRestrictedGoodsFlowStatus = exports.ApiRestrictedGoodsFlowCategory = exports.ApiPartnerFlowApprovedPartnerServices = exports.ApiCbdFlowStatus = exports.AccountStatus = void 0;
var AccountStatus;
(function (AccountStatus) {
    AccountStatus["UNDEFINED"] = "UNDEFINED";
    AccountStatus["CLAIMED"] = "CLAIMED";
    AccountStatus["NOT_CLAIMED"] = "NOT_CLAIMED";
    AccountStatus["MUST_BE_CLAIMED"] = "MUST_BE_CLAIMED";
    AccountStatus["FUNDS_HELD"] = "FUNDS_HELD";
    AccountStatus["BLOCKED"] = "BLOCKED";
    AccountStatus["SUSPENDED"] = "SUSPENDED";
    AccountStatus["NA"] = "NA";
})(AccountStatus = exports.AccountStatus || (exports.AccountStatus = {}));
var ApiCbdFlowStatus;
(function (ApiCbdFlowStatus) {
    ApiCbdFlowStatus["UNDEFINED"] = "UNDEFINED";
    ApiCbdFlowStatus["COMPLETED_ATTESTATION_FORM"] = "COMPLETED_ATTESTATION_FORM";
    ApiCbdFlowStatus["CONFIRMED_CBD"] = "CONFIRMED_CBD";
    ApiCbdFlowStatus["POSSIBLE_CBD"] = "POSSIBLE_CBD";
    ApiCbdFlowStatus["NO_CBD"] = "NO_CBD";
})(ApiCbdFlowStatus = exports.ApiCbdFlowStatus || (exports.ApiCbdFlowStatus = {}));
var ApiPartnerFlowApprovedPartnerServices;
(function (ApiPartnerFlowApprovedPartnerServices) {
    ApiPartnerFlowApprovedPartnerServices["NOT_ONLY_APPROVED_PARTNER_SERVICES_V1"] = "NOT_ONLY_APPROVED_PARTNER_SERVICES_V1";
    ApiPartnerFlowApprovedPartnerServices["ONLY_APPROVED_PARTNER_SERVICES_V1"] = "ONLY_APPROVED_PARTNER_SERVICES_V1";
    ApiPartnerFlowApprovedPartnerServices["UNDEFINED_APPROVED_PARTNER_SERVICES"] = "UNDEFINED_APPROVED_PARTNER_SERVICES";
})(ApiPartnerFlowApprovedPartnerServices = exports.ApiPartnerFlowApprovedPartnerServices || (exports.ApiPartnerFlowApprovedPartnerServices = {}));
var ApiRestrictedGoodsFlowCategory;
(function (ApiRestrictedGoodsFlowCategory) {
    ApiRestrictedGoodsFlowCategory["TRAVEL_AGENCIES"] = "TRAVEL_AGENCIES";
    ApiRestrictedGoodsFlowCategory["ADULT"] = "ADULT";
    ApiRestrictedGoodsFlowCategory["TOBACCO_ALCOHOL"] = "TOBACCO_ALCOHOL";
    ApiRestrictedGoodsFlowCategory["FIREARMS_WEAPONS"] = "FIREARMS_WEAPONS";
    ApiRestrictedGoodsFlowCategory["UNDEFINED_CATEGORY"] = "UNDEFINED_CATEGORY";
    ApiRestrictedGoodsFlowCategory["FINANCIAL"] = "FINANCIAL";
    ApiRestrictedGoodsFlowCategory["GAMBLING_LOTTERIES_SKILL_GAMES"] = "GAMBLING_LOTTERIES_SKILL_GAMES";
    ApiRestrictedGoodsFlowCategory["BINARY_OPTIONS_CRYPTOCURRENCIES"] = "BINARY_OPTIONS_CRYPTOCURRENCIES";
    ApiRestrictedGoodsFlowCategory["MEDICAL"] = "MEDICAL";
    ApiRestrictedGoodsFlowCategory["OTHER"] = "OTHER";
    ApiRestrictedGoodsFlowCategory["MARKETPLACES"] = "MARKETPLACES";
    ApiRestrictedGoodsFlowCategory["CBD"] = "CBD";
})(ApiRestrictedGoodsFlowCategory = exports.ApiRestrictedGoodsFlowCategory || (exports.ApiRestrictedGoodsFlowCategory = {}));
var ApiRestrictedGoodsFlowStatus;
(function (ApiRestrictedGoodsFlowStatus) {
    ApiRestrictedGoodsFlowStatus["NOT_RESTRICTED"] = "NOT_RESTRICTED";
    ApiRestrictedGoodsFlowStatus["RESTRICTED"] = "RESTRICTED";
    ApiRestrictedGoodsFlowStatus["UNDEFINED"] = "UNDEFINED";
})(ApiRestrictedGoodsFlowStatus = exports.ApiRestrictedGoodsFlowStatus || (exports.ApiRestrictedGoodsFlowStatus = {}));
var FieldType;
(function (FieldType) {
    FieldType["CHECKBOX"] = "CHECKBOX";
    FieldType["DROPDOWN"] = "DROPDOWN";
    FieldType["STRING"] = "STRING";
    FieldType["RICH_TEXT"] = "RICH_TEXT";
})(FieldType = exports.FieldType || (exports.FieldType = {}));
var PaymentMethodStatus;
(function (PaymentMethodStatus) {
    PaymentMethodStatus["UNDEFINED"] = "UNDEFINED";
    PaymentMethodStatus["NOT_CONNECTED"] = "NOT_CONNECTED";
    PaymentMethodStatus["SELECTED"] = "SELECTED";
    PaymentMethodStatus["CONNECTED"] = "CONNECTED";
})(PaymentMethodStatus = exports.PaymentMethodStatus || (exports.PaymentMethodStatus = {}));
var ProviderAccountStatus;
(function (ProviderAccountStatus) {
    ProviderAccountStatus["OK"] = "OK";
    ProviderAccountStatus["ACCOUNT_CREATION_PENDING"] = "ACCOUNT_CREATION_PENDING";
    ProviderAccountStatus["ACCOUNT_CREATION_FAILED"] = "ACCOUNT_CREATION_FAILED";
    ProviderAccountStatus["PAYMENTS_BLOCKED"] = "PAYMENTS_BLOCKED";
    ProviderAccountStatus["BLOCKED"] = "BLOCKED";
    ProviderAccountStatus["ACCEPTING_PAYINS_ONLY"] = "ACCEPTING_PAYINS_ONLY";
    ProviderAccountStatus["UNDEFINED"] = "UNDEFINED";
    ProviderAccountStatus["SUSPENDED"] = "SUSPENDED";
})(ProviderAccountStatus = exports.ProviderAccountStatus || (exports.ProviderAccountStatus = {}));
var RequiredAction;
(function (RequiredAction) {
    RequiredAction["CLAIM_ACCOUNT_TO_ENABLE_PAYIN"] = "CLAIM_ACCOUNT_TO_ENABLE_PAYIN";
    RequiredAction["NONE"] = "NONE";
    RequiredAction["CLAIM_ACCOUNT_TO_ENABLE_PAYOUT"] = "CLAIM_ACCOUNT_TO_ENABLE_PAYOUT";
    RequiredAction["CLAIM_ACCOUNT_TO_ENABLE_PAYOUT_AND_AVOID_SUSPENSION"] = "CLAIM_ACCOUNT_TO_ENABLE_PAYOUT_AND_AVOID_SUSPENSION";
})(RequiredAction = exports.RequiredAction || (exports.RequiredAction = {}));
var V1CbdFlowStatus;
(function (V1CbdFlowStatus) {
    V1CbdFlowStatus["UNDEFINED"] = "UNDEFINED";
    V1CbdFlowStatus["NO_CBD"] = "NO_CBD";
    V1CbdFlowStatus["POSSIBLE_CBD"] = "POSSIBLE_CBD";
    V1CbdFlowStatus["CONFIRMED_CBD"] = "CONFIRMED_CBD";
    V1CbdFlowStatus["COMPLETED_ATTESTATION_FORM"] = "COMPLETED_ATTESTATION_FORM";
})(V1CbdFlowStatus = exports.V1CbdFlowStatus || (exports.V1CbdFlowStatus = {}));
var V1PartnerFlowApprovedPartnerServices;
(function (V1PartnerFlowApprovedPartnerServices) {
    V1PartnerFlowApprovedPartnerServices["UNDEFINED_APPROVED_PARTNER_SERVICES"] = "UNDEFINED_APPROVED_PARTNER_SERVICES";
    V1PartnerFlowApprovedPartnerServices["ONLY_APPROVED_PARTNER_SERVICES_V1"] = "ONLY_APPROVED_PARTNER_SERVICES_V1";
    V1PartnerFlowApprovedPartnerServices["NOT_ONLY_APPROVED_PARTNER_SERVICES_V1"] = "NOT_ONLY_APPROVED_PARTNER_SERVICES_V1";
})(V1PartnerFlowApprovedPartnerServices = exports.V1PartnerFlowApprovedPartnerServices || (exports.V1PartnerFlowApprovedPartnerServices = {}));
var V1RestrictedGoodsFlowCategory;
(function (V1RestrictedGoodsFlowCategory) {
    V1RestrictedGoodsFlowCategory["UNDEFINED_CATEGORY"] = "UNDEFINED_CATEGORY";
    V1RestrictedGoodsFlowCategory["TOBACCO_ALCOHOL"] = "TOBACCO_ALCOHOL";
    V1RestrictedGoodsFlowCategory["FIREARMS_WEAPONS"] = "FIREARMS_WEAPONS";
    V1RestrictedGoodsFlowCategory["ADULT"] = "ADULT";
    V1RestrictedGoodsFlowCategory["MEDICAL"] = "MEDICAL";
    V1RestrictedGoodsFlowCategory["FINANCIAL"] = "FINANCIAL";
    V1RestrictedGoodsFlowCategory["TRAVEL_AGENCIES"] = "TRAVEL_AGENCIES";
    V1RestrictedGoodsFlowCategory["GAMBLING_LOTTERIES_SKILL_GAMES"] = "GAMBLING_LOTTERIES_SKILL_GAMES";
    V1RestrictedGoodsFlowCategory["BINARY_OPTIONS_CRYPTOCURRENCIES"] = "BINARY_OPTIONS_CRYPTOCURRENCIES";
    V1RestrictedGoodsFlowCategory["MARKETPLACES"] = "MARKETPLACES";
    V1RestrictedGoodsFlowCategory["OTHER"] = "OTHER";
    V1RestrictedGoodsFlowCategory["CBD"] = "CBD";
})(V1RestrictedGoodsFlowCategory = exports.V1RestrictedGoodsFlowCategory || (exports.V1RestrictedGoodsFlowCategory = {}));
var V1RestrictedGoodsFlowStatus;
(function (V1RestrictedGoodsFlowStatus) {
    V1RestrictedGoodsFlowStatus["UNDEFINED"] = "UNDEFINED";
    V1RestrictedGoodsFlowStatus["NOT_RESTRICTED"] = "NOT_RESTRICTED";
    V1RestrictedGoodsFlowStatus["RESTRICTED"] = "RESTRICTED";
})(V1RestrictedGoodsFlowStatus = exports.V1RestrictedGoodsFlowStatus || (exports.V1RestrictedGoodsFlowStatus = {}));
