// @wix/wix-axios-config/dist/es/src/wix-axios-instance-config.js
import interceptorsLoader from './axios-interceptors-loader';
import wixResultInterceptor from './axios-intercept-wix-result';
import baseurlInterceptor from './axios-intercept-baseurl';
import multilingualInterceptor from './axios-intercept-multilingual';
import artifactIdInterceptor from './axios-intercept-artifactid';
import siteRevisionInterceptor from './axios-intercept-site-revision';
import wixHeadersInterceptor from './axios-intercept-wix-headers';
export default (function (axios, _ref) {
  var req = _ref.req,
      adapter = _ref.adapter,
      baseURL = _ref.baseURL,
      lang = _ref.lang,
      locale = _ref.locale,
      isPrimaryLanguage = _ref.isPrimaryLanguage,
      siteInstance = _ref.siteInstance,
      signedInstance = _ref.signedInstance,
      _ref$loadInterceptors = _ref.loadInterceptors,
      loadInterceptors = _ref$loadInterceptors === void 0 ? true : _ref$loadInterceptors,
      wixCodeApi = _ref.wixCodeApi,
      _ref$shouldAddArtifac = _ref.shouldAddArtifactIdOverride,
      shouldAddArtifactIdOverride = _ref$shouldAddArtifac === void 0 ? false : _ref$shouldAddArtifac;
  var config = {
    baseURL: baseURL || axios.defaults.baseURL,
    adapter: adapter,
    lang: lang,
    locale: locale,
    isPrimaryLanguage: isPrimaryLanguage,
    siteInstance: siteInstance
  };

  if (req) {
    config.headers = req.headers;
  }

  var instance = axios.create(config);
  instance.unregisterWixInterceptors = interceptorsLoader.loadIfNeeded(instance, {
    shouldLoadInterceptors: loadInterceptors,
    interceptors: [wixResultInterceptor, baseurlInterceptor, multilingualInterceptor.getInterceptor({
      lang: lang,
      siteInstance: siteInstance,
      locale: locale,
      isPrimaryLanguage: isPrimaryLanguage
    }), artifactIdInterceptor.getInterceptor({
      shouldAddArtifactIdOverride: shouldAddArtifactIdOverride
    }), siteRevisionInterceptor.getInterceptor({
      wixCodeApi: wixCodeApi
    }), wixHeadersInterceptor.getInterceptor({
      signedInstance: signedInstance
    })]
  });
  return instance;
});
