// @wix/features-wizard-modal/dist/esm/hooks/use-features-wizard.hook.js
var _a;
import React, { useContext, createContext, useState } from 'react';
import { FontUpgrade } from 'wix-style-react';
import { FeaturesWizardModal } from '../components/modal';
var FeaturesWizardContext = createContext({});
export var ModalComponents;
(function (ModalComponents) {
    ModalComponents["FeaturesWizard"] = "features-wizard-app.components.features-wizard";
    ModalComponents["SiteType"] = "features-wizard-app.components.site-type";
})(ModalComponents || (ModalComponents = {}));
var componentsSizes = (_a = {},
    _a[ModalComponents.FeaturesWizard] = {
        width: 980,
        height: 660,
    },
    _a[ModalComponents.SiteType] = {
        width: 700,
        height: 475,
    },
    _a);
export var useFeaturesWizard = function () { return useContext(FeaturesWizardContext); };
export var FeaturesWizardProvider = function (_a) {
    var children = _a.children, componentName = _a.componentName;
    var _b = useState(false), isOpen = _b[0], setIsOpen = _b[1];
    var _c = componentsSizes[componentName], width = _c.width, height = _c.height;
    var open = function () {
        setIsOpen(true);
    };
    var close = function () {
        setIsOpen(false);
    };
    return (React.createElement(FeaturesWizardContext.Provider, { value: { open: open, close: close, width: width, height: height } },
        children,
        React.createElement(FontUpgrade, { active: true },
            React.createElement(FeaturesWizardModal, { componentName: componentName, isOpen: isOpen, onClose: close }))));
};
