// @wix/ambassador-serverless-ascend-calendar-service/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuggestionStatus = exports.SocialPostStatus = exports.EmailMarketingStatus = exports.DistributionOption = exports.Channel = exports.BlogPostStatus = void 0;
var BlogPostStatus;
(function (BlogPostStatus) {
    BlogPostStatus["Scheduled"] = "Scheduled";
    BlogPostStatus["Published"] = "Published";
    BlogPostStatus["Failed"] = "Failed";
})(BlogPostStatus = exports.BlogPostStatus || (exports.BlogPostStatus = {}));
var Channel;
(function (Channel) {
    Channel["EmailMarketing"] = "EmailMarketing";
    Channel["SocialPosts"] = "SocialPosts";
    Channel["BlogPosts"] = "BlogPosts";
    Channel["VideoMaker"] = "VideoMaker";
    Channel["Other"] = "Other";
})(Channel = exports.Channel || (exports.Channel = {}));
var DistributionOption;
(function (DistributionOption) {
    DistributionOption["Facebook"] = "Facebook";
    DistributionOption["Instagram"] = "Instagram";
    DistributionOption["Email"] = "Email";
    DistributionOption["Blog"] = "Blog";
})(DistributionOption = exports.DistributionOption || (exports.DistributionOption = {}));
var EmailMarketingStatus;
(function (EmailMarketingStatus) {
    EmailMarketingStatus["Scheduled"] = "Scheduled";
    EmailMarketingStatus["Published"] = "Published";
    EmailMarketingStatus["Failed"] = "Failed";
    EmailMarketingStatus["NotSentToAll"] = "NotSentToAll";
    EmailMarketingStatus["Sent"] = "Sent";
    EmailMarketingStatus["Declined"] = "Declined";
    EmailMarketingStatus["Other"] = "Other";
})(EmailMarketingStatus = exports.EmailMarketingStatus || (exports.EmailMarketingStatus = {}));
var SocialPostStatus;
(function (SocialPostStatus) {
    SocialPostStatus["Scheduled"] = "Scheduled";
    SocialPostStatus["Published"] = "Published";
    SocialPostStatus["Failed"] = "Failed";
})(SocialPostStatus = exports.SocialPostStatus || (exports.SocialPostStatus = {}));
var SuggestionStatus;
(function (SuggestionStatus) {
    SuggestionStatus["Suggested"] = "Suggested";
    SuggestionStatus["Completed"] = "Completed";
    SuggestionStatus["Dismissed"] = "Dismissed";
})(SuggestionStatus = exports.SuggestionStatus || (exports.SuggestionStatus = {}));
