// @wix/my-account-header/dist/es/src/client/components/HireAnExpertLink/HireAnExpertLink.js
import { __extends } from "tslib";
import * as React from 'react';
import { withTranslation } from '@wix/wix-i18n-config';
import HeaderLink from '../HeaderLink/HeaderLink.container';
import * as style from './HireAnExpertLink.scss';
import { connect } from 'react-redux';
import { showExpertExperience } from '../../selectors/expertsSelectors';
import { isNotCoBranded } from '../../selectors/coBrandingSelector';
import { isNtt } from '../../services/cookieService/cookieService';
import { experimentsSelectors } from '@wix/wix-experiments-redux';
export var deprecated_ARENA_URL = 'https://www.wix.com/marketplace/dynamic-brief?entry=generalMenu&serviceType=1&innerType=1&origin=os_header_non_partner';
export var ARENA_URL = 'https://www.wix.com/marketplace/general/site/briefRoot?origin=os_header_non_partner';
export var ARENA_MULTI_URL = ' https://www.wix.com/marketplace/multi/site/briefRoot?origin=os_header_non_partner';
var HireAnExpertLink = /** @class */ (function (_super) {
    __extends(HireAnExpertLink, _super);
    function HireAnExpertLink() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HireAnExpertLink.prototype.render = function () {
        var _a = this.props, show = _a.show, t = _a.t, shouldShowNewArenaLink = _a.shouldShowNewArenaLink, shouldSupportMultiService = _a.shouldSupportMultiService;
        if (show) {
            return (React.createElement("span", { className: style.inlineBlockContainer },
                React.createElement(HeaderLink, { text: t('hire_an_expert_link'), biLinkName: "hire-an-expert", url: shouldShowNewArenaLink ? (shouldSupportMultiService ? ARENA_MULTI_URL : ARENA_URL) : deprecated_ARENA_URL, target: "_blank" })));
        }
        return null;
    };
    return HireAnExpertLink;
}(React.PureComponent));
var mapStateToProps = function (state) {
    return ({
        show: !showExpertExperience(state) && !isNtt() && isNotCoBranded(state),
        shouldShowNewArenaLink: experimentsSelectors(state).enabled('specs.wos2.newHireAProfessionalLink'),
        shouldSupportMultiService: experimentsSelectors(state).enabled('specs.SupportMultipleServicesBrief')
    });
};
export default connect(mapStateToProps)(withTranslation()(HireAnExpertLink));
