// factory.ts/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeFactory = exports.each = exports.val = exports.Factory = exports.Derived = exports.Generator = exports.Pipeline = exports.Sync = exports.Async = void 0;
const Async = require("./async");
exports.Async = Async;
const Sync = require("./sync");
exports.Sync = Sync;
const Pipeline = require("./pipeline");
exports.Pipeline = Pipeline;
// for now, for backwards compat
var sync_1 = require("./sync");
Object.defineProperty(exports, "Generator", { enumerable: true, get: function () { return sync_1.Generator; } });
Object.defineProperty(exports, "Derived", { enumerable: true, get: function () { return sync_1.Derived; } });
Object.defineProperty(exports, "Factory", { enumerable: true, get: function () { return sync_1.Factory; } });
Object.defineProperty(exports, "val", { enumerable: true, get: function () { return sync_1.val; } });
Object.defineProperty(exports, "each", { enumerable: true, get: function () { return sync_1.each; } });
Object.defineProperty(exports, "makeFactory", { enumerable: true, get: function () { return sync_1.makeFactory; } });
