// @wix/ambassador-payment-dashboard-web/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubscriptionStatus = exports.RefundStatus = exports.RefundOrigin = exports.RefundCapabilityReason = exports.ProductType = exports.PaymentType = exports.PaymentStatus = exports.Namespace = exports.Frequency = exports.EvidenceType = exports.DisputeStatus = exports.DisputeDisputeReason = exports.DisputeCapabilityReason = exports.DashboardSorting = exports.DashboardDisputeReason = exports.CaptureStatus = exports.BoSorting = exports.AuthorizationStatus = exports.AccountProfileStatus = exports.AcceptDisputeCapabilityReason = void 0;
var AcceptDisputeCapabilityReason;
(function (AcceptDisputeCapabilityReason) {
    AcceptDisputeCapabilityReason["UNKNOWN"] = "UNKNOWN";
    AcceptDisputeCapabilityReason["NOT_AUTHORIZED"] = "NOT_AUTHORIZED";
})(AcceptDisputeCapabilityReason = exports.AcceptDisputeCapabilityReason || (exports.AcceptDisputeCapabilityReason = {}));
var AccountProfileStatus;
(function (AccountProfileStatus) {
    AccountProfileStatus["UNKNOWN"] = "UNKNOWN";
    AccountProfileStatus["PENDING"] = "PENDING";
    AccountProfileStatus["ACCEPTING_PAYMENTS"] = "ACCEPTING_PAYMENTS";
    AccountProfileStatus["ACTIVE"] = "ACTIVE";
    AccountProfileStatus["SUSPENDED"] = "SUSPENDED";
    AccountProfileStatus["BLOCKED"] = "BLOCKED";
    AccountProfileStatus["DENIED"] = "DENIED";
    AccountProfileStatus["UNDER_REVIEW"] = "UNDER_REVIEW";
    AccountProfileStatus["NOT_ACTIVE"] = "NOT_ACTIVE";
})(AccountProfileStatus = exports.AccountProfileStatus || (exports.AccountProfileStatus = {}));
var AuthorizationStatus;
(function (AuthorizationStatus) {
    AuthorizationStatus["UNDEFINED"] = "UNDEFINED";
    AuthorizationStatus["PENDING"] = "PENDING";
    AuthorizationStatus["DECLINED"] = "DECLINED";
    AuthorizationStatus["SUCCEEDED"] = "SUCCEEDED";
})(AuthorizationStatus = exports.AuthorizationStatus || (exports.AuthorizationStatus = {}));
var BoSorting;
(function (BoSorting) {
    BoSorting["ASC"] = "ASC";
    BoSorting["DESC"] = "DESC";
})(BoSorting = exports.BoSorting || (exports.BoSorting = {}));
var CaptureStatus;
(function (CaptureStatus) {
    CaptureStatus["UNDEFINED"] = "UNDEFINED";
    CaptureStatus["PENDING"] = "PENDING";
    CaptureStatus["DECLINED"] = "DECLINED";
    CaptureStatus["SUCCEEDED"] = "SUCCEEDED";
})(CaptureStatus = exports.CaptureStatus || (exports.CaptureStatus = {}));
var DashboardDisputeReason;
(function (DashboardDisputeReason) {
    DashboardDisputeReason["OTHER"] = "OTHER";
    DashboardDisputeReason["FRAUD_CARD_PRESENT"] = "FRAUD_CARD_PRESENT";
    DashboardDisputeReason["FRAUD_CARD_ABSENT"] = "FRAUD_CARD_ABSENT";
    DashboardDisputeReason["DUPLICATE_PROCESSING"] = "DUPLICATE_PROCESSING";
    DashboardDisputeReason["SERVICE_NOT_PROVIDED"] = "SERVICE_NOT_PROVIDED";
    DashboardDisputeReason["CANCELLED_RECURRING"] = "CANCELLED_RECURRING";
    DashboardDisputeReason["NOT_AS_DESCRIBED"] = "NOT_AS_DESCRIBED";
    DashboardDisputeReason["COUNTERFEIT"] = "COUNTERFEIT";
    DashboardDisputeReason["MISREPRESENTATION"] = "MISREPRESENTATION";
    DashboardDisputeReason["CANCELLED"] = "CANCELLED";
})(DashboardDisputeReason = exports.DashboardDisputeReason || (exports.DashboardDisputeReason = {}));
var DashboardSorting;
(function (DashboardSorting) {
    DashboardSorting["ASC"] = "ASC";
    DashboardSorting["DESC"] = "DESC";
})(DashboardSorting = exports.DashboardSorting || (exports.DashboardSorting = {}));
var DisputeCapabilityReason;
(function (DisputeCapabilityReason) {
    DisputeCapabilityReason["UNKNOWN"] = "UNKNOWN";
    DisputeCapabilityReason["NOT_AUTHORIZED"] = "NOT_AUTHORIZED";
})(DisputeCapabilityReason = exports.DisputeCapabilityReason || (exports.DisputeCapabilityReason = {}));
var DisputeDisputeReason;
(function (DisputeDisputeReason) {
    DisputeDisputeReason["OTHER"] = "OTHER";
    DisputeDisputeReason["FRAUD_CARD_PRESENT"] = "FRAUD_CARD_PRESENT";
    DisputeDisputeReason["FRAUD_CARD_ABSENT"] = "FRAUD_CARD_ABSENT";
    DisputeDisputeReason["DUPLICATE_PROCESSING"] = "DUPLICATE_PROCESSING";
    DisputeDisputeReason["SERVICES_NOT_PROVIDED"] = "SERVICES_NOT_PROVIDED";
    DisputeDisputeReason["CANCELED_RECURRING"] = "CANCELED_RECURRING";
    DisputeDisputeReason["NOT_AS_DESCRIBED"] = "NOT_AS_DESCRIBED";
    DisputeDisputeReason["COUNTERFEIT"] = "COUNTERFEIT";
    DisputeDisputeReason["MISREPRESENTATION"] = "MISREPRESENTATION";
    DisputeDisputeReason["CANCELED"] = "CANCELED";
})(DisputeDisputeReason = exports.DisputeDisputeReason || (exports.DisputeDisputeReason = {}));
var DisputeStatus;
(function (DisputeStatus) {
    DisputeStatus["UNDEFINED"] = "UNDEFINED";
    DisputeStatus["OPEN"] = "OPEN";
    DisputeStatus["UNDER_REVIEW"] = "UNDER_REVIEW";
    DisputeStatus["WON"] = "WON";
    DisputeStatus["LOST"] = "LOST";
})(DisputeStatus = exports.DisputeStatus || (exports.DisputeStatus = {}));
var EvidenceType;
(function (EvidenceType) {
    EvidenceType["OTHER"] = "OTHER";
    EvidenceType["COMMUNICATION"] = "COMMUNICATION";
    EvidenceType["INVOICE"] = "INVOICE";
    EvidenceType["TERMS_AND_CONDITIONS"] = "TERMS_AND_CONDITIONS";
    EvidenceType["SEPARATE_ORDERS_PROOF"] = "SEPARATE_ORDERS_PROOF";
    EvidenceType["CONFIRMATION_EMAIL"] = "CONFIRMATION_EMAIL";
    EvidenceType["DELIVERY_PROOF"] = "DELIVERY_PROOF";
    EvidenceType["TRANSACTION_LOG"] = "TRANSACTION_LOG";
    EvidenceType["WEBSITE_OR_SERVICE_ACCESSED"] = "WEBSITE_OR_SERVICE_ACCESSED";
    EvidenceType["CERTIFICATE_OF_AUTHENTICITY"] = "CERTIFICATE_OF_AUTHENTICITY";
})(EvidenceType = exports.EvidenceType || (exports.EvidenceType = {}));
var Frequency;
(function (Frequency) {
    Frequency["UNDEFINED"] = "UNDEFINED";
    Frequency["DAY"] = "DAY";
    Frequency["WEEK"] = "WEEK";
    Frequency["MONTH"] = "MONTH";
    Frequency["YEAR"] = "YEAR";
})(Frequency = exports.Frequency || (exports.Frequency = {}));
var Namespace;
(function (Namespace) {
    Namespace["WIX"] = "WIX";
    Namespace["CLUBS"] = "CLUBS";
    Namespace["PARTNERS"] = "PARTNERS";
})(Namespace = exports.Namespace || (exports.Namespace = {}));
var PaymentStatus;
(function (PaymentStatus) {
    PaymentStatus["UNKNOWN"] = "UNKNOWN";
    PaymentStatus["SUCCEEDED"] = "SUCCEEDED";
    PaymentStatus["PENDING"] = "PENDING";
    PaymentStatus["DECLINED"] = "DECLINED";
    PaymentStatus["PENDING_REFUND"] = "PENDING_REFUND";
    PaymentStatus["PARTIALLY_REFUNDED"] = "PARTIALLY_REFUNDED";
    PaymentStatus["REFUNDED"] = "REFUNDED";
    PaymentStatus["DISPUTE_OPEN"] = "DISPUTE_OPEN";
    PaymentStatus["DISPUTE_UNDER_REVIEW"] = "DISPUTE_UNDER_REVIEW";
    PaymentStatus["DISPUTE_WON"] = "DISPUTE_WON";
    PaymentStatus["DISPUTE_LOST"] = "DISPUTE_LOST";
    PaymentStatus["CHARGEBACK"] = "CHARGEBACK";
    PaymentStatus["REFUND_FAILED"] = "REFUND_FAILED";
    PaymentStatus["PARTIAL_REFUND_FAILED"] = "PARTIAL_REFUND_FAILED";
})(PaymentStatus = exports.PaymentStatus || (exports.PaymentStatus = {}));
var PaymentType;
(function (PaymentType) {
    PaymentType["ONE_TIME"] = "ONE_TIME";
    PaymentType["RECURRING"] = "RECURRING";
})(PaymentType = exports.PaymentType || (exports.PaymentType = {}));
var ProductType;
(function (ProductType) {
    ProductType["OTHER"] = "OTHER";
    ProductType["MERCHANDISE"] = "MERCHANDISE";
    ProductType["DIGITAL_GOODS"] = "DIGITAL_GOODS";
    ProductType["SERVICE"] = "SERVICE";
    ProductType["EVENT"] = "EVENT";
})(ProductType = exports.ProductType || (exports.ProductType = {}));
var RefundCapabilityReason;
(function (RefundCapabilityReason) {
    RefundCapabilityReason["UNKNOWN"] = "UNKNOWN";
    RefundCapabilityReason["PROVIDER_NOT_SUPPORTED"] = "PROVIDER_NOT_SUPPORTED";
    RefundCapabilityReason["PENDING_REFUND"] = "PENDING_REFUND";
    RefundCapabilityReason["METHOD_NOT_SUPPORTED"] = "METHOD_NOT_SUPPORTED";
    RefundCapabilityReason["PAYMENT_CHARGEBACKED"] = "PAYMENT_CHARGEBACKED";
    RefundCapabilityReason["NOT_AUTHORIZED"] = "NOT_AUTHORIZED";
})(RefundCapabilityReason = exports.RefundCapabilityReason || (exports.RefundCapabilityReason = {}));
var RefundOrigin;
(function (RefundOrigin) {
    RefundOrigin["UNDEFINED"] = "UNDEFINED";
    RefundOrigin["INITIATED_BY_PROVIDER"] = "INITIATED_BY_PROVIDER";
    RefundOrigin["MERCHANT"] = "MERCHANT";
    RefundOrigin["RISK_TEAM"] = "RISK_TEAM";
})(RefundOrigin = exports.RefundOrigin || (exports.RefundOrigin = {}));
var RefundStatus;
(function (RefundStatus) {
    RefundStatus["UNDEFINED"] = "UNDEFINED";
    RefundStatus["PENDING"] = "PENDING";
    RefundStatus["DECLINED"] = "DECLINED";
    RefundStatus["SUCCEEDED"] = "SUCCEEDED";
    RefundStatus["REVERSED"] = "REVERSED";
})(RefundStatus = exports.RefundStatus || (exports.RefundStatus = {}));
var SubscriptionStatus;
(function (SubscriptionStatus) {
    SubscriptionStatus["SUBSCRIPTION_UNDEFINED"] = "SUBSCRIPTION_UNDEFINED";
    SubscriptionStatus["SUBSCRIPTION_ACTIVE"] = "SUBSCRIPTION_ACTIVE";
    SubscriptionStatus["SUBSCRIPTION_CANCELLED"] = "SUBSCRIPTION_CANCELLED";
    SubscriptionStatus["SUBSCRIPTION_FINISHED"] = "SUBSCRIPTION_FINISHED";
    SubscriptionStatus["SUBSCRIPTION_PENDING"] = "SUBSCRIPTION_PENDING";
    SubscriptionStatus["SUBSCRIPTION_SUSPENDED"] = "SUBSCRIPTION_SUSPENDED";
})(SubscriptionStatus = exports.SubscriptionStatus || (exports.SubscriptionStatus = {}));
