// @wix/wixstores-client-core/dist/src/dashboard/topology.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUBSCRIPTION_URL = exports.BULK_UNSET_FULFILLERS_FROM_PRODUCTS_BY_FILTER = exports.BULK_ASSIGN_FULFILLERS_TO_PRODUCTS_BY_FILTER = exports.ASSIGN_FULFILLER_TO_PRODUCT = exports.ADD_FULFILLMENT_SERVICE = exports.FULFILLMENTS_URL = exports.CALCULATE_REFUND = exports.PREVIEW_REFUND_EMAIL = exports.PREVIEW_CANCEL_EMAIL = exports.REFUND_ORDER = exports.CANCEL_ORDER = exports.EXPORT_ORDERS_URL = exports.UPDATE_FULFILLMENT_URL = exports.ADD_FULFILLMENT_URL = exports.BULK_MARK_ORDERS_AS_PAID_COMMAND = exports.ORDER_EMAIL_PREVIEW_URL = exports.ORDERS_COMMANDS_URL = exports.ABANDONED_CARTS_API_ENDPOINT = exports.SUBMIT_NEW_TAX_RULE_ACCOUNT_REQUEST = exports.DELETE_TAX_RULE_URL = exports.DELETE_TAX_GROUP_URL = exports.UPDATE_TAX_RULE_URL = exports.CREATE_TAX_RULE_URL = exports.UPDATE_TAX_GROUP_URL = exports.TAX_RULES_API_URL = exports.ENABLED_TAX_GROUPS = exports.SET_TAX_CALCULATOR = exports.CREATE_TAX_GROUP_URL = exports.EXECUTE_BULK_COMMANDS = exports.GRAPHQL_NODE_ENDPOINT = exports.STORE_SETTINGS_COMMANDS = exports.CATALOG_COMMANDS_URL = exports.GRAPHQL_ENDPOINT = exports.PROVISION_STORES = void 0;
// General
var SM_ROOT = "/_api/wix-ecommerce-renderer-web/store-manager/";
exports.PROVISION_STORES = SM_ROOT + "provision-store";
exports.GRAPHQL_ENDPOINT = "/_api/wix-ecommerce-graphql-web/api";
exports.CATALOG_COMMANDS_URL = SM_ROOT + "catalog/{commandName}";
exports.STORE_SETTINGS_COMMANDS = SM_ROOT + "store-settings/{commandName}";
exports.GRAPHQL_NODE_ENDPOINT = "/_api/wixstores-graphql-server/graphql";
exports.EXECUTE_BULK_COMMANDS = SM_ROOT + "{commandName}";
// Tax
exports.CREATE_TAX_GROUP_URL = SM_ROOT + "createTaxGroup";
exports.SET_TAX_CALCULATOR = SM_ROOT + "changeTaxCalculator";
exports.ENABLED_TAX_GROUPS = SM_ROOT + "enableTaxGroups";
exports.TAX_RULES_API_URL = SM_ROOT + "tax-rules";
exports.UPDATE_TAX_GROUP_URL = SM_ROOT + "updateTaxGroup";
exports.CREATE_TAX_RULE_URL = SM_ROOT + "createTaxRules";
exports.UPDATE_TAX_RULE_URL = SM_ROOT + "updateTaxRules";
exports.DELETE_TAX_GROUP_URL = SM_ROOT + "deleteTaxGroup";
exports.DELETE_TAX_RULE_URL = SM_ROOT + "deleteTaxRules";
exports.SUBMIT_NEW_TAX_RULE_ACCOUNT_REQUEST = SM_ROOT + "submitNewTaxAccountRequest";
// Abandoned Cart
exports.ABANDONED_CARTS_API_ENDPOINT = '_api/abandoned-carts/v1/abandonedCarts';
// Orders Manager
exports.ORDERS_COMMANDS_URL = SM_ROOT + "order/{commandName}";
exports.ORDER_EMAIL_PREVIEW_URL = SM_ROOT + "order-email-preview";
exports.BULK_MARK_ORDERS_AS_PAID_COMMAND = 'BulkMarkOrderAsPaidByStoreManager';
exports.ADD_FULFILLMENT_URL = SM_ROOT + "orders/createFulfillment";
exports.UPDATE_FULFILLMENT_URL = SM_ROOT + "orders/updateFulfillment";
exports.EXPORT_ORDERS_URL = 'https://ecom.wix.com/storemanager/orders-export';
exports.CANCEL_ORDER = SM_ROOT + "orders/{orderId}/cancel";
exports.REFUND_ORDER = SM_ROOT + "orders/{orderId}/refund";
exports.PREVIEW_CANCEL_EMAIL = SM_ROOT + "orders/{orderId}/preview-email/cancel";
exports.PREVIEW_REFUND_EMAIL = SM_ROOT + "orders/{orderId}/preview-email/refund";
exports.CALCULATE_REFUND = SM_ROOT + "orders/{orderId}/calculate-refund";
// Fulfillment Services
exports.FULFILLMENTS_URL = '/_api/wix-ecommerce-fulfillment/v1/';
exports.ADD_FULFILLMENT_SERVICE = exports.FULFILLMENTS_URL + "fulfillers/unified";
exports.ASSIGN_FULFILLER_TO_PRODUCT = exports.FULFILLMENTS_URL + "fulfillers/setProductsAssociations/byIds";
exports.BULK_ASSIGN_FULFILLERS_TO_PRODUCTS_BY_FILTER = exports.FULFILLMENTS_URL + "fulfillers/setProductsAssociations/byFilter";
exports.BULK_UNSET_FULFILLERS_FROM_PRODUCTS_BY_FILTER = exports.FULFILLMENTS_URL + "fulfillers/unsetProductsAssociations/byFilter";
// Purchase Plans
exports.SUBSCRIPTION_URL = '/_api/subscriptions-server/v1/subscriptions';
