// @wix/my-account-header/dist/es/src/client/components/MySitesDropdown/Footer/MySitesFooter.js
import * as React from 'react';
import { memo, useCallback } from 'react';
import { Box, TextButton } from 'wix-style-react';
import { MySitesDropdownDataHooks } from '../MySitesDropdown.data-hooks';
import * as s from '../MySitesDropdown.scss';
import { withTranslation } from '@wix/wix-i18n-config';
import { BiEvents } from '../../../constants/biEvents';
import { provideLogger } from '../../../services/BiLogger';
import { getCurrentSiteId } from '../../../selectors/currentSiteSelectors';
import { getBiPageView } from '../../../selectors/biPageViewSelectors';
import { connect } from 'react-redux';
import * as URI from 'urijs';
var allSitesLink = function (biPageView) {
    return new URI("/account/sites").addQuery({ referralAdditionalInfo: biPageView }).toString();
};
export var MySitesFooter = memo(function (_a) {
    var biLogger = _a.biLogger, currentSiteId = _a.currentSiteId, biPageView = _a.biPageView, t = _a.t;
    var goToAllSitesClicked = useCallback(function () {
        biLogger.logEvent(BiEvents.MY_SITES_DROPDOWN_ACTION, {
            msid: currentSiteId,
            page_view: biPageView,
            item_name: 'all-sites'
        });
    }, [biLogger, currentSiteId]);
    return (React.createElement(Box, { className: s.footer, height: 60, verticalAlign: "middle", align: "center" },
        React.createElement(TextButton, { size: "medium", weight: "normal", dataHook: MySitesDropdownDataHooks.GoToAllSites, as: "a", href: allSitesLink(biPageView), onClick: goToAllSitesClicked }, t('site_list.dropdown.go_to_all_sites'))));
});
var mapStateToProps = function (state) { return ({
    currentSiteId: getCurrentSiteId(state),
    biPageView: getBiPageView(state),
}); };
export default provideLogger(withTranslation()(connect(mapStateToProps)(MySitesFooter)));
