// @wix/quick-actions-dropdown/dist/es/src/services/actionsPermissions.js
import { userHasPermission } from './permissions';
import { PERMISSIONS } from '../consts/permissions';
var always = function () { return true; };
var never = function () { return false; };
var isOwner = function (quickActionsData) { return quickActionsData.isOwner; };
var htmlSiteAndNotEditorXAndHasGetFeedbackPermission = function (quickActionsData) {
    return quickActionsData.isHtmlSite && quickActionsData.editorType !== 'EDITORX' && userHasPermission(quickActionsData, PERMISSIONS.GET_FEEDBACK);
};
var htmlSiteAndNotEditorXAndHasGetFeedbackPermissionDeprecated = function (quickActionsData) {
    return quickActionsData.isHtmlSite && quickActionsData.editorType !== 'EDITORX' && userHasPermission(quickActionsData, PERMISSIONS.GET_FEEDBACK_DEPRECATED);
};
var htmlSiteAndHasGetFeedbackPermission = function (quickActionsData) {
    return quickActionsData.isHtmlSite && userHasPermission(quickActionsData, PERMISSIONS.GET_FEEDBACK);
};
var htmlSiteAndHasGetFeedbackPermissionDeprecated = function (quickActionsData) {
    return quickActionsData.isHtmlSite && userHasPermission(quickActionsData, PERMISSIONS.GET_FEEDBACK_DEPRECATED);
};
var htmlSiteAndNotInTransferAndHasTransferSitePermission = function (quickActionsData) {
    return quickActionsData.isHtmlSite && !quickActionsData.isInTransfer && userHasPermission(quickActionsData, PERMISSIONS.TRANSFER_SITE);
};
var isInTransferAndHasTransferSitePermission = function (quickActionsData) {
    return quickActionsData.isInTransfer && userHasPermission(quickActionsData, PERMISSIONS.TRANSFER_SITE);
};
var withPermissions = function (permission) {
    return function (quickActionsData) { return userHasPermission(quickActionsData, permission); };
};
var htmlSiteWithPermissions = function (permission) {
    return function (quickActionsData) { return quickActionsData.isHtmlSite && userHasPermission(quickActionsData, permission); };
};
var isSiteBlocked = function (quickActionsData) { return quickActionsData.blocked; };
export var PermissionMethods = {
    always: always,
    never: never,
    isOwner: isOwner,
    isSiteBlocked: isSiteBlocked,
    htmlSiteAndNotEditorXAndHasGetFeedbackPermission: htmlSiteAndNotEditorXAndHasGetFeedbackPermission,
    htmlSiteAndNotEditorXAndHasGetFeedbackPermissionDeprecated: htmlSiteAndNotEditorXAndHasGetFeedbackPermissionDeprecated,
    htmlSiteAndHasGetFeedbackPermission: htmlSiteAndHasGetFeedbackPermission,
    htmlSiteAndHasGetFeedbackPermissionDeprecated: htmlSiteAndHasGetFeedbackPermissionDeprecated,
    htmlSiteAndNotInTransferAndHasTransferSitePermission: htmlSiteAndNotInTransferAndHasTransferSitePermission,
    isInTransferAndHasTransferSitePermission: isInTransferAndHasTransferSitePermission,
    htmlSiteWithPermissions: htmlSiteWithPermissions,
    withPermissions: withPermissions
};
