// @wix/fedops-logger/dist/es/src/publishers/env.js
'use strict';

exports.__esModule = true;
exports.setHost = setHost;
exports.getHost = getHost;
exports.isWebWorker = isWebWorker;
exports.hasSendBeacon = hasSendBeacon;
exports.hasFetch = hasFetch;
exports.provideBrowsingContext = provideBrowsingContext;

var _env = require('../env');

var host = 'frog.wix.com';

function setHost(_host) {
  host = _host;
}

function getHost() {
  return host;
}

function isWebWorker() {
  /* global WorkerGlobalScope */

  //At the moment, this is the most reliable way of detecting if current code is run in a worker
  //Taken from here: https://stackoverflow.com/a/18002694
  return typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope;
}

function hasSendBeacon() {
  return typeof navigator !== 'undefined' && navigator && navigator.sendBeacon;
}

function hasFetch() {
  return !!self.fetch;
}

/**
 * Provides a window object to the callback iff current window is the topmost one
 * Otherwise returns null
 */
function provideBrowsingContext(cb) {
  if (typeof (0, _env.env)() !== 'undefined' && (0, _env.env)()) {
    try {
      return (0, _env.env)().top === (0, _env.env)().self ? cb ? cb((0, _env.env)()) : (0, _env.env)() : null;
    } catch (e) {
      return null;
    }
  }

  return null;
}
