// @wix/ambassador-analytics-ng-server/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailboxSaleType = exports.LimitationSourceType = exports.Label = exports.JSType = exports.InvoiceStatus = exports.IntervalUnit = exports.InsightsCategories = exports.InitialPurchaseParamsParameters = exports.IdempotencyLevel = exports.HttpStatusCode = exports.Format = exports.FieldDescriptorProtoType = exports.FeatureUnits = exports.FeatureType = exports.FeatureStatus = exports.FeaturePeriod = exports.FeatureContextEnumFeatureContext = exports.Exposure = exports.EventType = exports.EventName = exports.EntityType = exports.DownloadFormat = exports.DirectDebitBrand = exports.DeviceType = exports.DashboardMode = exports.CurrentPeriod = exports.CreditCardType = exports.ContractSwitchTypeEnumContractSwitchType = exports.ComparisonPeriod = exports.CompareType = exports.CommonNotificationParamsParameters = exports.ChargebackParamsParameters = exports.ChargeFailureParamsParameters = exports.CashType = exports.CardBrand = exports.CancelImmediateByUserParamsParameters = exports.CancelDueToPaymentProblemsParamsParameters = exports.CacheControl = exports.CType = exports.BulkActionType = exports.BillingCycle = exports.BillingAccountType = exports.BenchmarksMode = exports.BenchmarkTipType = exports.BankTransferType = exports.AutoRenewOffParamsParameters = exports.AttributeType = exports.AttachmentFormat = exports.AlertType = exports.AggregationType = void 0;
exports.TransferServiceParamsParameters = exports.ThreeDsProviderEnum = exports.ThreeDsChallengeDisplaySizeEnum = exports.TaxTypeConfigurationTaxType = exports.TaxSettingsTaxType = exports.TaxIdFieldMode = exports.SwitchContractParamsParameters = exports.SubscriptionUpgradeType = exports.SubdivisionType = exports.StreamViewType = exports.SpeedState = exports.SortType = exports.SortOrder = exports.SortDirection = exports.SettingsPaymentMethod = exports.ServiceStatus = exports.ServiceAppendixType = exports.SegmentSegment = exports.SearchDetailsMode = exports.ScalarType = exports.RuleType = exports.RefundReason = exports.RecurringTypeMessageRecurringType = exports.RecurringChargeSuccessParamsParameters = exports.QuantityType = exports.PurchaseScenario = exports.ProductTypeContext = exports.ProductStatus = exports.PredefinedTimeframe = exports.PermissionType = exports.PeriodType = exports.PeriodEnumPeriod = exports.PaymentTypeRecurringType = exports.PaymentMethodSubType = exports.PaymentFiltersPaymentType = exports.PagingType = exports.OrderSessionItemContractSwitchType = exports.OptimizeMode = exports.Operator = exports.OfflinePaymentParamsParameters = exports.NullValue = exports.NotificationTypes = exports.NoActiveBillingAccountParamsParameters = exports.NearRenewalUpgradeCycleParamsParameters = exports.NearRenewalParamsParameters = exports.NearExpirationParamsParameters = exports.Namespace = exports.Method = exports.Maturity = exports.ManualRecurringReminderParamsParameters = void 0;
exports.VatType = exports.ValidateNoPreviousPurchaseType = exports.V2OrderSessionPaymentMethodType = exports.V1OrderSessionPaymentMethodType = exports.V1Mode = exports.V1ContractSwitchType = exports.TranslatableType = void 0;
var AggregationType;
(function (AggregationType) {
    AggregationType["UNKNOWN_AGGREGATION_TYPE"] = "UNKNOWN_AGGREGATION_TYPE";
    AggregationType["VALUE"] = "VALUE";
    AggregationType["RANGE"] = "RANGE";
    AggregationType["SCALAR"] = "SCALAR";
})(AggregationType = exports.AggregationType || (exports.AggregationType = {}));
var AlertType;
(function (AlertType) {
    AlertType["UNDEFINED"] = "UNDEFINED";
    AlertType["ANOMALY_DETECTION"] = "ANOMALY_DETECTION";
    AlertType["NUMBER_HIGHER"] = "NUMBER_HIGHER";
    AlertType["NUMBER_LOWER"] = "NUMBER_LOWER";
    AlertType["PERCENTAGE_INCREASE"] = "PERCENTAGE_INCREASE";
    AlertType["PERCENTAGE_DECREASE"] = "PERCENTAGE_DECREASE";
})(AlertType = exports.AlertType || (exports.AlertType = {}));
var AttachmentFormat;
(function (AttachmentFormat) {
    AttachmentFormat["undefined"] = "undefined";
    AttachmentFormat["csv"] = "csv";
    AttachmentFormat["xlsx"] = "xlsx";
})(AttachmentFormat = exports.AttachmentFormat || (exports.AttachmentFormat = {}));
var AttributeType;
(function (AttributeType) {
    AttributeType["STRING"] = "STRING";
    AttributeType["NUMBER"] = "NUMBER";
    AttributeType["BOOLEAN"] = "BOOLEAN";
})(AttributeType = exports.AttributeType || (exports.AttributeType = {}));
var AutoRenewOffParamsParameters;
(function (AutoRenewOffParamsParameters) {
    AutoRenewOffParamsParameters["UNUSED"] = "UNUSED";
    AutoRenewOffParamsParameters["NEXT_BILLING_DATE"] = "NEXT_BILLING_DATE";
})(AutoRenewOffParamsParameters = exports.AutoRenewOffParamsParameters || (exports.AutoRenewOffParamsParameters = {}));
var BankTransferType;
(function (BankTransferType) {
    BankTransferType["UNKNOWN"] = "UNKNOWN";
    BankTransferType["SOFORT"] = "SOFORT";
})(BankTransferType = exports.BankTransferType || (exports.BankTransferType = {}));
var BenchmarkTipType;
(function (BenchmarkTipType) {
    BenchmarkTipType["TRAFFIC"] = "TRAFFIC";
    BenchmarkTipType["SALES"] = "SALES";
})(BenchmarkTipType = exports.BenchmarkTipType || (exports.BenchmarkTipType = {}));
var BenchmarksMode;
(function (BenchmarksMode) {
    BenchmarksMode["notEmpty"] = "notEmpty";
    BenchmarksMode["emptyNoTraffic"] = "emptyNoTraffic";
    BenchmarksMode["emptyNoIndustry"] = "emptyNoIndustry";
})(BenchmarksMode = exports.BenchmarksMode || (exports.BenchmarksMode = {}));
var BillingAccountType;
(function (BillingAccountType) {
    BillingAccountType["CREDIT_CARD"] = "CREDIT_CARD";
    BillingAccountType["E_CHECK"] = "E_CHECK";
    BillingAccountType["BANK_DRAFT_ACCOUNT"] = "BANK_DRAFT_ACCOUNT";
    BillingAccountType["MANUAL_PAYMENT_ACCOUNT"] = "MANUAL_PAYMENT_ACCOUNT";
    BillingAccountType["PAY_PAL"] = "PAY_PAL";
    BillingAccountType["WIRE_TRANSFER_ACCOUNT"] = "WIRE_TRANSFER_ACCOUNT";
    BillingAccountType["WALLET_YANDEX"] = "WALLET_YANDEX";
    BillingAccountType["MERCADO_PAGO_RECURRING"] = "MERCADO_PAGO_RECURRING";
    BillingAccountType["CREDITS_PREPAID_CARD"] = "CREDITS_PREPAID_CARD";
    BillingAccountType["REG_DOT_RU"] = "REG_DOT_RU";
    BillingAccountType["SEPA_DIRECT_DEBIT"] = "SEPA_DIRECT_DEBIT";
    BillingAccountType["VIRTUAL_MONEY"] = "VIRTUAL_MONEY";
    BillingAccountType["ACCOUNT_CREDIT"] = "ACCOUNT_CREDIT";
    BillingAccountType["CASH"] = "CASH";
    BillingAccountType["BANK_TRANSFER"] = "BANK_TRANSFER";
    BillingAccountType["LOCAL_CREDIT_DEBIT_CARD"] = "LOCAL_CREDIT_DEBIT_CARD";
})(BillingAccountType = exports.BillingAccountType || (exports.BillingAccountType = {}));
var BillingCycle;
(function (BillingCycle) {
    BillingCycle["ONE_TIME"] = "ONE_TIME";
    BillingCycle["DAILY"] = "DAILY";
    BillingCycle["WEEKLY"] = "WEEKLY";
    BillingCycle["MONTHLY"] = "MONTHLY";
    BillingCycle["QUARTERLY"] = "QUARTERLY";
    BillingCycle["SEMI_ANNUALLY"] = "SEMI_ANNUALLY";
    BillingCycle["YEARLY"] = "YEARLY";
    BillingCycle["TWO_YEARS"] = "TWO_YEARS";
    BillingCycle["THREE_YEARS"] = "THREE_YEARS";
    BillingCycle["FOUR_YEARS"] = "FOUR_YEARS";
    BillingCycle["FIVE_YEARS"] = "FIVE_YEARS";
})(BillingCycle = exports.BillingCycle || (exports.BillingCycle = {}));
var BulkActionType;
(function (BulkActionType) {
    BulkActionType["UNKNOWN_ACTION_TYPE"] = "UNKNOWN_ACTION_TYPE";
    BulkActionType["INSERT"] = "INSERT";
    BulkActionType["UPDATE"] = "UPDATE";
    BulkActionType["DELETE"] = "DELETE";
})(BulkActionType = exports.BulkActionType || (exports.BulkActionType = {}));
var CType;
(function (CType) {
    CType["STRING"] = "STRING";
    CType["CORD"] = "CORD";
    CType["STRING_PIECE"] = "STRING_PIECE";
})(CType = exports.CType || (exports.CType = {}));
var CacheControl;
(function (CacheControl) {
    CacheControl["UNDEFINED"] = "UNDEFINED";
    CacheControl["INVALIDATE"] = "INVALIDATE";
    CacheControl["BYPASS"] = "BYPASS";
})(CacheControl = exports.CacheControl || (exports.CacheControl = {}));
var CancelDueToPaymentProblemsParamsParameters;
(function (CancelDueToPaymentProblemsParamsParameters) {
    CancelDueToPaymentProblemsParamsParameters["UNUSED"] = "UNUSED";
})(CancelDueToPaymentProblemsParamsParameters = exports.CancelDueToPaymentProblemsParamsParameters || (exports.CancelDueToPaymentProblemsParamsParameters = {}));
var CancelImmediateByUserParamsParameters;
(function (CancelImmediateByUserParamsParameters) {
    CancelImmediateByUserParamsParameters["UNUSED"] = "UNUSED";
})(CancelImmediateByUserParamsParameters = exports.CancelImmediateByUserParamsParameters || (exports.CancelImmediateByUserParamsParameters = {}));
var CardBrand;
(function (CardBrand) {
    CardBrand["UNKNOWN"] = "UNKNOWN";
    CardBrand["VISA"] = "VISA";
    CardBrand["MASTERCARD"] = "MASTERCARD";
    CardBrand["AMEX"] = "AMEX";
    CardBrand["DISCOVER"] = "DISCOVER";
    CardBrand["JCB"] = "JCB";
    CardBrand["DINERS"] = "DINERS";
    CardBrand["CARTE_BLUE"] = "CARTE_BLUE";
    CardBrand["ELO"] = "ELO";
    CardBrand["UNION_PAY"] = "UNION_PAY";
})(CardBrand = exports.CardBrand || (exports.CardBrand = {}));
var CashType;
(function (CashType) {
    CashType["UNKNOWN"] = "UNKNOWN";
    CashType["BOLETO"] = "BOLETO";
})(CashType = exports.CashType || (exports.CashType = {}));
var ChargeFailureParamsParameters;
(function (ChargeFailureParamsParameters) {
    ChargeFailureParamsParameters["UNUSED"] = "UNUSED";
    ChargeFailureParamsParameters["PACKAGE_ID"] = "PACKAGE_ID";
    ChargeFailureParamsParameters["SUBSCRIPTION_TYPE"] = "SUBSCRIPTION_TYPE";
    ChargeFailureParamsParameters["NEXT_BILLING_DATE"] = "NEXT_BILLING_DATE";
    ChargeFailureParamsParameters["AMOUNT"] = "AMOUNT";
    ChargeFailureParamsParameters["CURRENCY_SYMBOL"] = "CURRENCY_SYMBOL";
})(ChargeFailureParamsParameters = exports.ChargeFailureParamsParameters || (exports.ChargeFailureParamsParameters = {}));
var ChargebackParamsParameters;
(function (ChargebackParamsParameters) {
    ChargebackParamsParameters["UNUSED"] = "UNUSED";
})(ChargebackParamsParameters = exports.ChargebackParamsParameters || (exports.ChargebackParamsParameters = {}));
var CommonNotificationParamsParameters;
(function (CommonNotificationParamsParameters) {
    CommonNotificationParamsParameters["UNUSED"] = "UNUSED";
    CommonNotificationParamsParameters["PRODUCT_NAME"] = "PRODUCT_NAME";
    CommonNotificationParamsParameters["SITE_NAME"] = "SITE_NAME";
    CommonNotificationParamsParameters["DOMAIN_NAME"] = "DOMAIN_NAME";
    CommonNotificationParamsParameters["PRODUCT_TYPE_NAME"] = "PRODUCT_TYPE_NAME";
})(CommonNotificationParamsParameters = exports.CommonNotificationParamsParameters || (exports.CommonNotificationParamsParameters = {}));
var CompareType;
(function (CompareType) {
    CompareType["UNDEFINED"] = "UNDEFINED";
    CompareType["PREVIOUS_PERIOD"] = "PREVIOUS_PERIOD";
    CompareType["PREVIOUS_YEAR"] = "PREVIOUS_YEAR";
})(CompareType = exports.CompareType || (exports.CompareType = {}));
var ComparisonPeriod;
(function (ComparisonPeriod) {
    ComparisonPeriod["PREVIOUS_PERIOD"] = "PREVIOUS_PERIOD";
    ComparisonPeriod["SELECTED_PERIOD"] = "SELECTED_PERIOD";
    ComparisonPeriod["LAST_YEAR"] = "LAST_YEAR";
})(ComparisonPeriod = exports.ComparisonPeriod || (exports.ComparisonPeriod = {}));
var ContractSwitchTypeEnumContractSwitchType;
(function (ContractSwitchTypeEnumContractSwitchType) {
    ContractSwitchTypeEnumContractSwitchType["UNKNOWN"] = "UNKNOWN";
    ContractSwitchTypeEnumContractSwitchType["CREDIT_UNUSED_PERIOD"] = "CREDIT_UNUSED_PERIOD";
    ContractSwitchTypeEnumContractSwitchType["EXTEND"] = "EXTEND";
    ContractSwitchTypeEnumContractSwitchType["FULL_AMOUNT_PERIOD"] = "FULL_AMOUNT_PERIOD";
})(ContractSwitchTypeEnumContractSwitchType = exports.ContractSwitchTypeEnumContractSwitchType || (exports.ContractSwitchTypeEnumContractSwitchType = {}));
var CreditCardType;
(function (CreditCardType) {
    CreditCardType["Visa"] = "Visa";
    CreditCardType["MasterCard"] = "MasterCard";
    CreditCardType["AMEX"] = "AMEX";
    CreditCardType["Discover"] = "Discover";
    CreditCardType["JCB"] = "JCB";
    CreditCardType["ENROUTE"] = "ENROUTE";
    CreditCardType["Diners"] = "Diners";
    CreditCardType["SOLO"] = "SOLO";
    CreditCardType["MASTERO"] = "MASTERO";
    CreditCardType["VISA_DELTA"] = "VISA_DELTA";
    CreditCardType["BANCARD"] = "BANCARD";
    CreditCardType["ELECTRON"] = "ELECTRON";
    CreditCardType["MAESTR_UK"] = "MAESTR_UK";
    CreditCardType["CARTE_BLUE"] = "CARTE_BLUE";
    CreditCardType["ELO"] = "ELO";
    CreditCardType["SBI_MAESTRO"] = "SBI_MAESTRO";
    CreditCardType["UnionPay"] = "UnionPay";
})(CreditCardType = exports.CreditCardType || (exports.CreditCardType = {}));
var CurrentPeriod;
(function (CurrentPeriod) {
    CurrentPeriod["CUSTOM"] = "CUSTOM";
    CurrentPeriod["TODAY"] = "TODAY";
    CurrentPeriod["YESTERDAY"] = "YESTERDAY";
    CurrentPeriod["LAST_7_DAYS"] = "LAST_7_DAYS";
    CurrentPeriod["LAST_14_DAYS"] = "LAST_14_DAYS";
    CurrentPeriod["LAST_30_DAYS"] = "LAST_30_DAYS";
    CurrentPeriod["LAST_90_DAYS"] = "LAST_90_DAYS";
    CurrentPeriod["LAST_28_DAYS"] = "LAST_28_DAYS";
    CurrentPeriod["LAST_180_DAYS"] = "LAST_180_DAYS";
    CurrentPeriod["LAST_365_DAYS"] = "LAST_365_DAYS";
    CurrentPeriod["THIS_WEEK"] = "THIS_WEEK";
    CurrentPeriod["THIS_MONTH"] = "THIS_MONTH";
    CurrentPeriod["THIS_QUATER"] = "THIS_QUATER";
    CurrentPeriod["THIS_YEAR"] = "THIS_YEAR";
    CurrentPeriod["LAST_WEEK"] = "LAST_WEEK";
    CurrentPeriod["LAST_MONTH"] = "LAST_MONTH";
    CurrentPeriod["LAST_QUATER"] = "LAST_QUATER";
    CurrentPeriod["LAST_YEAR"] = "LAST_YEAR";
})(CurrentPeriod = exports.CurrentPeriod || (exports.CurrentPeriod = {}));
var DashboardMode;
(function (DashboardMode) {
    DashboardMode["desktop"] = "desktop";
    DashboardMode["mobile"] = "mobile";
    DashboardMode["advanced"] = "advanced";
})(DashboardMode = exports.DashboardMode || (exports.DashboardMode = {}));
var DeviceType;
(function (DeviceType) {
    DeviceType["ALL"] = "ALL";
    DeviceType["MOBILE"] = "MOBILE";
    DeviceType["COMPUTER"] = "COMPUTER";
})(DeviceType = exports.DeviceType || (exports.DeviceType = {}));
var DirectDebitBrand;
(function (DirectDebitBrand) {
    DirectDebitBrand["UNKNOWN"] = "UNKNOWN";
    DirectDebitBrand["SEPA"] = "SEPA";
})(DirectDebitBrand = exports.DirectDebitBrand || (exports.DirectDebitBrand = {}));
var DownloadFormat;
(function (DownloadFormat) {
    DownloadFormat["CSV"] = "CSV";
    DownloadFormat["JSON"] = "JSON";
})(DownloadFormat = exports.DownloadFormat || (exports.DownloadFormat = {}));
var EntityType;
(function (EntityType) {
    EntityType["NOTHING"] = "NOTHING";
    EntityType["PRODUCT"] = "PRODUCT";
    EntityType["FEATURE"] = "FEATURE";
    EntityType["PRODUCT_TYPE"] = "PRODUCT_TYPE";
    EntityType["PRODUCT_FAMILY"] = "PRODUCT_FAMILY";
})(EntityType = exports.EntityType || (exports.EntityType = {}));
var EventName;
(function (EventName) {
    EventName["INSERT"] = "INSERT";
    EventName["MODIFY"] = "MODIFY";
    EventName["REMOVE"] = "REMOVE";
})(EventName = exports.EventName || (exports.EventName = {}));
var EventType;
(function (EventType) {
    EventType["UNKNOWN"] = "UNKNOWN";
    EventType["CREATED"] = "CREATED";
    EventType["UPDATED"] = "UPDATED";
    EventType["DELETED"] = "DELETED";
    EventType["ACTION"] = "ACTION";
})(EventType = exports.EventType || (exports.EventType = {}));
var Exposure;
(function (Exposure) {
    Exposure["PRIVATE"] = "PRIVATE";
    Exposure["INTERNAL"] = "INTERNAL";
    Exposure["PUBLIC"] = "PUBLIC";
})(Exposure = exports.Exposure || (exports.Exposure = {}));
var FeatureContextEnumFeatureContext;
(function (FeatureContextEnumFeatureContext) {
    FeatureContextEnumFeatureContext["UNKNOWN"] = "UNKNOWN";
    FeatureContextEnumFeatureContext["ACCOUNT"] = "ACCOUNT";
    FeatureContextEnumFeatureContext["SITE"] = "SITE";
    FeatureContextEnumFeatureContext["EXTERNAL_MANAGEMENT"] = "EXTERNAL_MANAGEMENT";
})(FeatureContextEnumFeatureContext = exports.FeatureContextEnumFeatureContext || (exports.FeatureContextEnumFeatureContext = {}));
var FeaturePeriod;
(function (FeaturePeriod) {
    FeaturePeriod["NO_PERIOD"] = "NO_PERIOD";
    FeaturePeriod["MILLISECOND"] = "MILLISECOND";
    FeaturePeriod["SECOND"] = "SECOND";
    FeaturePeriod["MINUTE"] = "MINUTE";
    FeaturePeriod["HOUR"] = "HOUR";
    FeaturePeriod["DAY"] = "DAY";
    FeaturePeriod["WEEK"] = "WEEK";
    FeaturePeriod["MONTH"] = "MONTH";
    FeaturePeriod["YEAR"] = "YEAR";
})(FeaturePeriod = exports.FeaturePeriod || (exports.FeaturePeriod = {}));
var FeatureStatus;
(function (FeatureStatus) {
    FeatureStatus["UNKNOWN"] = "UNKNOWN";
    FeatureStatus["DRAFT"] = "DRAFT";
    FeatureStatus["PUBLISHED"] = "PUBLISHED";
    FeatureStatus["ARCHIVED"] = "ARCHIVED";
})(FeatureStatus = exports.FeatureStatus || (exports.FeatureStatus = {}));
var FeatureType;
(function (FeatureType) {
    FeatureType["UNKNOWN"] = "UNKNOWN";
    FeatureType["GENERIC"] = "GENERIC";
    FeatureType["DOMAIN"] = "DOMAIN";
})(FeatureType = exports.FeatureType || (exports.FeatureType = {}));
var FeatureUnits;
(function (FeatureUnits) {
    FeatureUnits["UNITLESS"] = "UNITLESS";
    FeatureUnits["MINUTE"] = "MINUTE";
    FeatureUnits["HOUR"] = "HOUR";
    FeatureUnits["KB"] = "KB";
    FeatureUnits["MB"] = "MB";
    FeatureUnits["GB"] = "GB";
    FeatureUnits["SECOND"] = "SECOND";
})(FeatureUnits = exports.FeatureUnits || (exports.FeatureUnits = {}));
var FieldDescriptorProtoType;
(function (FieldDescriptorProtoType) {
    FieldDescriptorProtoType["TYPE_DOUBLE"] = "TYPE_DOUBLE";
    FieldDescriptorProtoType["TYPE_FLOAT"] = "TYPE_FLOAT";
    FieldDescriptorProtoType["TYPE_INT64"] = "TYPE_INT64";
    FieldDescriptorProtoType["TYPE_UINT64"] = "TYPE_UINT64";
    FieldDescriptorProtoType["TYPE_INT32"] = "TYPE_INT32";
    FieldDescriptorProtoType["TYPE_FIXED64"] = "TYPE_FIXED64";
    FieldDescriptorProtoType["TYPE_FIXED32"] = "TYPE_FIXED32";
    FieldDescriptorProtoType["TYPE_BOOL"] = "TYPE_BOOL";
    FieldDescriptorProtoType["TYPE_STRING"] = "TYPE_STRING";
    FieldDescriptorProtoType["TYPE_GROUP"] = "TYPE_GROUP";
    FieldDescriptorProtoType["TYPE_MESSAGE"] = "TYPE_MESSAGE";
    FieldDescriptorProtoType["TYPE_BYTES"] = "TYPE_BYTES";
    FieldDescriptorProtoType["TYPE_UINT32"] = "TYPE_UINT32";
    FieldDescriptorProtoType["TYPE_ENUM"] = "TYPE_ENUM";
    FieldDescriptorProtoType["TYPE_SFIXED32"] = "TYPE_SFIXED32";
    FieldDescriptorProtoType["TYPE_SFIXED64"] = "TYPE_SFIXED64";
    FieldDescriptorProtoType["TYPE_SINT32"] = "TYPE_SINT32";
    FieldDescriptorProtoType["TYPE_SINT64"] = "TYPE_SINT64";
})(FieldDescriptorProtoType = exports.FieldDescriptorProtoType || (exports.FieldDescriptorProtoType = {}));
var Format;
(function (Format) {
    Format["EMAIL"] = "EMAIL";
    Format["HOSTNAME"] = "HOSTNAME";
    Format["IPV4"] = "IPV4";
    Format["IPV6"] = "IPV6";
    Format["URI"] = "URI";
    Format["WEB_URL"] = "WEB_URL";
    Format["PHONE"] = "PHONE";
    Format["CREDIT_CARD"] = "CREDIT_CARD";
    Format["GUID"] = "GUID";
    Format["COUNTRY"] = "COUNTRY";
    Format["LANGUAGE"] = "LANGUAGE";
    Format["DECIMAL_VALUE"] = "DECIMAL_VALUE";
    Format["CURRENCY"] = "CURRENCY";
    Format["LANGUAGE_TAG"] = "LANGUAGE_TAG";
    Format["COLOR_HEX"] = "COLOR_HEX";
    Format["LOCAL_DATE"] = "LOCAL_DATE";
    Format["LOCAL_TIME"] = "LOCAL_TIME";
    Format["LOCAL_DATE_TIME"] = "LOCAL_DATE_TIME";
})(Format = exports.Format || (exports.Format = {}));
var HttpStatusCode;
(function (HttpStatusCode) {
    HttpStatusCode["OK"] = "OK";
    HttpStatusCode["UNAUTHENTICATED"] = "UNAUTHENTICATED";
    HttpStatusCode["PERMISSION_DENIED"] = "PERMISSION_DENIED";
    HttpStatusCode["INVALID_ARGUMENT"] = "INVALID_ARGUMENT";
    HttpStatusCode["NOT_FOUND"] = "NOT_FOUND";
    HttpStatusCode["INTERNAL"] = "INTERNAL";
    HttpStatusCode["UNAVAILABLE"] = "UNAVAILABLE";
    HttpStatusCode["RESOURCE_EXHAUSTED"] = "RESOURCE_EXHAUSTED";
    HttpStatusCode["CANCELED"] = "CANCELED";
    HttpStatusCode["ALREADY_EXISTS"] = "ALREADY_EXISTS";
    HttpStatusCode["FAILED_PRECONDITION"] = "FAILED_PRECONDITION";
})(HttpStatusCode = exports.HttpStatusCode || (exports.HttpStatusCode = {}));
var IdempotencyLevel;
(function (IdempotencyLevel) {
    IdempotencyLevel["IDEMPOTENCY_UNKNOWN"] = "IDEMPOTENCY_UNKNOWN";
    IdempotencyLevel["NO_SIDE_EFFECTS"] = "NO_SIDE_EFFECTS";
    IdempotencyLevel["IDEMPOTENT"] = "IDEMPOTENT";
})(IdempotencyLevel = exports.IdempotencyLevel || (exports.IdempotencyLevel = {}));
var InitialPurchaseParamsParameters;
(function (InitialPurchaseParamsParameters) {
    InitialPurchaseParamsParameters["UNUSED"] = "UNUSED";
    InitialPurchaseParamsParameters["INVOICE_NUM"] = "INVOICE_NUM";
    InitialPurchaseParamsParameters["BILLINGVENDOR"] = "BILLINGVENDOR";
    InitialPurchaseParamsParameters["PAYMENT_CYCLE"] = "PAYMENT_CYCLE";
    InitialPurchaseParamsParameters["CYCLE_ID"] = "CYCLE_ID";
    InitialPurchaseParamsParameters["SBS_PRODUCT_ID"] = "SBS_PRODUCT_ID";
    InitialPurchaseParamsParameters["NEXT_RENEWAL_AMOUNT"] = "NEXT_RENEWAL_AMOUNT";
    InitialPurchaseParamsParameters["NEXT_RENEWAL_CURRENCY"] = "NEXT_RENEWAL_CURRENCY";
    InitialPurchaseParamsParameters["PAYMENT_TYPE"] = "PAYMENT_TYPE";
})(InitialPurchaseParamsParameters = exports.InitialPurchaseParamsParameters || (exports.InitialPurchaseParamsParameters = {}));
var InsightsCategories;
(function (InsightsCategories) {
    InsightsCategories["TRAFFIC"] = "TRAFFIC";
    InsightsCategories["SALES"] = "SALES";
})(InsightsCategories = exports.InsightsCategories || (exports.InsightsCategories = {}));
var IntervalUnit;
(function (IntervalUnit) {
    IntervalUnit["UNKNOWN"] = "UNKNOWN";
    IntervalUnit["DAY"] = "DAY";
    IntervalUnit["WEEK"] = "WEEK";
    IntervalUnit["MONTH"] = "MONTH";
    IntervalUnit["YEAR"] = "YEAR";
})(IntervalUnit = exports.IntervalUnit || (exports.IntervalUnit = {}));
var InvoiceStatus;
(function (InvoiceStatus) {
    InvoiceStatus["UNKNOWN"] = "UNKNOWN";
    InvoiceStatus["DRAFT"] = "DRAFT";
    InvoiceStatus["PAID"] = "PAID";
    InvoiceStatus["VOID"] = "VOID";
    InvoiceStatus["REFUNDED"] = "REFUNDED";
    InvoiceStatus["DECLINED"] = "DECLINED";
    InvoiceStatus["PAYMENT_IN_PROGRESS"] = "PAYMENT_IN_PROGRESS";
})(InvoiceStatus = exports.InvoiceStatus || (exports.InvoiceStatus = {}));
var JSType;
(function (JSType) {
    JSType["JS_NORMAL"] = "JS_NORMAL";
    JSType["JS_STRING"] = "JS_STRING";
    JSType["JS_NUMBER"] = "JS_NUMBER";
})(JSType = exports.JSType || (exports.JSType = {}));
var Label;
(function (Label) {
    Label["LABEL_OPTIONAL"] = "LABEL_OPTIONAL";
    Label["LABEL_REQUIRED"] = "LABEL_REQUIRED";
    Label["LABEL_REPEATED"] = "LABEL_REPEATED";
})(Label = exports.Label || (exports.Label = {}));
var LimitationSourceType;
(function (LimitationSourceType) {
    LimitationSourceType["UNKNOWN"] = "UNKNOWN";
    LimitationSourceType["Default_Quota_Limit"] = "Default_Quota_Limit";
    LimitationSourceType["Feature_Quota_limit"] = "Feature_Quota_limit";
})(LimitationSourceType = exports.LimitationSourceType || (exports.LimitationSourceType = {}));
var MailboxSaleType;
(function (MailboxSaleType) {
    MailboxSaleType["UNKNOWN_SALE_TYPE"] = "UNKNOWN_SALE_TYPE";
    MailboxSaleType["NO_SALE"] = "NO_SALE";
    MailboxSaleType["SEASONAL_SALE"] = "SEASONAL_SALE";
    MailboxSaleType["PERSONAL_SALE"] = "PERSONAL_SALE";
})(MailboxSaleType = exports.MailboxSaleType || (exports.MailboxSaleType = {}));
var ManualRecurringReminderParamsParameters;
(function (ManualRecurringReminderParamsParameters) {
    ManualRecurringReminderParamsParameters["UNUSED"] = "UNUSED";
    ManualRecurringReminderParamsParameters["PACKAGE_ID"] = "PACKAGE_ID";
    ManualRecurringReminderParamsParameters["EXPIRATION_DATE"] = "EXPIRATION_DATE";
})(ManualRecurringReminderParamsParameters = exports.ManualRecurringReminderParamsParameters || (exports.ManualRecurringReminderParamsParameters = {}));
var Maturity;
(function (Maturity) {
    Maturity["ALPHA"] = "ALPHA";
    Maturity["BETA"] = "BETA";
    Maturity["IA"] = "IA";
    Maturity["GA"] = "GA";
    Maturity["NOT_IMPLEMENTED"] = "NOT_IMPLEMENTED";
})(Maturity = exports.Maturity || (exports.Maturity = {}));
var Method;
(function (Method) {
    Method["UNKNOWN_METHOD"] = "UNKNOWN_METHOD";
    Method["CREATE"] = "CREATE";
    Method["GET_ITEM"] = "GET_ITEM";
    Method["UPDATE"] = "UPDATE";
    Method["DELETE"] = "DELETE";
    Method["UPSERT"] = "UPSERT";
    Method["LIST"] = "LIST";
    Method["QUERY"] = "QUERY";
    Method["BULK_CREATE"] = "BULK_CREATE";
    Method["BULK_UPDATE"] = "BULK_UPDATE";
    Method["BULK_UPDATE_BY_FILTER"] = "BULK_UPDATE_BY_FILTER";
    Method["BULK_DELETE"] = "BULK_DELETE";
    Method["BULK_DELETE_BY_FILTER"] = "BULK_DELETE_BY_FILTER";
    Method["BULK_UPSERT"] = "BULK_UPSERT";
})(Method = exports.Method || (exports.Method = {}));
var Namespace;
(function (Namespace) {
    Namespace["DataStreams"] = "DataStreams";
    Namespace["WixStores"] = "WixStores";
    Namespace["Premium"] = "Premium";
    Namespace["VelocityInfra"] = "VelocityInfra";
    Namespace["Global"] = "Global";
    Namespace["Testing"] = "Testing";
    Namespace["Testing2"] = "Testing2";
    Namespace["AppMarket"] = "AppMarket";
    Namespace["Chat"] = "Chat";
    Namespace["Ping"] = "Ping";
    Namespace["CRM"] = "CRM";
    Namespace["Serverless"] = "Serverless";
    Namespace["DevEx"] = "DevEx";
    Namespace["Contacts"] = "Contacts";
    Namespace["Promote"] = "Promote";
    Namespace["DataStreamsLocal"] = "DataStreamsLocal";
    Namespace["Blog"] = "Blog";
    Namespace["Photography"] = "Photography";
    Namespace["Velo"] = "Velo";
    Namespace["Funnel"] = "Funnel";
    Namespace["BIVertical"] = "BIVertical";
    Namespace["Partners"] = "Partners";
    Namespace["OneApp"] = "OneApp";
    Namespace["DataStreamsOnDemand"] = "DataStreamsOnDemand";
    Namespace["WixCacheMainDcs"] = "WixCacheMainDcs";
    Namespace["WixCacheAllDcs"] = "WixCacheAllDcs";
    Namespace["WixCacheLocal"] = "WixCacheLocal";
    Namespace["MetaSite"] = "MetaSite";
})(Namespace = exports.Namespace || (exports.Namespace = {}));
var NearExpirationParamsParameters;
(function (NearExpirationParamsParameters) {
    NearExpirationParamsParameters["UNUSED"] = "UNUSED";
    NearExpirationParamsParameters["EXPIRATION_DATE"] = "EXPIRATION_DATE";
})(NearExpirationParamsParameters = exports.NearExpirationParamsParameters || (exports.NearExpirationParamsParameters = {}));
var NearRenewalParamsParameters;
(function (NearRenewalParamsParameters) {
    NearRenewalParamsParameters["UNUSED"] = "UNUSED";
    NearRenewalParamsParameters["PACKAGE_ID"] = "PACKAGE_ID";
    NearRenewalParamsParameters["CURRENCY"] = "CURRENCY";
    NearRenewalParamsParameters["NEXT_RENEWAL_DATE"] = "NEXT_RENEWAL_DATE";
    NearRenewalParamsParameters["CYCLE"] = "CYCLE";
    NearRenewalParamsParameters["PRICE"] = "PRICE";
})(NearRenewalParamsParameters = exports.NearRenewalParamsParameters || (exports.NearRenewalParamsParameters = {}));
var NearRenewalUpgradeCycleParamsParameters;
(function (NearRenewalUpgradeCycleParamsParameters) {
    NearRenewalUpgradeCycleParamsParameters["UNUSED"] = "UNUSED";
})(NearRenewalUpgradeCycleParamsParameters = exports.NearRenewalUpgradeCycleParamsParameters || (exports.NearRenewalUpgradeCycleParamsParameters = {}));
var NoActiveBillingAccountParamsParameters;
(function (NoActiveBillingAccountParamsParameters) {
    NoActiveBillingAccountParamsParameters["UNUSED"] = "UNUSED";
    NoActiveBillingAccountParamsParameters["PACKAGE_ID"] = "PACKAGE_ID";
    NoActiveBillingAccountParamsParameters["EMAIL"] = "EMAIL";
})(NoActiveBillingAccountParamsParameters = exports.NoActiveBillingAccountParamsParameters || (exports.NoActiveBillingAccountParamsParameters = {}));
var NotificationTypes;
(function (NotificationTypes) {
    NotificationTypes["UNKNOWN"] = "UNKNOWN";
    NotificationTypes["CREATED"] = "CREATED";
    NotificationTypes["ABOUT_TO_BE_RENEWED"] = "ABOUT_TO_BE_RENEWED";
    NotificationTypes["ABOUT_TO_BE_RENEWED_UPGRADE_CYCLE"] = "ABOUT_TO_BE_RENEWED_UPGRADE_CYCLE";
    NotificationTypes["ABOUT_TO_EXPIRE"] = "ABOUT_TO_EXPIRE";
    NotificationTypes["RECURRING_SUCCESS"] = "RECURRING_SUCCESS";
    NotificationTypes["AUTO_RENEW_TURNED_OFF"] = "AUTO_RENEW_TURNED_OFF";
    NotificationTypes["CANCELLED_BY_USER"] = "CANCELLED_BY_USER";
    NotificationTypes["CANCELLED_PAYMENT_PROBLEMS"] = "CANCELLED_PAYMENT_PROBLEMS";
    NotificationTypes["NO_ACTIVE_BILLING_ACCOUNT"] = "NO_ACTIVE_BILLING_ACCOUNT";
    NotificationTypes["CHARGE_FAILURE"] = "CHARGE_FAILURE";
    NotificationTypes["CHARGEBACK"] = "CHARGEBACK";
    NotificationTypes["MANUAL_RECURRING_REMINDER"] = "MANUAL_RECURRING_REMINDER";
    NotificationTypes["OFFLINE_PAYMENT"] = "OFFLINE_PAYMENT";
    NotificationTypes["SWITCH_CONTRACT"] = "SWITCH_CONTRACT";
    NotificationTypes["TRANSFER_SERVICE"] = "TRANSFER_SERVICE";
    NotificationTypes["ALERT_DISMISSED"] = "ALERT_DISMISSED";
})(NotificationTypes = exports.NotificationTypes || (exports.NotificationTypes = {}));
var NullValue;
(function (NullValue) {
    NullValue["NULL_VALUE"] = "NULL_VALUE";
})(NullValue = exports.NullValue || (exports.NullValue = {}));
var OfflinePaymentParamsParameters;
(function (OfflinePaymentParamsParameters) {
    OfflinePaymentParamsParameters["UNUSED"] = "UNUSED";
    OfflinePaymentParamsParameters["VOUCHERS"] = "VOUCHERS";
})(OfflinePaymentParamsParameters = exports.OfflinePaymentParamsParameters || (exports.OfflinePaymentParamsParameters = {}));
var Operator;
(function (Operator) {
    Operator["EQ"] = "EQ";
    Operator["LE"] = "LE";
    Operator["LT"] = "LT";
    Operator["GE"] = "GE";
    Operator["GT"] = "GT";
    Operator["BEGINS_WITH"] = "BEGINS_WITH";
    Operator["BETWEEN"] = "BETWEEN";
})(Operator = exports.Operator || (exports.Operator = {}));
var OptimizeMode;
(function (OptimizeMode) {
    OptimizeMode["SPEED"] = "SPEED";
    OptimizeMode["CODE_SIZE"] = "CODE_SIZE";
    OptimizeMode["LITE_RUNTIME"] = "LITE_RUNTIME";
})(OptimizeMode = exports.OptimizeMode || (exports.OptimizeMode = {}));
var OrderSessionItemContractSwitchType;
(function (OrderSessionItemContractSwitchType) {
    OrderSessionItemContractSwitchType["UNKNOWN"] = "UNKNOWN";
    OrderSessionItemContractSwitchType["REGULAR"] = "REGULAR";
    OrderSessionItemContractSwitchType["FULL_AMOUNT_PERIOD"] = "FULL_AMOUNT_PERIOD";
})(OrderSessionItemContractSwitchType = exports.OrderSessionItemContractSwitchType || (exports.OrderSessionItemContractSwitchType = {}));
var PagingType;
(function (PagingType) {
    PagingType["UNKNOWN_TYPE"] = "UNKNOWN_TYPE";
    PagingType["OFFSET"] = "OFFSET";
    PagingType["CURSOR"] = "CURSOR";
})(PagingType = exports.PagingType || (exports.PagingType = {}));
var PaymentFiltersPaymentType;
(function (PaymentFiltersPaymentType) {
    PaymentFiltersPaymentType["UNKNOWN"] = "UNKNOWN";
    PaymentFiltersPaymentType["RECURRING"] = "RECURRING";
    PaymentFiltersPaymentType["REAL_TIME"] = "REAL_TIME";
})(PaymentFiltersPaymentType = exports.PaymentFiltersPaymentType || (exports.PaymentFiltersPaymentType = {}));
var PaymentMethodSubType;
(function (PaymentMethodSubType) {
    PaymentMethodSubType["MP_OXXO"] = "MP_OXXO";
    PaymentMethodSubType["MP_BANAMEX"] = "MP_BANAMEX";
    PaymentMethodSubType["MP_BANCOMER"] = "MP_BANCOMER";
    PaymentMethodSubType["MP_SERFIN"] = "MP_SERFIN";
    PaymentMethodSubType["AP_BOLETO"] = "AP_BOLETO";
    PaymentMethodSubType["PAYU_CREDIT_CARD"] = "PAYU_CREDIT_CARD";
    PaymentMethodSubType["PAYU_DEBIT_CARD"] = "PAYU_DEBIT_CARD";
    PaymentMethodSubType["PAYU_NET_BANKING"] = "PAYU_NET_BANKING";
    PaymentMethodSubType["SOFORT"] = "SOFORT";
})(PaymentMethodSubType = exports.PaymentMethodSubType || (exports.PaymentMethodSubType = {}));
var PaymentTypeRecurringType;
(function (PaymentTypeRecurringType) {
    PaymentTypeRecurringType["None"] = "None";
    PaymentTypeRecurringType["AUTO"] = "AUTO";
    PaymentTypeRecurringType["HYBRID"] = "HYBRID";
})(PaymentTypeRecurringType = exports.PaymentTypeRecurringType || (exports.PaymentTypeRecurringType = {}));
var PeriodEnumPeriod;
(function (PeriodEnumPeriod) {
    PeriodEnumPeriod["UNDEFINED"] = "UNDEFINED";
    PeriodEnumPeriod["DAILY"] = "DAILY";
    PeriodEnumPeriod["WEEKLY"] = "WEEKLY";
    PeriodEnumPeriod["MONTHLY"] = "MONTHLY";
})(PeriodEnumPeriod = exports.PeriodEnumPeriod || (exports.PeriodEnumPeriod = {}));
var PeriodType;
(function (PeriodType) {
    PeriodType["CUSTOM"] = "CUSTOM";
    PeriodType["TODAY"] = "TODAY";
    PeriodType["YESTERDAY"] = "YESTERDAY";
    PeriodType["LAST_7_DAYS"] = "LAST_7_DAYS";
    PeriodType["LAST_14_DAYS"] = "LAST_14_DAYS";
    PeriodType["LAST_28_DAYS"] = "LAST_28_DAYS";
    PeriodType["LAST_30_DAYS"] = "LAST_30_DAYS";
    PeriodType["LAST_90_DAYS"] = "LAST_90_DAYS";
    PeriodType["LAST_180_DAYS"] = "LAST_180_DAYS";
    PeriodType["LAST_365_DAYS"] = "LAST_365_DAYS";
    PeriodType["THIS_WEEK"] = "THIS_WEEK";
    PeriodType["THIS_MONTH"] = "THIS_MONTH";
    PeriodType["THIS_QUARTER"] = "THIS_QUARTER";
    PeriodType["THIS_YEAR"] = "THIS_YEAR";
    PeriodType["LAST_WEEK"] = "LAST_WEEK";
    PeriodType["LAST_MONTH"] = "LAST_MONTH";
    PeriodType["LAST_QUARTER"] = "LAST_QUARTER";
    PeriodType["LAST_YEAR"] = "LAST_YEAR";
})(PeriodType = exports.PeriodType || (exports.PeriodType = {}));
var PermissionType;
(function (PermissionType) {
    PermissionType["AUTO"] = "AUTO";
    PermissionType["MANUAL"] = "MANUAL";
})(PermissionType = exports.PermissionType || (exports.PermissionType = {}));
var PredefinedTimeframe;
(function (PredefinedTimeframe) {
    PredefinedTimeframe["LAST_7_DAYS"] = "LAST_7_DAYS";
    PredefinedTimeframe["LAST_14_DAYS"] = "LAST_14_DAYS";
    PredefinedTimeframe["LAST_30_DAYS"] = "LAST_30_DAYS";
    PredefinedTimeframe["LAST_90_DAYS"] = "LAST_90_DAYS";
    PredefinedTimeframe["THIS_WEEK"] = "THIS_WEEK";
    PredefinedTimeframe["THIS_MONTH"] = "THIS_MONTH";
})(PredefinedTimeframe = exports.PredefinedTimeframe || (exports.PredefinedTimeframe = {}));
var ProductStatus;
(function (ProductStatus) {
    ProductStatus["UNKNOWN"] = "UNKNOWN";
    ProductStatus["DRAFT"] = "DRAFT";
    ProductStatus["PUBLISHED"] = "PUBLISHED";
    ProductStatus["ARCHIVED"] = "ARCHIVED";
})(ProductStatus = exports.ProductStatus || (exports.ProductStatus = {}));
var ProductTypeContext;
(function (ProductTypeContext) {
    ProductTypeContext["UNKNOWN"] = "UNKNOWN";
    ProductTypeContext["ACCOUNT"] = "ACCOUNT";
    ProductTypeContext["SITE"] = "SITE";
})(ProductTypeContext = exports.ProductTypeContext || (exports.ProductTypeContext = {}));
var PurchaseScenario;
(function (PurchaseScenario) {
    PurchaseScenario["UNAUTHORIZED"] = "UNAUTHORIZED";
    PurchaseScenario["ACCOUNT_LEVEL"] = "ACCOUNT_LEVEL";
    PurchaseScenario["PURCHASE"] = "PURCHASE";
    PurchaseScenario["VIEW"] = "VIEW";
    PurchaseScenario["EMPTY_OFFERING"] = "EMPTY_OFFERING";
    PurchaseScenario["ANONYMOUS"] = "ANONYMOUS";
})(PurchaseScenario = exports.PurchaseScenario || (exports.PurchaseScenario = {}));
var QuantityType;
(function (QuantityType) {
    QuantityType["UNKNOWN"] = "UNKNOWN";
    QuantityType["BOOLEAN"] = "BOOLEAN";
    QuantityType["QUOTA"] = "QUOTA";
})(QuantityType = exports.QuantityType || (exports.QuantityType = {}));
var RecurringChargeSuccessParamsParameters;
(function (RecurringChargeSuccessParamsParameters) {
    RecurringChargeSuccessParamsParameters["UNUSED"] = "UNUSED";
})(RecurringChargeSuccessParamsParameters = exports.RecurringChargeSuccessParamsParameters || (exports.RecurringChargeSuccessParamsParameters = {}));
var RecurringTypeMessageRecurringType;
(function (RecurringTypeMessageRecurringType) {
    RecurringTypeMessageRecurringType["UNKNOWN"] = "UNKNOWN";
    RecurringTypeMessageRecurringType["MANUAL"] = "MANUAL";
    RecurringTypeMessageRecurringType["OFF"] = "OFF";
    RecurringTypeMessageRecurringType["AUTO"] = "AUTO";
})(RecurringTypeMessageRecurringType = exports.RecurringTypeMessageRecurringType || (exports.RecurringTypeMessageRecurringType = {}));
var RefundReason;
(function (RefundReason) {
    RefundReason["UNKNOWN"] = "UNKNOWN";
    RefundReason["FRAUDULENT"] = "FRAUDULENT";
    RefundReason["REQUESTED_BY_CUSTOMER"] = "REQUESTED_BY_CUSTOMER";
})(RefundReason = exports.RefundReason || (exports.RefundReason = {}));
var RuleType;
(function (RuleType) {
    RuleType["VALIDATION"] = "VALIDATION";
    RuleType["OTHER"] = "OTHER";
    RuleType["MAX"] = "MAX";
    RuleType["MIN"] = "MIN";
    RuleType["MAX_LENGTH"] = "MAX_LENGTH";
    RuleType["MIN_LENGTH"] = "MIN_LENGTH";
    RuleType["MAX_SIZE"] = "MAX_SIZE";
    RuleType["MIN_SIZE"] = "MIN_SIZE";
    RuleType["FORMAT"] = "FORMAT";
    RuleType["DECIMAL_LTE"] = "DECIMAL_LTE";
    RuleType["DECIMAL_GTE"] = "DECIMAL_GTE";
    RuleType["DECIMAL_LT"] = "DECIMAL_LT";
    RuleType["DECIMAL_GT"] = "DECIMAL_GT";
    RuleType["DECIMAL_MAX_SCALE"] = "DECIMAL_MAX_SCALE";
    RuleType["INVALID_ENUM_VALUE"] = "INVALID_ENUM_VALUE";
    RuleType["REQUIRED_FIELD"] = "REQUIRED_FIELD";
})(RuleType = exports.RuleType || (exports.RuleType = {}));
var ScalarType;
(function (ScalarType) {
    ScalarType["UNKNOWN_SCALAR_TYPE"] = "UNKNOWN_SCALAR_TYPE";
    ScalarType["COUNT_DISTINCT"] = "COUNT_DISTINCT";
    ScalarType["MIN"] = "MIN";
    ScalarType["MAX"] = "MAX";
    ScalarType["SUM"] = "SUM";
    ScalarType["AVG"] = "AVG";
})(ScalarType = exports.ScalarType || (exports.ScalarType = {}));
var SearchDetailsMode;
(function (SearchDetailsMode) {
    SearchDetailsMode["OR"] = "OR";
    SearchDetailsMode["AND"] = "AND";
})(SearchDetailsMode = exports.SearchDetailsMode || (exports.SearchDetailsMode = {}));
var SegmentSegment;
(function (SegmentSegment) {
    SegmentSegment["UNSPECIFIED"] = "UNSPECIFIED";
    SegmentSegment["USERS"] = "USERS";
    SegmentSegment["PUBLIC"] = "PUBLIC";
    SegmentSegment["OTHERS"] = "OTHERS";
})(SegmentSegment = exports.SegmentSegment || (exports.SegmentSegment = {}));
var ServiceAppendixType;
(function (ServiceAppendixType) {
    ServiceAppendixType["UNKNOWN_APPENDIX_TYPE"] = "UNKNOWN_APPENDIX_TYPE";
    ServiceAppendixType["NILE"] = "NILE";
    ServiceAppendixType["MESSAGE_BUS"] = "MESSAGE_BUS";
})(ServiceAppendixType = exports.ServiceAppendixType || (exports.ServiceAppendixType = {}));
var ServiceStatus;
(function (ServiceStatus) {
    ServiceStatus["UNKNOWN"] = "UNKNOWN";
    ServiceStatus["ACTIVE"] = "ACTIVE";
    ServiceStatus["CANCELED"] = "CANCELED";
    ServiceStatus["FRAUDULENT"] = "FRAUDULENT";
    ServiceStatus["SUSPENDED"] = "SUSPENDED";
    ServiceStatus["UNPAID"] = "UNPAID";
})(ServiceStatus = exports.ServiceStatus || (exports.ServiceStatus = {}));
var SettingsPaymentMethod;
(function (SettingsPaymentMethod) {
    SettingsPaymentMethod["UNKNOWN"] = "UNKNOWN";
    SettingsPaymentMethod["CARD"] = "CARD";
    SettingsPaymentMethod["DIRECT_DEBIT"] = "DIRECT_DEBIT";
    SettingsPaymentMethod["CASH"] = "CASH";
})(SettingsPaymentMethod = exports.SettingsPaymentMethod || (exports.SettingsPaymentMethod = {}));
var SortDirection;
(function (SortDirection) {
    SortDirection["DESC"] = "DESC";
    SortDirection["ASC"] = "ASC";
})(SortDirection = exports.SortDirection || (exports.SortDirection = {}));
var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder = exports.SortOrder || (exports.SortOrder = {}));
var SortType;
(function (SortType) {
    SortType["COUNT"] = "COUNT";
    SortType["VALUE"] = "VALUE";
})(SortType = exports.SortType || (exports.SortType = {}));
var SpeedState;
(function (SpeedState) {
    SpeedState["SUPER_FAST"] = "SUPER_FAST";
    SpeedState["PRETTY_FAST"] = "PRETTY_FAST";
    SpeedState["MODERATE"] = "MODERATE";
    SpeedState["COULD_LOAD_FASTER"] = "COULD_LOAD_FASTER";
})(SpeedState = exports.SpeedState || (exports.SpeedState = {}));
var StreamViewType;
(function (StreamViewType) {
    StreamViewType["NEW_AND_OLD_IMAGES"] = "NEW_AND_OLD_IMAGES";
    StreamViewType["KEYS_ONLY"] = "KEYS_ONLY";
    StreamViewType["NEW_IMAGE"] = "NEW_IMAGE";
    StreamViewType["OLD_IMAGE"] = "OLD_IMAGE";
})(StreamViewType = exports.StreamViewType || (exports.StreamViewType = {}));
var SubdivisionType;
(function (SubdivisionType) {
    SubdivisionType["UNKNOWN_SUBDIVISION_TYPE"] = "UNKNOWN_SUBDIVISION_TYPE";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_1"] = "ADMINISTRATIVE_AREA_LEVEL_1";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_2"] = "ADMINISTRATIVE_AREA_LEVEL_2";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_3"] = "ADMINISTRATIVE_AREA_LEVEL_3";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_4"] = "ADMINISTRATIVE_AREA_LEVEL_4";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_5"] = "ADMINISTRATIVE_AREA_LEVEL_5";
    SubdivisionType["COUNTRY"] = "COUNTRY";
})(SubdivisionType = exports.SubdivisionType || (exports.SubdivisionType = {}));
var SubscriptionUpgradeType;
(function (SubscriptionUpgradeType) {
    SubscriptionUpgradeType["UNKNOWN"] = "UNKNOWN";
    SubscriptionUpgradeType["CREDIT_UNUSED_PERIOD"] = "CREDIT_UNUSED_PERIOD";
    SubscriptionUpgradeType["EXTEND"] = "EXTEND";
})(SubscriptionUpgradeType = exports.SubscriptionUpgradeType || (exports.SubscriptionUpgradeType = {}));
var SwitchContractParamsParameters;
(function (SwitchContractParamsParameters) {
    SwitchContractParamsParameters["UNUSED"] = "UNUSED";
    SwitchContractParamsParameters["INVOICE_NUM"] = "INVOICE_NUM";
    SwitchContractParamsParameters["BILLINGVENDOR"] = "BILLINGVENDOR";
    SwitchContractParamsParameters["PAYMENT_CYCLE"] = "PAYMENT_CYCLE";
    SwitchContractParamsParameters["CYCLE_ID"] = "CYCLE_ID";
    SwitchContractParamsParameters["SBS_PRODUCT_ID"] = "SBS_PRODUCT_ID";
    SwitchContractParamsParameters["NEXT_RENEWAL_AMOUNT"] = "NEXT_RENEWAL_AMOUNT";
    SwitchContractParamsParameters["NEXT_RENEWAL_CURRENCY"] = "NEXT_RENEWAL_CURRENCY";
    SwitchContractParamsParameters["PAYMENT_TYPE"] = "PAYMENT_TYPE";
})(SwitchContractParamsParameters = exports.SwitchContractParamsParameters || (exports.SwitchContractParamsParameters = {}));
var TaxIdFieldMode;
(function (TaxIdFieldMode) {
    TaxIdFieldMode["UNKNOWN"] = "UNKNOWN";
    TaxIdFieldMode["HIDE"] = "HIDE";
    TaxIdFieldMode["OPTIONAL"] = "OPTIONAL";
    TaxIdFieldMode["MANDATORY"] = "MANDATORY";
})(TaxIdFieldMode = exports.TaxIdFieldMode || (exports.TaxIdFieldMode = {}));
var TaxSettingsTaxType;
(function (TaxSettingsTaxType) {
    TaxSettingsTaxType["NOT_APPLICABLE"] = "NOT_APPLICABLE";
    TaxSettingsTaxType["VAT"] = "VAT";
    TaxSettingsTaxType["GST"] = "GST";
})(TaxSettingsTaxType = exports.TaxSettingsTaxType || (exports.TaxSettingsTaxType = {}));
var TaxTypeConfigurationTaxType;
(function (TaxTypeConfigurationTaxType) {
    TaxTypeConfigurationTaxType["NOT_APPLICABLE"] = "NOT_APPLICABLE";
    TaxTypeConfigurationTaxType["REGULAR"] = "REGULAR";
    TaxTypeConfigurationTaxType["EU_VAT"] = "EU_VAT";
    TaxTypeConfigurationTaxType["JAPAN_JCT"] = "JAPAN_JCT";
    TaxTypeConfigurationTaxType["AU_GST"] = "AU_GST";
    TaxTypeConfigurationTaxType["BR_CNPJ"] = "BR_CNPJ";
    TaxTypeConfigurationTaxType["RU_TIN"] = "RU_TIN";
    TaxTypeConfigurationTaxType["IN_RNR"] = "IN_RNR";
    TaxTypeConfigurationTaxType["IL_VAT"] = "IL_VAT";
    TaxTypeConfigurationTaxType["US_TAX"] = "US_TAX";
    TaxTypeConfigurationTaxType["SG_GST"] = "SG_GST";
    TaxTypeConfigurationTaxType["CA_GST"] = "CA_GST";
    TaxTypeConfigurationTaxType["CA_HST"] = "CA_HST";
    TaxTypeConfigurationTaxType["CA_PST"] = "CA_PST";
    TaxTypeConfigurationTaxType["CA_QST"] = "CA_QST";
    TaxTypeConfigurationTaxType["NZ_GST"] = "NZ_GST";
    TaxTypeConfigurationTaxType["MD_VAT"] = "MD_VAT";
    TaxTypeConfigurationTaxType["CL_VAT"] = "CL_VAT";
    TaxTypeConfigurationTaxType["MY_VAT"] = "MY_VAT";
    TaxTypeConfigurationTaxType["KR_VAT"] = "KR_VAT";
})(TaxTypeConfigurationTaxType = exports.TaxTypeConfigurationTaxType || (exports.TaxTypeConfigurationTaxType = {}));
var ThreeDsChallengeDisplaySizeEnum;
(function (ThreeDsChallengeDisplaySizeEnum) {
    ThreeDsChallengeDisplaySizeEnum["UNKNOWN"] = "UNKNOWN";
    ThreeDsChallengeDisplaySizeEnum["SMALL"] = "SMALL";
    ThreeDsChallengeDisplaySizeEnum["MEDIUM"] = "MEDIUM";
    ThreeDsChallengeDisplaySizeEnum["LARGE"] = "LARGE";
})(ThreeDsChallengeDisplaySizeEnum = exports.ThreeDsChallengeDisplaySizeEnum || (exports.ThreeDsChallengeDisplaySizeEnum = {}));
var ThreeDsProviderEnum;
(function (ThreeDsProviderEnum) {
    ThreeDsProviderEnum["UNKNOWN"] = "UNKNOWN";
    ThreeDsProviderEnum["ADYEN"] = "ADYEN";
    ThreeDsProviderEnum["WORLDPAY"] = "WORLDPAY";
})(ThreeDsProviderEnum = exports.ThreeDsProviderEnum || (exports.ThreeDsProviderEnum = {}));
var TransferServiceParamsParameters;
(function (TransferServiceParamsParameters) {
    TransferServiceParamsParameters["UNUSED"] = "UNUSED";
})(TransferServiceParamsParameters = exports.TransferServiceParamsParameters || (exports.TransferServiceParamsParameters = {}));
var TranslatableType;
(function (TranslatableType) {
    TranslatableType["UNDEFINED_TYPE"] = "UNDEFINED_TYPE";
    TranslatableType["SHORT_TEXT"] = "SHORT_TEXT";
    TranslatableType["LONG_TEXT"] = "LONG_TEXT";
    TranslatableType["RICH_TEXT"] = "RICH_TEXT";
    TranslatableType["RICH_CONTENT_EDITOR"] = "RICH_CONTENT_EDITOR";
    TranslatableType["SELECTION"] = "SELECTION";
    TranslatableType["MULTI_SELECTION"] = "MULTI_SELECTION";
    TranslatableType["DOCUMENT"] = "DOCUMENT";
    TranslatableType["IMAGE"] = "IMAGE";
    TranslatableType["VIDEO"] = "VIDEO";
})(TranslatableType = exports.TranslatableType || (exports.TranslatableType = {}));
var V1ContractSwitchType;
(function (V1ContractSwitchType) {
    V1ContractSwitchType["UNKNOWN"] = "UNKNOWN";
    V1ContractSwitchType["REGULAR"] = "REGULAR";
    V1ContractSwitchType["FULL_AMOUNT_PERIOD"] = "FULL_AMOUNT_PERIOD";
})(V1ContractSwitchType = exports.V1ContractSwitchType || (exports.V1ContractSwitchType = {}));
var V1Mode;
(function (V1Mode) {
    V1Mode["DESKTOP"] = "DESKTOP";
    V1Mode["MOBILE"] = "MOBILE";
    V1Mode["ADVANCED"] = "ADVANCED";
})(V1Mode = exports.V1Mode || (exports.V1Mode = {}));
var V1OrderSessionPaymentMethodType;
(function (V1OrderSessionPaymentMethodType) {
    V1OrderSessionPaymentMethodType["UNKNOWN"] = "UNKNOWN";
    V1OrderSessionPaymentMethodType["CARD"] = "CARD";
    V1OrderSessionPaymentMethodType["DIRECT_DEBIT"] = "DIRECT_DEBIT";
    V1OrderSessionPaymentMethodType["CASH"] = "CASH";
    V1OrderSessionPaymentMethodType["BANK_TRANSFER"] = "BANK_TRANSFER";
})(V1OrderSessionPaymentMethodType = exports.V1OrderSessionPaymentMethodType || (exports.V1OrderSessionPaymentMethodType = {}));
var V2OrderSessionPaymentMethodType;
(function (V2OrderSessionPaymentMethodType) {
    V2OrderSessionPaymentMethodType["UNKNOWN"] = "UNKNOWN";
    V2OrderSessionPaymentMethodType["CARD"] = "CARD";
    V2OrderSessionPaymentMethodType["DIRECT_DEBIT"] = "DIRECT_DEBIT";
    V2OrderSessionPaymentMethodType["CASH"] = "CASH";
    V2OrderSessionPaymentMethodType["BANK_TRANSFER"] = "BANK_TRANSFER";
})(V2OrderSessionPaymentMethodType = exports.V2OrderSessionPaymentMethodType || (exports.V2OrderSessionPaymentMethodType = {}));
var ValidateNoPreviousPurchaseType;
(function (ValidateNoPreviousPurchaseType) {
    ValidateNoPreviousPurchaseType["UNKNOWN"] = "UNKNOWN";
    ValidateNoPreviousPurchaseType["OK"] = "OK";
    ValidateNoPreviousPurchaseType["PENDING"] = "PENDING";
    ValidateNoPreviousPurchaseType["DELIVERED"] = "DELIVERED";
    ValidateNoPreviousPurchaseType["SERVER_ERROR"] = "SERVER_ERROR";
})(ValidateNoPreviousPurchaseType = exports.ValidateNoPreviousPurchaseType || (exports.ValidateNoPreviousPurchaseType = {}));
var VatType;
(function (VatType) {
    VatType["UNSPECIFIED"] = "UNSPECIFIED";
    VatType["CPF"] = "CPF";
    VatType["CNPJ"] = "CNPJ";
})(VatType = exports.VatType || (exports.VatType = {}));
