// @wix/ambassador-wix-quotes-web/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifyInvoiceErrorCodes = exports.ValueType = exports.SubdivisionType = exports.StructNullValue = exports.SourceType = exports.QuotesSortOrder = exports.PriceQuoteStatus = exports.Placement = exports.PaymentTermType = exports.InvoicesSortableField = exports.InvoiceStatus = exports.IdlSortableField = exports.ErrorResult = exports.DisplayStatus = exports.CustomFieldGroup = exports.CommonSortOrder = void 0;
var CommonSortOrder;
(function (CommonSortOrder) {
    CommonSortOrder["ASC"] = "ASC";
    CommonSortOrder["DESC"] = "DESC";
})(CommonSortOrder = exports.CommonSortOrder || (exports.CommonSortOrder = {}));
var CustomFieldGroup;
(function (CustomFieldGroup) {
    CustomFieldGroup["UNKNOWN_CUSTOM_FIELD_GROUP"] = "UNKNOWN_CUSTOM_FIELD_GROUP";
    CustomFieldGroup["BUSINESS_DETAILS"] = "BUSINESS_DETAILS";
    CustomFieldGroup["CUSTOMER_DETAILS"] = "CUSTOMER_DETAILS";
    CustomFieldGroup["DOCUMENT"] = "DOCUMENT";
    CustomFieldGroup["FOOTER"] = "FOOTER";
    CustomFieldGroup["OTHER"] = "OTHER";
})(CustomFieldGroup = exports.CustomFieldGroup || (exports.CustomFieldGroup = {}));
var DisplayStatus;
(function (DisplayStatus) {
    DisplayStatus["Draft"] = "Draft";
    DisplayStatus["Sent"] = "Sent";
    DisplayStatus["Processing"] = "Processing";
    DisplayStatus["Paid"] = "Paid";
    DisplayStatus["Overdue"] = "Overdue";
    DisplayStatus["Void"] = "Void";
    DisplayStatus["Deleted"] = "Deleted";
    DisplayStatus["PartiallyPaid"] = "PartiallyPaid";
    DisplayStatus["PartialAndOverdue"] = "PartialAndOverdue";
})(DisplayStatus = exports.DisplayStatus || (exports.DisplayStatus = {}));
var ErrorResult;
(function (ErrorResult) {
    ErrorResult["NoError"] = "NoError";
    ErrorResult["InvalidTpaInstance"] = "InvalidTpaInstance";
    ErrorResult["EntityNotFound"] = "EntityNotFound";
    ErrorResult["TokenValidationError"] = "TokenValidationError";
    ErrorResult["AppProvisionError"] = "AppProvisionError";
    ErrorResult["DuplicateAggregate"] = "DuplicateAggregate";
    ErrorResult["UnknownTimeRange"] = "UnknownTimeRange";
    ErrorResult["IllegalParameter"] = "IllegalParameter";
    ErrorResult["DuplicateInvoiceNumber"] = "DuplicateInvoiceNumber";
    ErrorResult["ConcurrentUpdate"] = "ConcurrentUpdate";
    ErrorResult["NotSupported"] = "NotSupported";
})(ErrorResult = exports.ErrorResult || (exports.ErrorResult = {}));
var IdlSortableField;
(function (IdlSortableField) {
    IdlSortableField["QuoteNumber"] = "QuoteNumber";
    IdlSortableField["ByCustomer"] = "ByCustomer";
    IdlSortableField["Total"] = "Total";
    IdlSortableField["IssueDate"] = "IssueDate";
    IdlSortableField["ValidThroughDate"] = "ValidThroughDate";
    IdlSortableField["AcceptanceDate"] = "AcceptanceDate";
    IdlSortableField["Status"] = "Status";
    IdlSortableField["Title"] = "Title";
})(IdlSortableField = exports.IdlSortableField || (exports.IdlSortableField = {}));
var InvoiceStatus;
(function (InvoiceStatus) {
    InvoiceStatus["Draft"] = "Draft";
    InvoiceStatus["Sent"] = "Sent";
    InvoiceStatus["Processing"] = "Processing";
    InvoiceStatus["Paid"] = "Paid";
    InvoiceStatus["Overdue"] = "Overdue";
    InvoiceStatus["Void"] = "Void";
    InvoiceStatus["Deleted"] = "Deleted";
    InvoiceStatus["PartiallyPaid"] = "PartiallyPaid";
    InvoiceStatus["PartialAndOverdue"] = "PartialAndOverdue";
})(InvoiceStatus = exports.InvoiceStatus || (exports.InvoiceStatus = {}));
var InvoicesSortableField;
(function (InvoicesSortableField) {
    InvoicesSortableField["QuoteNumber"] = "QuoteNumber";
    InvoicesSortableField["ByCustomer"] = "ByCustomer";
    InvoicesSortableField["Total"] = "Total";
    InvoicesSortableField["IssueDate"] = "IssueDate";
    InvoicesSortableField["DueDate"] = "DueDate";
    InvoicesSortableField["Status"] = "Status";
})(InvoicesSortableField = exports.InvoicesSortableField || (exports.InvoicesSortableField = {}));
var PaymentTermType;
(function (PaymentTermType) {
    PaymentTermType["DueOnReceipt"] = "DueOnReceipt";
    PaymentTermType["NetPlus"] = "NetPlus";
    PaymentTermType["TimeStamp"] = "TimeStamp";
    PaymentTermType["Other"] = "Other";
})(PaymentTermType = exports.PaymentTermType || (exports.PaymentTermType = {}));
var Placement;
(function (Placement) {
    Placement["Unknown"] = "Unknown";
    Placement["Replace"] = "Replace";
    Placement["Before"] = "Before";
    Placement["After"] = "After";
})(Placement = exports.Placement || (exports.Placement = {}));
var PriceQuoteStatus;
(function (PriceQuoteStatus) {
    PriceQuoteStatus["Draft"] = "Draft";
    PriceQuoteStatus["Sent"] = "Sent";
    PriceQuoteStatus["Processing"] = "Processing";
    PriceQuoteStatus["Accepted"] = "Accepted";
    PriceQuoteStatus["Rejected"] = "Rejected";
    PriceQuoteStatus["Expired"] = "Expired";
    PriceQuoteStatus["Void"] = "Void";
    PriceQuoteStatus["Deleted"] = "Deleted";
    PriceQuoteStatus["Invoiced"] = "Invoiced";
})(PriceQuoteStatus = exports.PriceQuoteStatus || (exports.PriceQuoteStatus = {}));
var QuotesSortOrder;
(function (QuotesSortOrder) {
    QuotesSortOrder["Ascending"] = "Ascending";
    QuotesSortOrder["Descending"] = "Descending";
})(QuotesSortOrder = exports.QuotesSortOrder || (exports.QuotesSortOrder = {}));
var SourceType;
(function (SourceType) {
    SourceType["UNKNOWN_SOURCE_TYPE"] = "UNKNOWN_SOURCE_TYPE";
    SourceType["WIX_APP"] = "WIX_APP";
    SourceType["EXTERNAL"] = "EXTERNAL";
    SourceType["ADMIN"] = "ADMIN";
    SourceType["OTHER"] = "OTHER";
})(SourceType = exports.SourceType || (exports.SourceType = {}));
var StructNullValue;
(function (StructNullValue) {
    StructNullValue["NULL_VALUE"] = "NULL_VALUE";
})(StructNullValue = exports.StructNullValue || (exports.StructNullValue = {}));
var SubdivisionType;
(function (SubdivisionType) {
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_2"] = "ADMINISTRATIVE_AREA_LEVEL_2";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_4"] = "ADMINISTRATIVE_AREA_LEVEL_4";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_3"] = "ADMINISTRATIVE_AREA_LEVEL_3";
    SubdivisionType["UNKNOWN_SUBDIVISION_TYPE"] = "UNKNOWN_SUBDIVISION_TYPE";
    SubdivisionType["COUNTRY"] = "COUNTRY";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_1"] = "ADMINISTRATIVE_AREA_LEVEL_1";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_5"] = "ADMINISTRATIVE_AREA_LEVEL_5";
})(SubdivisionType = exports.SubdivisionType || (exports.SubdivisionType = {}));
var ValueType;
(function (ValueType) {
    ValueType["UNKNOWN_VALUE_TYPE"] = "UNKNOWN_VALUE_TYPE";
    ValueType["STRING"] = "STRING";
    ValueType["DATE"] = "DATE";
    ValueType["BOOLEAN"] = "BOOLEAN";
    ValueType["NUMBER"] = "NUMBER";
})(ValueType = exports.ValueType || (exports.ValueType = {}));
var VerifyInvoiceErrorCodes;
(function (VerifyInvoiceErrorCodes) {
    VerifyInvoiceErrorCodes["UNKNOWN_VERIFICATION_ERROR"] = "UNKNOWN_VERIFICATION_ERROR";
    VerifyInvoiceErrorCodes["FAILED_READING_BUSINESS_INFO"] = "FAILED_READING_BUSINESS_INFO";
    VerifyInvoiceErrorCodes["MISSING_INVOICE_CUSTOMER"] = "MISSING_INVOICE_CUSTOMER";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_CUSTOMER_EMAIL"] = "INVALID_INVOICE_CUSTOMER_EMAIL";
    VerifyInvoiceErrorCodes["MISSING_INVOICE_CUSTOMER_CONTACT_ID"] = "MISSING_INVOICE_CUSTOMER_CONTACT_ID";
    VerifyInvoiceErrorCodes["MISSING_INVOICE_LINE_ITEMS"] = "MISSING_INVOICE_LINE_ITEMS";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_LINE_ITEM_EMPTY_NAME"] = "INVALID_INVOICE_LINE_ITEM_EMPTY_NAME";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_LINE_ITEM_TAX"] = "INVALID_INVOICE_LINE_ITEM_TAX";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_ISSUE_DATE"] = "INVALID_INVOICE_ISSUE_DATE";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_DUE_DATE_BEFORE_ISSUE_DATE"] = "INVALID_INVOICE_DUE_DATE_BEFORE_ISSUE_DATE";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_DISCOUNT_RATE"] = "INVALID_INVOICE_DISCOUNT_RATE";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_TOTAL_IS_NEGATIVE"] = "INVALID_INVOICE_TOTAL_IS_NEGATIVE";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_TOTAL_TOO_LARGE"] = "INVALID_INVOICE_TOTAL_TOO_LARGE";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_BALANCE_IS_NEGATIVE"] = "INVALID_INVOICE_BALANCE_IS_NEGATIVE";
    VerifyInvoiceErrorCodes["MISSING_INVOICE_CURRENCY"] = "MISSING_INVOICE_CURRENCY";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_CURRENCY"] = "INVALID_INVOICE_CURRENCY";
    VerifyInvoiceErrorCodes["INVALID_INVOICE_FIELDS"] = "INVALID_INVOICE_FIELDS";
})(VerifyInvoiceErrorCodes = exports.VerifyInvoiceErrorCodes || (exports.VerifyInvoiceErrorCodes = {}));
