// @wix/my-account-header/dist/es/src/client/selectors/expertsSelectors.js
import { experimentsSelectors } from '@wix/wix-experiments-redux';
import { ExpertType } from './../types/experts';
import { ExpertsAccountLevel } from '@wix/header-common';
export var isSiteBuilder = function (state) { return state.expertTypes.types.has(ExpertType.SITE_BUILDER); };
export var isPro = function (state) { return state.expertTypes.types.has(ExpertType.PRO); };
export var isPotentialExpert = function (state) { return state.expertTypes.types.has(ExpertType.POTENTIAL_EXPERT); };
export var isExpertDashboardEnabled = function (state) { return state.expertTypes.isExpertDashboardEnabled; };
export var showExpertExperience = function (state) {
    return experimentsSelectors(state).enabled('specs.partners.OptOutDashboard')
        || experimentsSelectors(state).enabled('specs.wosbm.expertsServerDeprecation') ?
        isExpertDashboardEnabled(state) :
        isSiteBuilder(state) || isPotentialExpert(state);
};
export var showProHelp = function (state) {
    return experimentsSelectors(state).enabled('specs.wosbm.expertsServerDeprecation') ?
        isExpertDashboardEnabled(state) :
        isSiteBuilder(state) || isPro(state);
};
export var isAccountLevelAboveExplorer = function (state) { return (state.expertTypes.accountLevel &&
    state.expertTypes.accountLevel !== ExpertsAccountLevel.NOT_PARTNER &&
    state.expertTypes.accountLevel !== ExpertsAccountLevel.EXPLORER); };
