// @wix/my-account-header/dist/es/src/client/selectors/logoSelector.js
import { isAccountCoBrandingDataReceived } from './accountSelectors';
import { siteCoBrandingDataExists } from './siteCoBrandingSelector';
import { getCoBrandingLogoUrl, isCoBranded } from './coBrandingSelector';
import { isEditorXBrand } from './isEditorXSelector';
import { isNtt } from '../services/cookieService/cookieService';
export var getLogoType = function (state) {
    var NTT = LogoTypes.NTT, CO_BRANDING = LogoTypes.CO_BRANDING, EDITORX = LogoTypes.EDITORX, WIX = LogoTypes.WIX;
    if (isNtt()) {
        return NTT;
    }
    else if (isCoBranded(state) && !!getCoBrandingLogoUrl(state)) {
        return CO_BRANDING;
    }
    else if (isEditorXBrand(state)) {
        return EDITORX;
    }
    else {
        return WIX;
    }
};
export var isLogoTypeDetermined = function (state) {
    return isAccountCoBrandingDataReceived(state) || siteCoBrandingDataExists(state);
};
export var LogoTypes;
(function (LogoTypes) {
    LogoTypes["WIX"] = "wix";
    LogoTypes["NTT"] = "ntt";
    LogoTypes["CO_BRANDING"] = "coBranding";
    LogoTypes["EDITORX"] = "editorX";
})(LogoTypes || (LogoTypes = {}));
