// @wix/my-account-header/dist/es/src/client/selectors/recentSitesListSelector.js
export var getRecentSitesList = function (state) { var _a; return (_a = state.recentSitesList) === null || _a === void 0 ? void 0 : _a.list; };
export var getCurrentSite = function (_a) {
    var currentSite = _a.currentSite, recentSitesList = _a.recentSitesList;
    return currentSite && (recentSitesList === null || recentSitesList === void 0 ? void 0 : recentSitesList.list) &&
        recentSitesList.list.find(function (site) { return site.msid === currentSite.id; });
};
export var getSitesWithoutCurrent = function (_a) {
    var currentSite = _a.currentSite, recentSitesList = _a.recentSitesList;
    return (recentSitesList === null || recentSitesList === void 0 ? void 0 : recentSitesList.list) &&
        (currentSite ? recentSitesList.list.filter(function (site) { return site.msid !== currentSite.id; }) : recentSitesList.list);
};
export var hasRecentSitesListError = function (state) { var _a; return (_a = state.recentSitesList) === null || _a === void 0 ? void 0 : _a.hasError; };
export var hasNoSites = function (state) { var _a; return ((_a = getRecentSitesList(state)) === null || _a === void 0 ? void 0 : _a.length) === 0; };
export var isLoadingSitesList = function (state) { var _a; return (_a = state.recentSitesList) === null || _a === void 0 ? void 0 : _a.isLoading; };
