// @wix/dashboard-smart-answers-widget/dist/esm/services/prediction/prediction-services.js
import { __awaiter, __generator } from "tslib";
import { PredictorPrimeService } from '@wix/ambassador-predictor-prime-service/http';
import { DEFAULT_ARTICLES_IDS, NUMBER_OF_FEATURED_ARTICLES, } from '../../common/constants';
export var PREDICTION_API_BASE_URL = '/_api/predictor-service';
var service = PredictorPrimeService(PREDICTION_API_BASE_URL).Predictor();
export var getPrediction = function () { return __awaiter(void 0, void 0, void 0, function () {
    var articles, output, error_1;
    var _a, _b, _c;
    return __generator(this, function (_d) {
        switch (_d.label) {
            case 0:
                _d.trys.push([0, 2, , 3]);
                return [4 /*yield*/, service().invokeByModel({
                        modelId: 'pars',
                        features: {
                            values: {
                                audience: { strVal: 'dashboard-single-site' },
                            },
                        },
                    })];
            case 1:
                output = (_d.sent()).output;
                articles = (_c = (_b = (_a = output === null || output === void 0 ? void 0 : output.values) === null || _a === void 0 ? void 0 : _a.recommended_article_ids) === null || _b === void 0 ? void 0 : _b.propertyValList) === null || _c === void 0 ? void 0 : _c.value;
                return [3 /*break*/, 3];
            case 2:
                error_1 = _d.sent();
                // TODO: handle Errors
                console.log(error_1);
                return [3 /*break*/, 3];
            case 3: return [2 /*return*/, articles
                    ? articles
                        .map(function (article) { return article.strVal; })
                        .slice(0, NUMBER_OF_FEATURED_ARTICLES)
                    : DEFAULT_ARTICLES_IDS];
        }
    });
}); };
