// @wix/dashboard-answers-widget/dist/es/src/services/IframeLoadingSequence.js
var IFRAME_LOAD = 'IFRAME_LOAD';
var UNIQUE_SCRIPT_LOAD = 'UNIQUE_SCRIPT_LOAD';
var UNIQUE_CSS_LOAD = 'UNIQUE_CSS_LOAD';
var IframeLoadingSequence = /** @class */ (function () {
    function IframeLoadingSequence(debug) {
        if (debug === void 0) { debug = false; }
        this.debug = debug;
        this.resolves = {};
        this.promises = {};
        this.add(IFRAME_LOAD);
        this.add(UNIQUE_SCRIPT_LOAD);
        this.add(UNIQUE_CSS_LOAD);
    }
    IframeLoadingSequence.prototype.add = function (name) {
        var _this = this;
        this.promises[name] = new Promise(function (resolve) { return (_this.resolves[name] = resolve); });
        return this;
    };
    IframeLoadingSequence.prototype.resolve = function (name) {
        if (this.debug) {
            console.log("iframeLoadingSequence: " + name + " was resolved");
        }
        this.resolves[name] && this.resolves[name]();
        return this;
    };
    IframeLoadingSequence.prototype.reportOnIframeLoad = function () {
        return this.resolve(IFRAME_LOAD);
    };
    IframeLoadingSequence.prototype.reportUniqueScriptLoad = function () {
        return this.resolve(UNIQUE_SCRIPT_LOAD);
    };
    IframeLoadingSequence.prototype.reportUniqueCssLoad = function () {
        return this.resolve(UNIQUE_CSS_LOAD);
    };
    IframeLoadingSequence.prototype.whenUniqueScriptLoaded = function () {
        return this.promises[UNIQUE_SCRIPT_LOAD];
    };
    IframeLoadingSequence.prototype.whenAll = function () {
        var _this = this;
        return Promise.all(Object.keys(this.promises).map(function (key) { return _this.promises[key]; }));
    };
    return IframeLoadingSequence;
}());
export { IframeLoadingSequence };
