// @wix/account-manager-api/dist/src/bi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.viewFinishedLoading = exports.viewStartLoading = exports.notifyViewFinishedLoading = exports.notifyViewStartLoading = exports.VIEW_FINISHED_LOADING = exports.VIEW_START_LOADING = void 0;
var react_module_container_1 = require("react-module-container");
exports.VIEW_START_LOADING = 'accountManager.viewStartLoading';
exports.VIEW_FINISHED_LOADING = 'accountManager.viewFinishedLoading';
/**
 * Call when a your application page component starts loading or loads a new view. (In the future may show a loader)
 *
 * @doc API
 */
exports.notifyViewStartLoading = function () { return react_module_container_1.ModuleRegistry.notifyListeners(exports.VIEW_START_LOADING); };
/**
 * Call when a your application page component is finished loading. (In the future may remove a loader)
 *
 * @doc API
 */
exports.notifyViewFinishedLoading = function () { return react_module_container_1.ModuleRegistry.notifyListeners(exports.VIEW_FINISHED_LOADING); };
// Backwards compatibility
exports.viewStartLoading = exports.notifyViewStartLoading;
exports.viewFinishedLoading = exports.notifyViewFinishedLoading;
