// @wix/account-manager-api/dist/src/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageComponentId = exports.getPageLinkComponent = exports.getPageUrl = exports.navigateTo = exports.PAGE_LINK_COMPONENT = exports.GET_PAGE_URL = exports.NAVIGATE_TO = void 0;
var react_module_container_1 = require("react-module-container");
exports.NAVIGATE_TO = 'accountManager.navigateTo';
exports.GET_PAGE_URL = 'accountManager.getPageUrl';
exports.PAGE_LINK_COMPONENT = 'accountManager.pageLinkComponent';
exports.navigateTo = function (navigateToPageConfig) { return react_module_container_1.ModuleRegistry.invoke(exports.NAVIGATE_TO, navigateToPageConfig); };
exports.getPageUrl = function (navigateToPageConfig) { return react_module_container_1.ModuleRegistry.invoke(exports.GET_PAGE_URL, navigateToPageConfig); };
exports.getPageLinkComponent = function () { return react_module_container_1.ModuleRegistry.component(exports.PAGE_LINK_COMPONENT); };
var PageComponentId;
(function (PageComponentId) {
    PageComponentId["InspiringWebsite"] = "account-manager-explore-inspiring-websites";
    PageComponentId["UpdatesReleases"] = "account-manager-explore-updates-releases";
    PageComponentId["Ascend"] = "account-manager-explore-ascend";
    PageComponentId["AscendNew"] = "account-manager-explore-ascend-new";
    PageComponentId["SiteScanner"] = "account-manager-explore-site-scanner";
    PageComponentId["Domains"] = "account-manager-domains";
    PageComponentId["Vouchers"] = "account-manager-vouchers";
    PageComponentId["AccountSettings"] = "account-settings-lazy-component";
    PageComponentId["Team"] = "account-team-app-lazy-component-id";
    PageComponentId["Mailboxes"] = "account-mailboxes-lazy-component";
    PageComponentId["Subscriptions"] = "account-subscriptions-lazy-component";
    PageComponentId["PaymentMethods"] = "account-payment-methods-lazy-component";
    PageComponentId["WixPayments"] = "wix-payments-accounts-lazy-component-id";
    PageComponentId["Billing"] = "account-billing-lazy-component";
    PageComponentId["HireAPartner"] = "user-projects-lazy-component-id";
})(PageComponentId = exports.PageComponentId || (exports.PageComponentId = {}));
