// @wix/analytics-ng-common/dist/es/src/TimeFrames/DaysTimeFrame.js
import { __extends } from "tslib";
import { TimeFrame } from './TimeFrame';
var DaysTimeFrame = /** @class */ (function (_super) {
    __extends(DaysTimeFrame, _super);
    function DaysTimeFrame(numberOfDays) {
        var _this = _super.call(this) || this;
        _this._numberOfDays = numberOfDays;
        return _this;
    }
    DaysTimeFrame.prototype.getRequiredDataLength = function () {
        return this._numberOfDays;
    };
    DaysTimeFrame.prototype.getDateInterval = function (queryResult, options) {
        var data = this.getQueryResultData(queryResult);
        var todayDate = this.getTodayDate(queryResult, options);
        var startDate = todayDate.subtractDays(this._numberOfDays - 1);
        return {
            startDate: startDate.toDateString(),
            numberOfDays: this._numberOfDays,
        };
    };
    DaysTimeFrame.prototype.getPreviousDateInterval = function (queryResult, options) {
        var todayDate = this.getTodayDate(queryResult, options);
        var startDate = todayDate.subtractDays(this._numberOfDays * 2 - 1);
        return {
            startDate: startDate.toDateString(),
            numberOfDays: this._numberOfDays,
        };
    };
    return DaysTimeFrame;
}(TimeFrame));
export { DaysTimeFrame };
