// @wix/analytics-ng-common/dist/es/src/TimeFrames/ThisWeekTimeframe.js
import { __extends } from "tslib";
import { TimeFrame } from './TimeFrame';
var ThisWeekTimeframe = /** @class */ (function (_super) {
    __extends(ThisWeekTimeframe, _super);
    function ThisWeekTimeframe() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ThisWeekTimeframe.prototype.getRequiredDataLength = function () {
        return 7;
    };
    ThisWeekTimeframe.prototype.getDateInterval = function (queryResult, options) {
        var today = this.getTodayDate(queryResult, options);
        var startOfWeek = today.getStartOfWeek();
        var numberOfDays = today.differenceInDays(startOfWeek);
        return {
            startDate: startOfWeek.toDateString(),
            numberOfDays: numberOfDays,
        };
    };
    ThisWeekTimeframe.prototype.getPreviousDateInterval = function (queryResult, options) {
        var today = this.getTodayDate(queryResult, options);
        var lastWeek = today.subtractWeeks(1);
        var startOfWeek = lastWeek.getStartOfWeek();
        var numberOfDays = lastWeek.differenceInDays(startOfWeek);
        return {
            startDate: startOfWeek.toDateString(),
            numberOfDays: numberOfDays,
        };
    };
    return ThisWeekTimeframe;
}(TimeFrame));
export { ThisWeekTimeframe };
