// @wix/analytics-ng-common/dist/es/src/TimeFrames/ThisMonthTimeframe.js
import { __extends } from "tslib";
import { TimeFrame } from './TimeFrame';
var ThisMonthTimeframe = /** @class */ (function (_super) {
    __extends(ThisMonthTimeframe, _super);
    function ThisMonthTimeframe() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ThisMonthTimeframe.prototype.getRequiredDataLength = function () {
        return 30;
    };
    ThisMonthTimeframe.prototype.getDateInterval = function (queryResult, options) {
        var today = this.getTodayDate(queryResult, options);
        var startOfMonth = today.getStartOfTheMonth();
        var numberOfDays = today.differenceInDays(startOfMonth);
        return {
            startDate: startOfMonth.toDateString(),
            numberOfDays: numberOfDays,
        };
    };
    ThisMonthTimeframe.prototype.getPreviousDateInterval = function (queryResult, options) {
        var today = this.getTodayDate(queryResult, options);
        // This already takes care of the difference in the length of the month.
        // For example, if March ends in the 31 then the -1 of that day will result in the 29th of Feb (since it doesn't have a 31 or a 30th)
        var lastMonth = today.subtractMonths(1);
        var lastStartOfMonth = lastMonth.getStartOfTheMonth();
        var numberOfDays = lastMonth.differenceInDays(lastStartOfMonth);
        return {
            startDate: lastStartOfMonth.toDateString(),
            numberOfDays: numberOfDays,
        };
    };
    return ThisMonthTimeframe;
}(TimeFrame));
export { ThisMonthTimeframe };
