// @wix/my-account-header/dist/es/src/client/components/Header/ADI/AdiHeader.js
import { __extends } from "tslib";
import * as React from 'react';
import * as css from './AdiHeader.scss';
import { ModuleRegistry } from 'react-module-container';
import Help from '../../Help/Help';
import { withTranslation } from '@wix/wix-i18n-config';
import { ON_CLOSE_IN_EDITOR } from '@wix/business-manager-api';
import { withHeaderProvider } from '../../HeaderProvider/HeaderProvider';
import { Text } from 'wix-style-react';
import { dataHooks } from '../../../constants/dataHooks';
var AdiHeaderComponentClass = /** @class */ (function (_super) {
    __extends(AdiHeaderComponentClass, _super);
    function AdiHeaderComponentClass(props) {
        var _this = _super.call(this, props) || this;
        _this.notifyAndClose = function (event) { return _this.notifyAndCloseIfNotPrevented(event); };
        return _this;
    }
    AdiHeaderComponentClass.prototype.notifyAndCloseIfNotPrevented = function (event) {
        ModuleRegistry.notifyListeners(ON_CLOSE_IN_EDITOR, { event: event });
        if (!event.defaultPrevented) {
            this.closeBizMgr();
        }
    };
    /**
     * MEMO: if there will be more postMessages to send to the ADI - extract to a SDK
     */
    AdiHeaderComponentClass.prototype.closeBizMgr = function () {
        parent.postMessage('{"intent":"TPA2", "action":"CLOSE", "type":"closeWindow", "namespace": "Wix", "data": {}}', '*');
    };
    AdiHeaderComponentClass.prototype.render = function () {
        return React.createElement("header", { className: css.header, "data-hook": dataHooks.Header },
            React.createElement(Text, { weight: "bold", size: "small", dataHook: "header-title" }, this.props.t('header.adi-title')),
            React.createElement("div", { className: css.actionsWrapper },
                React.createElement(Help, { containerClassName: css.helpButton, iconClassName: css.helpIconAdi, locale: this.props.locale }),
                React.createElement("a", { "data-hook": "close-from-adi", onClick: this.notifyAndClose, className: css.button },
                    React.createElement("div", { className: css.closeButtonContent }))));
    };
    return AdiHeaderComponentClass;
}(React.Component));
export var AdiHeaderComponent = withTranslation()(AdiHeaderComponentClass);
export var AdiHeader = withHeaderProvider(AdiHeaderComponent, false);
export default AdiHeader;
