// @wix/yoshi-flow-bm-runtime/build/esm/createLazyComponent.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import React, { Suspense, useState } from 'react';
import { ReactLoadableComponent } from 'react-module-container';
export function createLazyComponent(requireableBMComponent, _a) {
    var _this = this;
    var flowAPIDelegate = _a.flowAPIDelegate, context = _a.context;
    var enableTranslationsHMR = context.enableTranslationsHMR;
    var component = requireableBMComponent.component, loadReactComponent = requireableBMComponent.loadReactComponent, filesHook = requireableBMComponent.filesHook, resolveHook = requireableBMComponent.resolveHook;
    var maybeTranslationsModel = component.maybeTranslationsModel;
    var createComponentHooksFlowAPI = function (ownProps) {
        var fedopsLogger = flowAPIDelegate.createFedopsLogger(component.componentId);
        var maybeComponentTranslationsAPI = maybeTranslationsModel
            ? flowAPIDelegate.createComponentTranslationsAPI(maybeTranslationsModel)
            : undefined;
        var componentHooksFlowAPI = flowAPIDelegate.getComponentHooksFlowAPI(component, {
            fedopsLogger: fedopsLogger,
            maybeComponentTranslationsAPI: maybeComponentTranslationsAPI,
            ownProps: ownProps,
        });
        return {
            componentHooksFlowAPI: componentHooksFlowAPI,
            fedopsLogger: fedopsLogger,
            maybeComponentTranslationsAPI: maybeComponentTranslationsAPI,
        };
    };
    return function (props) {
        var _a = __read(useState(), 2), error = _a[0], setError = _a[1];
        var _b = __read(useState(function () { return createComponentHooksFlowAPI(props); }), 1), _c = _b[0], componentHooksFlowAPI = _c.componentHooksFlowAPI, fedopsLogger = _c.fedopsLogger, maybeComponentTranslationsAPI = _c.maybeComponentTranslationsAPI;
        var _d = __read(useState(function () {
            return ReactLoadableComponent(component.componentId, function () { return __awaiter(_this, void 0, void 0, function () {
                var resolveDataPromise, reactComponentPromise, componentFlowAPIPromise, _a, resolvedData, reactComponent, _componentFlowAPI, e_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _b.trys.push([0, 2, , 3]);
                            // do not pre-fetch when translations are bundled
                            if (!enableTranslationsHMR) {
                                flowAPIDelegate.prefetchTranslations(maybeComponentTranslationsAPI);
                            }
                            flowAPIDelegate.prefetchExperiments();
                            resolveDataPromise = resolveHook === null || resolveHook === void 0 ? void 0 : resolveHook.call(flowAPIDelegate.bmModule, componentHooksFlowAPI);
                            reactComponentPromise = loadReactComponent();
                            componentFlowAPIPromise = flowAPIDelegate.createBMComponentAPI(component, {
                                fedopsLogger: fedopsLogger,
                                componentTranslationsAPI: maybeComponentTranslationsAPI,
                            });
                            return [4 /*yield*/, Promise.all([
                                    resolveDataPromise,
                                    reactComponentPromise,
                                    componentFlowAPIPromise,
                                ])];
                        case 1:
                            _a = __read.apply(void 0, [_b.sent(), 3]), resolvedData = _a[0], reactComponent = _a[1], _componentFlowAPI = _a[2];
                            return [2 /*return*/, __assign(__assign({ default: reactComponent }, resolvedData), { _componentFlowAPI: _componentFlowAPI })];
                        case 2:
                            e_1 = _b.sent();
                            setError(e_1);
                            throw e_1;
                        case 3: return [2 /*return*/];
                    }
                });
            }); }, filesHook === null || filesHook === void 0 ? void 0 : filesHook.call(flowAPIDelegate.bmModule, componentHooksFlowAPI));
        }), 1), Component = _d[0];
        if (error) {
            throw error;
        }
        return (React.createElement(Suspense, { fallback: null },
            React.createElement(Component, __assign({}, props))));
    };
}
