// @wix/my-account-header/dist/es/src/client/components/Inbox/InboxHeaderWidgetLazyComponent.container.js
import { __extends } from "tslib";
import { ReactLazyComponent } from 'react-module-container';
import { connect } from 'react-redux';
import { create as fedopsLoggerCreate } from '@wix/fedops-logger';
import { withErrorBoundary } from '@wix/business-manager-common-utils';
import { getLanguage } from '../../selectors/currentUserSelectors';
var InboxHeaderWidgetLazyComponent = /** @class */ (function (_super) {
    __extends(InboxHeaderWidgetLazyComponent, _super);
    function InboxHeaderWidgetLazyComponent(props) {
        var _this = this;
        var inboxHeaderWidgetStaticsUrl = props.inboxHeaderWidgetStaticsUrl, isDebug = props.isDebug;
        var manifest = {
            files: [
                inboxHeaderWidgetStaticsUrl + "module-registry-components" + (isDebug ? '' : '.min') + ".css",
                inboxHeaderWidgetStaticsUrl + "module-registry-components.bundle" + (isDebug ? '' : '.min') + ".js"
            ],
            resolve: function () {
                fedopsLoggerCreate('notifications-widget-component', { isWidgetLogger: true }).appLoadStarted();
            },
            component: 'chat.InboxHeaderWidgetComponent'
        };
        _this = _super.call(this, props, manifest) || this;
        return _this;
    }
    return InboxHeaderWidgetLazyComponent;
}(ReactLazyComponent));
export { InboxHeaderWidgetLazyComponent };
var mapStateToProps = function (state) { return ({
    inboxHeaderWidgetStaticsUrl: state.topology.inboxHeaderWidgetStaticsUrl,
    isDebug: state.isDebug,
    experiments: state.experiments,
    locale: getLanguage(state),
}); };
export default withErrorBoundary()(connect(mapStateToProps)(InboxHeaderWidgetLazyComponent));
