// @wix/ambassador-crm-financial-settings/http.js
/* eslint-disable */
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrmFinancialSettings = void 0;
var http_1 = require("@wix/ambassador/runtime/http");
var CrmFinancialSettings = function (baseUrl, httpClientConfig) {
    var httpClient = new http_1.AmbassadorHTTPClient(baseUrl, httpClientConfig);
    return {
        CustomFieldDefinitionsService: function () {
            return function (headers) {
                var service = {
                    createCustomFieldDefinition: function (createCustomFieldDefinitionRequest) {
                        return httpClient.invoke('POST', '/v1/custom-field-definitions', createCustomFieldDefinitionRequest, headers);
                    },
                    deleteCustomFieldDefinition: function (deleteCustomFieldDefinitionRequest) {
                        return httpClient.invoke('DELETE', '/v1/custom-field-definitions/{key}', deleteCustomFieldDefinitionRequest, headers);
                    },
                    getCustomFieldDefinition: function (getCustomFieldDefinitionRequest) {
                        return httpClient.invoke('GET', '/v1/custom-field-definitions/{key}', getCustomFieldDefinitionRequest, headers);
                    },
                    queryCustomFieldDefinitions: function (queryCustomFieldDefinitionsRequest) {
                        return httpClient.invoke('POST', '/v1/custom-field-definitions/query', queryCustomFieldDefinitionsRequest, headers);
                    },
                    updateCustomFieldDefinition: function (updateCustomFieldDefinitionRequest) {
                        return httpClient.invoke('PATCH', '/v1/custom-field-definitions/{custom_field_definition.key}', updateCustomFieldDefinitionRequest, headers);
                    },
                };
                service.createCustomFieldDefinition.__http_method = 'POST';
                service.createCustomFieldDefinition.__http_url = '/v1/custom-field-definitions';
                service.deleteCustomFieldDefinition.__http_method = 'DELETE';
                service.deleteCustomFieldDefinition.__http_url = '/v1/custom-field-definitions/{key}';
                service.getCustomFieldDefinition.__http_method = 'GET';
                service.getCustomFieldDefinition.__http_url = '/v1/custom-field-definitions/{key}';
                service.queryCustomFieldDefinitions.__http_method = 'POST';
                service.queryCustomFieldDefinitions.__http_url = '/v1/custom-field-definitions/query';
                service.updateCustomFieldDefinition.__http_method = 'PATCH';
                service.updateCustomFieldDefinition.__http_url = '/v1/custom-field-definitions/{custom_field_definition.key}';
                return service;
            };
        },
        CustomFieldTypesService: function () {
            return function (headers) {
                var service = {
                    createCustomFieldType: function (createCustomFieldTypeRequest) {
                        return httpClient.invoke('POST', '/v1/custom-field-types', createCustomFieldTypeRequest, headers);
                    },
                    deleteCustomFieldType: function (deleteCustomFieldTypeRequest) {
                        return httpClient.invoke('DELETE', '/v1/custom-field-types/{key}', deleteCustomFieldTypeRequest, headers);
                    },
                    getCustomFieldType: function (getCustomFieldTypeRequest) {
                        return httpClient.invoke('GET', '/v1/custom-field-types/{key}', getCustomFieldTypeRequest, headers);
                    },
                    queryCustomFieldTypes: function (queryCustomFieldTypesRequest) {
                        return httpClient.invoke('POST', '/v1/custom-field-types/query', queryCustomFieldTypesRequest, headers);
                    },
                    updateCustomFieldType: function (updateCustomFieldTypeRequest) {
                        return httpClient.invoke('PATCH', '/v1/custom-field-types/{custom_field_type.key}', updateCustomFieldTypeRequest, headers);
                    },
                };
                service.createCustomFieldType.__http_method = 'POST';
                service.createCustomFieldType.__http_url = '/v1/custom-field-types';
                service.deleteCustomFieldType.__http_method = 'DELETE';
                service.deleteCustomFieldType.__http_url = '/v1/custom-field-types/{key}';
                service.getCustomFieldType.__http_method = 'GET';
                service.getCustomFieldType.__http_url = '/v1/custom-field-types/{key}';
                service.queryCustomFieldTypes.__http_method = 'POST';
                service.queryCustomFieldTypes.__http_url = '/v1/custom-field-types/query';
                service.updateCustomFieldType.__http_method = 'PATCH';
                service.updateCustomFieldType.__http_url = '/v1/custom-field-types/{custom_field_type.key}';
                return service;
            };
        },
        DesignSettingsService: function () {
            return function (headers) {
                var service = {
                    getDesignSettings: function (getDesignSettingsRequest) {
                        return httpClient.invoke('GET', '/v1/design-settings', getDesignSettingsRequest, headers);
                    },
                    updateDesignSettings: function (updateDesignSettingsRequest) {
                        return httpClient.invoke('POST', '/v1/design-settings', updateDesignSettingsRequest, headers);
                    },
                };
                service.getDesignSettings.__http_method = 'GET';
                service.getDesignSettings.__http_url = '/v1/design-settings';
                service.updateDesignSettings.__http_method = 'POST';
                service.updateDesignSettings.__http_url = '/v1/design-settings';
                return service;
            };
        },
        GeneralSettingsService: function () {
            return function (headers) {
                var service = {
                    getGeneralSettings: function (getGeneralSettingsRequest) {
                        return httpClient.invoke('GET', '/v1/general-settings', getGeneralSettingsRequest, headers);
                    },
                    updateGeneralSettings: function (updateGeneralSettingsRequest) {
                        return httpClient.invoke('PATCH', '/v1/general-settings', updateGeneralSettingsRequest, headers);
                    },
                };
                service.getGeneralSettings.__http_method = 'GET';
                service.getGeneralSettings.__http_url = '/v1/general-settings';
                service.updateGeneralSettings.__http_method = 'PATCH';
                service.updateGeneralSettings.__http_url = '/v1/general-settings';
                return service;
            };
        },
    };
};
exports.CrmFinancialSettings = CrmFinancialSettings;
exports.CrmFinancialSettings.__transport = 'http';
__exportStar(require("./types"), exports);
