// @wix/my-account-header/dist/es/src/client/components/UserProfilePicture/UserProfilePicture.js
import { __extends } from "tslib";
import * as React from 'react';
import * as css from './UserProfilePicture.scss';
export var DEFAULT_IMAGE_TIMEOUT = 200;
var UserProfilePicture = /** @class */ (function (_super) {
    __extends(UserProfilePicture, _super);
    function UserProfilePicture(props) {
        var _this = _super.call(this, props) || this;
        _this.handleImageLoad = function () {
            _this.setState({ showDefaultImage: false });
            clearTimeout(_this.defaultImageTimeout);
        };
        _this.handleImageError = function () {
            _this.setState({ showDefaultImage: true, hideImage: true });
        };
        _this.state = { showDefaultImage: false, hideImage: false };
        return _this;
    }
    UserProfilePicture.prototype.componentDidMount = function () {
        var _this = this;
        this.defaultImageTimeout = setTimeout(function () { return _this.setState({ showDefaultImage: true }); }, DEFAULT_IMAGE_TIMEOUT);
    };
    UserProfilePicture.prototype.componentWillUnmount = function () {
        clearTimeout(this.defaultImageTimeout);
    };
    UserProfilePicture.prototype.render = function () {
        var profileImage = this.props.profileImage;
        var _a = this.state, showDefaultImage = _a.showDefaultImage, hideImage = _a.hideImage;
        return (React.createElement("div", { className: css.container },
            showDefaultImage && React.createElement("div", { "data-hook": "default-profile-image", className: css.defaultImage }),
            !hideImage && React.createElement("img", { "data-hook": "user-profile-image", src: profileImage, onLoad: this.handleImageLoad, onError: this.handleImageError, className: css.image })));
    };
    return UserProfilePicture;
}(React.Component));
export { UserProfilePicture };
