// @wix/my-account-header/dist/es/src/client/components/AccountMenu/AccountMenuHeader.js
import { __extends } from "tslib";
import * as React from 'react';
import { connect } from 'react-redux';
import { withTranslation } from '@wix/wix-i18n-config';
import { Text, Box } from 'wix-style-react';
import * as css from './AccountMenu.scss';
import { getAccountName } from '../../selectors/accountSelectors';
import UserProfilePicture from '../UserProfilePicture/UserProfilePicture.container';
import AccountMenuHeaderUserName from './AccountMenuHeaderUserName';
import { getDisplayName } from '../../selectors/currentUserSelectors';
var AccountMenuHeader = /** @class */ (function (_super) {
    __extends(AccountMenuHeader, _super);
    function AccountMenuHeader() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AccountMenuHeader.prototype.render = function () {
        var _a = this.props, accountName = _a.accountName, displayName = _a.displayName;
        return (React.createElement("div", { "data-hook": "account-menu-header", className: css.header },
            React.createElement(ProfilePicture, null),
            React.createElement(UserDetails, { userName: displayName, accountName: accountName })));
    };
    return AccountMenuHeader;
}(React.Component));
var mapStateToProps = function (state) {
    return {
        accountName: getAccountName(state),
        displayName: getDisplayName(state),
    };
};
export default withTranslation()(connect(mapStateToProps)(AccountMenuHeader));
function TeamName(_a) {
    var accountName = _a.accountName;
    return (React.createElement(Text, { dataHook: "team-name", size: "tiny", secondary: true }, accountName));
}
function UserDetails(_a) {
    var userName = _a.userName, accountName = _a.accountName;
    return (React.createElement(Box, { direction: "vertical", width: "232px" },
        React.createElement(AccountMenuHeaderUserName, { userName: userName }),
        accountName && React.createElement(TeamName, { accountName: accountName })));
}
function ProfilePicture() {
    return (React.createElement("div", { className: css.profilePicture },
        React.createElement(UserProfilePicture, null)));
}
