// @wix/my-account-header/dist/es/src/client/components/AccountMenu/MenuItems.js
var _a, _b;
import { __spreadArrays } from "tslib";
import { Permission } from '@wix/account-manager-common/dist/src';
import { PageComponentId } from '@wix/account-manager-api';
export var HeaderPermissions;
(function (HeaderPermissions) {
    HeaderPermissions["viewCreateSite"] = "HEADER.VIEW_CREATE_SITE_ON_HEADER";
})(HeaderPermissions || (HeaderPermissions = {}));
export var ACCOUNT_SETTINGS = 'accountSettings';
export var SUBSCRIPTIONS = 'subscriptions';
export var DOMAINS = 'domains';
export var MAILBOXES = 'mailboxes';
export var BILLING_HISTORY = 'billingHistory';
export var PAYMENT_METHODS = 'paymentMethods';
export var WIX_PAYMENTS_ACCOUNTS = 'wixPaymentsAccounts';
export var VOUCHERS = 'vouchers';
export var TEMPLATES = 'templates';
export var HELP_CENTER = 'helpCenter';
export var HIRE_AN_EXPERT = 'hireAnExpert';
export var HIRE_A_SPECIALIST = 'hireASpecialist';
export var SEPARATOR = 'separator';
export var LANGUAGES = 'languages';
export var isSeparator = function (item) { return item.key === SEPARATOR; };
export var MenuItems = function (_a) {
    var locale = _a.locale, experiments = _a.experiments, isEditorX = _a.isEditorX;
    return __spreadArrays([
        {
            key: ACCOUNT_SETTINGS,
            link: '/account/account-settings',
        },
        {
            key: SEPARATOR
        },
        {
            key: DOMAINS,
            link: '/wix/api/mpContainerStaticController',
        },
        {
            key: MAILBOXES,
            link: '/account/mailboxes',
        },
        {
            key: VOUCHERS,
            link: '/account/vouchers',
        },
        {
            key: BILLING_HISTORY,
            link: '/account/billing-history',
        },
        {
            key: SUBSCRIPTIONS,
            link: '/wix/api/PaymentConsole',
        },
        {
            key: PAYMENT_METHODS,
            link: '/account/payment-methods',
        }
    ], (experiments.enabled('specs.payments.EnableWixPaymentsAccountsInAccountManager')
        ? [
            {
                key: WIX_PAYMENTS_ACCOUNTS,
                link: '/account/wix-payments',
            },
        ]
        : []), [
        { key: SEPARATOR },
        {
            key: TEMPLATES,
            link: '/create-site',
        },
        {
            key: HELP_CENTER,
            link: "https://support.wix.com/" + locale,
        },
        {
            key: isEditorX && experiments.enabled('specs.wosbm.hireSpecialistInEditorX')
                ? HIRE_A_SPECIALIST : HIRE_AN_EXPERT,
            link: '/account/hire-a-partner',
        }
    ], (experiments.enabled('specs.wosbm.scrollableAccountMenu')
        ? []
        : [{ key: SEPARATOR }]));
};
export var permissionMap = (_a = {},
    _a[DOMAINS] = Permission.VIEW_DOMAINS,
    _a[MAILBOXES] = Permission.VIEW_MAILBOXES,
    _a[SUBSCRIPTIONS] = Permission.VIEW_SUBSCRIPTIONS,
    _a[BILLING_HISTORY] = Permission.VIEW_BILLING_HISTORY,
    _a[PAYMENT_METHODS] = Permission.VIEW_PAYMENT_METHODS,
    _a[WIX_PAYMENTS_ACCOUNTS] = Permission.VIEW_PAYMENT_METHODS,
    _a[VOUCHERS] = Permission.VIEW_VOUCHERS,
    _a[TEMPLATES] = HeaderPermissions.viewCreateSite,
    _a[HIRE_AN_EXPERT] = Permission.VIEW_HIRE_A_PARTNER,
    _a[HIRE_A_SPECIALIST] = Permission.VIEW_HIRE_A_PARTNER,
    _a);
export var accountManagerPageComponentIdMap = (_b = {},
    _b[ACCOUNT_SETTINGS] = PageComponentId.AccountSettings,
    _b[DOMAINS] = PageComponentId.Domains,
    _b[MAILBOXES] = PageComponentId.Mailboxes,
    _b[SUBSCRIPTIONS] = PageComponentId.Subscriptions,
    _b[BILLING_HISTORY] = PageComponentId.Billing,
    _b[PAYMENT_METHODS] = PageComponentId.PaymentMethods,
    _b[WIX_PAYMENTS_ACCOUNTS] = PageComponentId.WixPayments,
    _b[VOUCHERS] = PageComponentId.Vouchers,
    _b[HIRE_AN_EXPERT] = PageComponentId.HireAPartner,
    _b[HIRE_A_SPECIALIST] = PageComponentId.HireAPartner,
    _b);
