// @wix/wix-payments-server/dist/src/routes/api/accounts/accounts-endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountsEndpoints = void 0;
var AccountsEndpoints;
(function (AccountsEndpoints) {
    AccountsEndpoints["API_ACCOUNT"] = "api/accounts";
    AccountsEndpoints["UPDATE_BANK_ACCOUNT_DETAILS"] = "api/accounts/bank-account-details";
    AccountsEndpoints["UPDATE_INSTALLMENTS"] = "api/accounts/installments";
    AccountsEndpoints["UPDATE_ANTICIPATION"] = "api/accounts/anticipation";
    AccountsEndpoints["GET_ACTIVATION_STATUS"] = "api/accounts/activationStatus";
    AccountsEndpoints["ADD_PROFILE_DOCUMENTS"] = "api/accounts/documents";
    AccountsEndpoints["UPDATE_SOFT_DESCRIPTOR"] = "api/accounts/update-soft-descriptor";
    AccountsEndpoints["DELETE_SHAREHOLDER"] = "api/accounts/delete-shareholder";
    AccountsEndpoints["SAVE_AS_DRAFT"] = "api/accounts/save-as-draft";
    AccountsEndpoints["SAVE_AS_DRAFT_AND_CREATE_PROFILE"] = "api/accounts/save-as-draft-and-create-profile";
    AccountsEndpoints["GET_PAYOUTS_SETTINGS"] = "api/accounts/payouts-settings";
    AccountsEndpoints["UPDATE_PAYOUTS_SETTINGS"] = "api/accounts/update-payouts-settings";
    AccountsEndpoints["COMPLAIN_ABOUT_RESERVATION"] = "api/accounts/complain-about-reservation";
})(AccountsEndpoints = exports.AccountsEndpoints || (exports.AccountsEndpoints = {}));
