// @wix/fedops-logger/dist/es/src/reporter/reporter.js
'use strict';

exports.__esModule = true;

var _publishers = require('../publishers/publishers');

var _conf = require('../conf');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Reporter = function () {
  function Reporter() {
    var baseUrl = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _conf.BI_BASE_URL;
    var performanceBaseUrl = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _conf.PERFORMANCE_BASE_URL;
    var publisher = arguments[2];

    _classCallCheck(this, Reporter);

    this.baseUrl = baseUrl;
    this.performanceBaseUrl = performanceBaseUrl;
    this.mainUrl = this.baseUrl;
    this.publisher = publisher;

    if (!this.publisher) {
      this.publisher = (0, _publishers.getPublisherByType)();
    }
  }

  Reporter.prototype.getBaseUrl = function getBaseUrl() {
    return this.baseUrl;
  };

  Reporter.prototype.getPublisher = function getPublisher() {
    return this.publisher;
  };

  Reporter.prototype.useBaseUrl = function useBaseUrl() {
    this.mainUrl = this.baseUrl;
  };

  Reporter.prototype.usePerformanceUrl = function usePerformanceUrl() {
    this.mainUrl = this.performanceBaseUrl;
  };

  Reporter.prototype.report = function report(data) {
    if (!data) {
      return null;
    }

    return this.publisher(this.mainUrl, data);
  };

  return Reporter;
}();

exports.default = Reporter;
