// @wix/my-account-header/dist/es/src/client/components/AcademyX/AcademyXMenu.js
import { __spreadArrays } from "tslib";
import * as React from 'react';
import { withTranslation } from '@wix/wix-i18n-config';
import { dropdownLinkClasses } from '../DropdownMenu/DropdownMenu';
import * as css from '../DropdownMenu/DropdownMenu.scss';
import { connect } from 'react-redux';
import { experimentsSelectors } from '@wix/wix-experiments-redux';
import { isNotCoBranded } from '../../selectors/coBrandingSelector';
export var ACADEMY_X_MENU_ITEM = {
    key: 'academyXMenu.academy.X',
    link: 'https://www.editorx.com/academy?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-academy'
};
export var WEBINARS_MENU_ITEM = {
    key: 'academyXMenu.webinars',
    link: 'https://www.editorx.com/academy/webinars?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-webinars'
};
export var FORUM_MENU_ITEM = {
    key: 'academyXMenu.forum',
    link: 'https://www.editorxcommunity.com/?referralAdditionalInfo=dashboard_menu/',
    dataHook: 'academy-x-menu-forum'
};
var HIRE_SPECIALIST_MENU_ITEM = {
    key: 'academyXMenu.hire.a.specialist',
    link: 'https://www.wix.com/marketplace/dynamic-brief?entry=editor-x&serviceType=1&innerType=1&origin=editorx-os-header-resources',
    dataHook: 'academy-x-menu-hire-a-specialist',
};
export var INSPIRATION_MENU_ITEMS = {
    key: 'academyXMenu.inspiration',
    link: 'https://www.editorx.com/inspiration?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-inspiration'
};
export var FEATURES_MENU_ITEM = {
    key: 'academyXMenu.features',
    link: 'https://www.editorx.com/features?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-features'
};
export var DESIGN_FEATURES_MENU_ITEM = {
    key: 'academyXMenu.features',
    link: 'https://www.editorx.com/features/design?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-features'
};
export var COMING_SOON_MENU_ITEM = {
    key: 'academyXMenu.coming_soon',
    link: 'https://www.editorx.com/features/coming-soon?referralAdditionalInfo=dashboard_menu',
    dataHook: 'academy-x-menu-coming-soon'
};
var AcademyXMenu = function (_a) {
    var shouldShowDesignFeatuers = _a.shouldShowDesignFeatuers, shouldShowHireSpecialist = _a.shouldShowHireSpecialist, t = _a.t;
    var menuItems = __spreadArrays([
        ACADEMY_X_MENU_ITEM,
        WEBINARS_MENU_ITEM,
        FORUM_MENU_ITEM
    ], (shouldShowHireSpecialist ? [HIRE_SPECIALIST_MENU_ITEM] : []), [
        INSPIRATION_MENU_ITEMS
    ], (shouldShowDesignFeatuers ? [DESIGN_FEATURES_MENU_ITEM] : [FEATURES_MENU_ITEM]), [
        COMING_SOON_MENU_ITEM
    ]);
    return (React.createElement("div", { className: css.container, "data-hook": "academy-x-header-menu" }, menuItems.map(function (_a, index) {
        var key = _a.key, link = _a.link, dataHook = _a.dataHook;
        return (React.createElement("a", { "data-hook": dataHook, className: dropdownLinkClasses, href: link, key: index, target: "_blank" }, t(key)));
    })));
};
var mapStateToProps = function (state) { return ({
    shouldShowDesignFeatuers: experimentsSelectors(state).enabled('specs.wosbm.UpdateFeaturesEntryInHeader'),
    shouldShowHireSpecialist: experimentsSelectors(state)
        .enabled('specs.wosbm.hireSpecialistInEditorX')
        && isNotCoBranded(state),
}); };
export default connect(mapStateToProps)(withTranslation()(AcademyXMenu));
