// @wix/quick-actions-dropdown/dist/es/src/consts/data-hook.js
export var TRANSFERABLE_SITE_CONTENT = 'transferable-site-content';
export var newQuickActionsPopoverDataHook = 'quick-actions-popover';
export var oldQuickActionsDropdownDataHook = 'dropdown-layout-options';
export var editSiteActionDataHook = 'quick-actions-action-edit-site';
export var getFeedbackActionDataHook = 'quick-actions-action-get-feedback';
export var previewSiteActionDataHook = 'quick-actions-action-preview';
export var viewLiveSiteActionDataHook = 'quick-actions-action-view-site';
export var renameSiteActionDataHook = 'quick-actions-action-rename';
export var duplicateSiteActionDataHook = 'quick-actions-action-duplicate';
export var transferSiteActionDataHook = 'quick-actions-action-transfer';
export var inviteContributorActionDataHook = 'quick-actions-action-invite-contributor';
export var moveToTrashActionDataHook = 'quick-actions-action-move-to-trash';
export var createNewSiteActionDataHook = 'quick-actions-action-create-new-site';
export var renameSiteLoader = 'rename-site-loader';
export var renameComponentFunctionalLayout = 'rename-component-functional-layout';
export var renameComponentSiteDisplayNameInput = 'rename-component-site-display-name-input';
export var renameComponentDomainInput = 'rename-component-domain-input';
export var duplicateDialogFunctionaLayout = 'duplicate-dialog-functional-layout';
export var transferSiteInitialDialog = 'transfer-site-initial-dialog';
export var moveToTrashDialogFunctionalLayout = 'move-to-trash-dialog-functional-layout';
