// @wix/quick-actions-dropdown/dist/es/src/consts/permissions.js
export var PERMISSIONS;
(function (PERMISSIONS) {
    PERMISSIONS["OPEN_DASHBOARD"] = "my-account.open-dashboard";
    PERMISSIONS["RENAME_SITE"] = "my-account.rename-site";
    PERMISSIONS["DUPLICATE_SITE"] = "my-account.duplicate-site";
    PERMISSIONS["DELETE_SITE"] = "my-account.delete-site";
    PERMISSIONS["INVITE_CONTRIBUTOR"] = "my-account.invite-contributor";
    PERMISSIONS["CREATE_SITE"] = "site-actions.create-site";
    PERMISSIONS["MOVE_SITE_TO_FOLDER"] = "site-actions.move-site-to-folder";
    PERMISSIONS["TRANSFER_SITE"] = "site-actions.transfer-site";
    PERMISSIONS["REMOVE_SITE_FROM_TRASH"] = "SITE_ACTION.REMOVE_FROM_TRASH";
    PERMISSIONS["RESTORE_SITE"] = "site-actions.restore-site";
    PERMISSIONS["EDIT_SITE"] = "html-editor.edit";
    PERMISSIONS["GET_FEEDBACK_DEPRECATED"] = "SITE_ACTION.GET_FEEDBACK";
    PERMISSIONS["GET_FEEDBACK"] = "SITE_ACTION.SITE_GET_FEEDBACK";
    PERMISSIONS["PREVIEW_SITE"] = "html-editor.preview";
    PERMISSIONS["CREATE_FOLDER"] = "folders.create";
})(PERMISSIONS || (PERMISSIONS = {}));
