// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/InviteContributor/invite-contributor-site-action.js
import * as React from 'react';
import UserAddSmall from 'wix-ui-icons-common/UserAddSmall';
import { ActionType, TOrigin } from '../../../../types';
import { PERMISSIONS } from '../../../../consts/permissions';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { BIEvents } from '../../../../consts/bi';
import { sendSiteActionsBI } from '../../../../services/BIService';
import { navigateTo, PageComponentId } from '@wix/business-manager-api';
var onClick = function (params) {
    var origin = params.origin, msid = params.quickActionsData.metaSiteId;
    sendSiteActionsBI({
        evid: BIEvents.quickActionClicked,
        name: 'quick_actions.invite_contributor',
        msid: msid,
        origin: origin
    });
    if (origin === TOrigin.Dashboard) {
        navigateTo({
            pageComponentId: PageComponentId.RolesAndPermissions,
            contextData: {
                referrer: origin,
                appState: 'invite'
            }
        });
    }
    else if (origin === TOrigin.mySites) {
        window.open("/dashboard/" + msid + "/roles-and-permissions/invite?referralInfo=" + origin);
    }
};
export var getInviteContributorSiteAction = function (experiments, origin) {
    var textKey = experiments['specs.wos1.useNewDashboard2020'] === 'true' && origin === TOrigin.Dashboard ?
        'quickActions.invite-people' : 'quickActions.invite-contributor';
    return {
        isPermitted: PermissionMethods.withPermissions(PERMISSIONS.INVITE_CONTRIBUTOR),
        actionName: 'invite-contributor',
        textKey: textKey,
        type: ActionType.click,
        icon: React.createElement(UserAddSmall, null),
        onActionClick: onClick,
    };
};
