// @wix/wix-experiments-redux/dist/src/ExperimentsScopeConductor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRedux = require('react-redux');

var _wixExperiments = require('@wix/wix-experiments');

var _wixExperiments2 = _interopRequireDefault(_wixExperiments);

var _addExperiments = require('./addExperiments');

var _addExperiments2 = _interopRequireDefault(_addExperiments);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ExperimentsScopeConductor = function (_Component) {
  (0, _inherits3.default)(ExperimentsScopeConductor, _Component);

  function ExperimentsScopeConductor(props) {
    (0, _classCallCheck3.default)(this, ExperimentsScopeConductor);

    var _this = (0, _possibleConstructorReturn3.default)(this, (ExperimentsScopeConductor.__proto__ || Object.getPrototypeOf(ExperimentsScopeConductor)).call(this, props));

    _this.state = { scopeConducted: false };
    return _this;
  }

  (0, _createClass3.default)(ExperimentsScopeConductor, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this2 = this;

      var experiments = new _wixExperiments2.default();
      experiments.load(this.props.scope).then(function () {
        _this2.props.addExperiments(experiments.all());
        _this2.setState({ scopeConducted: true });
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return this.state.scopeConducted ? _react.Children.only(this.props.children) : null;
    }
  }]);
  return ExperimentsScopeConductor;
}(_react.Component);

ExperimentsScopeConductor.propTypes = {
  scope: _propTypes2.default.string.isRequired,
  children: _propTypes2.default.element.isRequired,
  addExperiments: _propTypes2.default.func.isRequired
};


var mapStateToProps = function mapStateToProps(store, ownProps) {
  return { scope: ownProps.scope };
};
exports.default = (0, _reactRedux.connect)(mapStateToProps, { addExperiments: _addExperiments2.default })(ExperimentsScopeConductor);
