// @wix/dashboard-smart-answers-widget/dist/esm/services/bi-logger/connect-client-bi-logger.js
import webBiLogger from '@wix/web-bi-logger';
import initSchemaLogger from '@wix/bi-logger-ans-bi';
import { REFERRAL_URL, SOURCE_URL } from '../../common/constants';
export var getConnectClientLogger = function (uuid, source_name) {
    return initSchemaLogger(webBiLogger)().util.updateDefaults({
        src: 4,
        uuid: uuid,
        source_name: source_name,
    });
};
export var logFeaturedArticleClicked = function (biLogger, _a) {
    var feature_article_position = _a.feature_article_position, item_id = _a.item_id, lang = _a.lang;
    biLogger.clickOnFeaturedArticleInHeaderHelpWidget({
        feature_article_position: feature_article_position,
        item_id: item_id,
        lang: lang,
    });
};
export var logSerachNoResult = function (biLogger, _a) {
    var question = _a.question;
    return biLogger.gotNoAnswer({
        question: question,
        source_url: SOURCE_URL,
    });
};
export var logSearchCompleted = function (biLogger, _a) {
    var question = _a.question, first_answer = _a.first_answer, lang = _a.lang;
    return biLogger.askingAQuestion({
        source_url: SOURCE_URL,
        question: question,
        first_answer: first_answer,
        lang: lang,
    });
};
export var logSearchResultClicked = function (biLogger, _a) {
    var answer = _a.answer, question = _a.question, order = _a.order, item_url = _a.item_url, lang = _a.lang;
    return biLogger.clickOnAnswer({
        source_url: SOURCE_URL,
        question: question,
        answer: answer,
        order: order,
        item_url: item_url,
        lang: lang,
    });
};
export var logArticleClicked = function (biLogger, _a) {
    var item_id = _a.item_id, item_name = _a.item_name, item_url = _a.item_url, lang = _a.lang;
    return biLogger.pageView({
        item_id: item_id,
        item_name: item_name,
        referral_url: REFERRAL_URL,
        lang: lang,
        item_url: item_url,
    });
};
export var logFeaturedArticlesViewd = function (biLogger, _a) {
    var featured_articles_item_ids = _a.featured_articles_item_ids, featured_articles_lang = _a.featured_articles_lang;
    return biLogger.viewFeaturedArticlesInHeaderHelpWidget({
        featured_articles_item_ids: featured_articles_item_ids,
        featured_articles_lang: featured_articles_lang,
    });
};
