// @wix/wix-axios-config/dist/src/wix-axios-config.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _axiosInterceptorsLoader = _interopRequireDefault(require("./axios-interceptors-loader"));

var _axiosInterceptWixResult = _interopRequireDefault(require("./axios-intercept-wix-result"));

var _axiosInterceptBaseurl = _interopRequireDefault(require("./axios-intercept-baseurl"));

var _axiosInterceptMultilingual = _interopRequireDefault(require("./axios-intercept-multilingual"));

var _axiosInterceptArtifactid = _interopRequireDefault(require("./axios-intercept-artifactid"));

var _axiosInterceptSiteRevision = _interopRequireDefault(require("./axios-intercept-site-revision"));

var _axiosInterceptWixHeaders = _interopRequireDefault(require("./axios-intercept-wix-headers"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _defaults = {
  loadInterceptors: true
};

var _default = function _default(axios, _ref) {
  var baseURL = _ref.baseURL,
      _ref$loadInterceptors = _ref.loadInterceptors,
      loadInterceptors = _ref$loadInterceptors === void 0 ? _defaults.loadInterceptors : _ref$loadInterceptors,
      lang = _ref.lang,
      locale = _ref.locale,
      isPrimaryLanguage = _ref.isPrimaryLanguage,
      siteInstance = _ref.siteInstance,
      signedInstance = _ref.signedInstance,
      wixCodeApi = _ref.wixCodeApi,
      _ref$shouldAddArtifac = _ref.shouldAddArtifactIdOverride,
      shouldAddArtifactIdOverride = _ref$shouldAddArtifac === void 0 ? false : _ref$shouldAddArtifac;
  axios.defaults.baseURL = baseURL;
  return _axiosInterceptorsLoader["default"].loadIfNeeded(axios, {
    shouldLoadInterceptors: loadInterceptors,
    interceptors: [_axiosInterceptWixResult["default"], _axiosInterceptBaseurl["default"], _axiosInterceptMultilingual["default"].getInterceptor({
      lang: lang,
      siteInstance: siteInstance,
      locale: locale,
      isPrimaryLanguage: isPrimaryLanguage
    }), _axiosInterceptArtifactid["default"].getInterceptor({
      shouldAddArtifactIdOverride: shouldAddArtifactIdOverride
    }), _axiosInterceptSiteRevision["default"].getInterceptor({
      wixCodeApi: wixCodeApi
    }), _axiosInterceptWixHeaders["default"].getInterceptor({
      signedInstance: signedInstance
    })]
  });
};

exports["default"] = _default;
