// @wix/ambassador-contacts-app/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubmitOperation = exports.SubdivisionType = exports.SourceType = exports.SortOrder = exports.PhonePhoneTag = exports.JobType = exports.JobStatus = exports.ImageProvider = exports.GetContactResponseType = exports.GetBy = exports.EmailEmailTag = exports.ContactsFacetType = exports.ContactSourceType = exports.ContactPhonePhoneTag = exports.ContactFieldSet = exports.ContactEmailEmailTag = exports.ContactAddressAddressTag = exports.ContactActivityType = exports.BlockType = exports.AddressAddressTag = exports.Action = void 0;
var Action;
(function (Action) {
    Action["UNKNOWN"] = "UNKNOWN";
    Action["UPDATED"] = "UPDATED";
    Action["CREATED"] = "CREATED";
})(Action = exports.Action || (exports.Action = {}));
var AddressAddressTag;
(function (AddressAddressTag) {
    AddressAddressTag["UNTAGGED"] = "UNTAGGED";
    AddressAddressTag["HOME"] = "HOME";
    AddressAddressTag["WORK"] = "WORK";
    AddressAddressTag["BILLING"] = "BILLING";
    AddressAddressTag["SHIPPING"] = "SHIPPING";
    AddressAddressTag["OTHER"] = "OTHER";
})(AddressAddressTag = exports.AddressAddressTag || (exports.AddressAddressTag = {}));
var BlockType;
(function (BlockType) {
    BlockType["UNKNOWN_BLOCK_TYPE"] = "UNKNOWN_BLOCK_TYPE";
    BlockType["MANUAL"] = "MANUAL";
    BlockType["TIMEOUT"] = "TIMEOUT";
})(BlockType = exports.BlockType || (exports.BlockType = {}));
var ContactActivityType;
(function (ContactActivityType) {
    ContactActivityType["GENERAL"] = "GENERAL";
    ContactActivityType["CONTACT_CREATED"] = "CONTACT_CREATED";
    ContactActivityType["MEMBER_LOGIN"] = "MEMBER_LOGIN";
    ContactActivityType["MEMBER_REGISTER"] = "MEMBER_REGISTER";
    ContactActivityType["MEMBER_STATUS_CHANGED"] = "MEMBER_STATUS_CHANGED";
    ContactActivityType["FORM_SUBMITTED"] = "FORM_SUBMITTED";
    ContactActivityType["INBOX_FORM_SUBMITTED"] = "INBOX_FORM_SUBMITTED";
    ContactActivityType["INBOX_PAYMENT_REQUEST_PAID"] = "INBOX_PAYMENT_REQUEST_PAID";
    ContactActivityType["INBOX_MESSAGE_TO_CUSTOMER"] = "INBOX_MESSAGE_TO_CUSTOMER";
    ContactActivityType["INBOX_MESSAGE_FROM_CUSTOMER"] = "INBOX_MESSAGE_FROM_CUSTOMER";
    ContactActivityType["NEWSLETTER_SUBSCRIPTION_FORM_SUBMITTED"] = "NEWSLETTER_SUBSCRIPTION_FORM_SUBMITTED";
    ContactActivityType["NEWSLETTER_SUBSCRIPTION_UNSUBSCRIBE"] = "NEWSLETTER_SUBSCRIPTION_UNSUBSCRIBE";
    ContactActivityType["ECOM_PURCHASE"] = "ECOM_PURCHASE";
    ContactActivityType["ECOM_CART_ABANDON"] = "ECOM_CART_ABANDON";
    ContactActivityType["ECOM_CHECKOUT_BUYER"] = "ECOM_CHECKOUT_BUYER";
    ContactActivityType["BOOKINGS_APPOINTMENT"] = "BOOKINGS_APPOINTMENT";
    ContactActivityType["HOTELS_RESERVATION"] = "HOTELS_RESERVATION";
    ContactActivityType["HOTELS_PURCHASE"] = "HOTELS_PURCHASE";
    ContactActivityType["HOTELS_CONFIRMATION"] = "HOTELS_CONFIRMATION";
    ContactActivityType["HOTELS_CANCEL"] = "HOTELS_CANCEL";
    ContactActivityType["VIDEO_PURCHASE"] = "VIDEO_PURCHASE";
    ContactActivityType["VIDEO_RENT"] = "VIDEO_RENT";
    ContactActivityType["CASHIER_BUTTON_PURCHASE"] = "CASHIER_BUTTON_PURCHASE";
    ContactActivityType["ARENA_NEW_LEAD"] = "ARENA_NEW_LEAD";
    ContactActivityType["EVENTS_RSVP"] = "EVENTS_RSVP";
    ContactActivityType["INVOICE_PAY"] = "INVOICE_PAY";
    ContactActivityType["INVOICE_OVERDUE"] = "INVOICE_OVERDUE";
    ContactActivityType["PRICE_QUOTE_ACCEPT"] = "PRICE_QUOTE_ACCEPT";
    ContactActivityType["PRICE_QUOTE_EXPIRE"] = "PRICE_QUOTE_EXPIRE";
    ContactActivityType["RESTAURANTS_ORDER"] = "RESTAURANTS_ORDER";
    ContactActivityType["RESTAURANTS_RESERVATION"] = "RESTAURANTS_RESERVATION";
    ContactActivityType["SHOUTOUT_OPEN"] = "SHOUTOUT_OPEN";
    ContactActivityType["SHOUTOUT_CLICK"] = "SHOUTOUT_CLICK";
    ContactActivityType["CONTACT_MERGED"] = "CONTACT_MERGED";
})(ContactActivityType = exports.ContactActivityType || (exports.ContactActivityType = {}));
var ContactAddressAddressTag;
(function (ContactAddressAddressTag) {
    ContactAddressAddressTag["UNTAGGED"] = "UNTAGGED";
    ContactAddressAddressTag["HOME"] = "HOME";
    ContactAddressAddressTag["WORK"] = "WORK";
    ContactAddressAddressTag["BILLING"] = "BILLING";
    ContactAddressAddressTag["SHIPPING"] = "SHIPPING";
})(ContactAddressAddressTag = exports.ContactAddressAddressTag || (exports.ContactAddressAddressTag = {}));
var ContactEmailEmailTag;
(function (ContactEmailEmailTag) {
    ContactEmailEmailTag["UNTAGGED"] = "UNTAGGED";
    ContactEmailEmailTag["MAIN"] = "MAIN";
    ContactEmailEmailTag["HOME"] = "HOME";
    ContactEmailEmailTag["WORK"] = "WORK";
})(ContactEmailEmailTag = exports.ContactEmailEmailTag || (exports.ContactEmailEmailTag = {}));
var ContactFieldSet;
(function (ContactFieldSet) {
    ContactFieldSet["BASIC"] = "BASIC";
    ContactFieldSet["COMMUNICATION_DETAILS"] = "COMMUNICATION_DETAILS";
    ContactFieldSet["EXTENDED"] = "EXTENDED";
    ContactFieldSet["FULL"] = "FULL";
})(ContactFieldSet = exports.ContactFieldSet || (exports.ContactFieldSet = {}));
var ContactPhonePhoneTag;
(function (ContactPhonePhoneTag) {
    ContactPhonePhoneTag["UNTAGGED"] = "UNTAGGED";
    ContactPhonePhoneTag["MAIN"] = "MAIN";
    ContactPhonePhoneTag["HOME"] = "HOME";
    ContactPhonePhoneTag["MOBILE"] = "MOBILE";
    ContactPhonePhoneTag["WORK"] = "WORK";
    ContactPhonePhoneTag["FAX"] = "FAX";
})(ContactPhonePhoneTag = exports.ContactPhonePhoneTag || (exports.ContactPhonePhoneTag = {}));
var ContactSourceType;
(function (ContactSourceType) {
    ContactSourceType["OTHER"] = "OTHER";
    ContactSourceType["ADMIN"] = "ADMIN";
    ContactSourceType["WIX_APP"] = "WIX_APP";
    ContactSourceType["IMPORT"] = "IMPORT";
    ContactSourceType["THIRD_PARTY"] = "THIRD_PARTY";
    ContactSourceType["WIX_BOOKINGS"] = "WIX_BOOKINGS";
    ContactSourceType["WIX_CHAT"] = "WIX_CHAT";
    ContactSourceType["WIX_EMAIL_MARKETING"] = "WIX_EMAIL_MARKETING";
    ContactSourceType["WIX_EVENTS"] = "WIX_EVENTS";
    ContactSourceType["WIX_FORMS"] = "WIX_FORMS";
    ContactSourceType["WIX_GROUPS"] = "WIX_GROUPS";
    ContactSourceType["WIX_HOTELS"] = "WIX_HOTELS";
    ContactSourceType["WIX_MARKET_PLACE"] = "WIX_MARKET_PLACE";
    ContactSourceType["WIX_MUSIC"] = "WIX_MUSIC";
    ContactSourceType["WIX_RESTAURANTS"] = "WIX_RESTAURANTS";
    ContactSourceType["WIX_SITE_MEMBERS"] = "WIX_SITE_MEMBERS";
    ContactSourceType["WIX_STORES"] = "WIX_STORES";
})(ContactSourceType = exports.ContactSourceType || (exports.ContactSourceType = {}));
var ContactsFacetType;
(function (ContactsFacetType) {
    ContactsFacetType["UNKNOWN"] = "UNKNOWN";
    ContactsFacetType["ALL_CONTACTS"] = "ALL_CONTACTS";
    ContactsFacetType["NOT_LABELED"] = "NOT_LABELED";
    ContactsFacetType["LABEL"] = "LABEL";
    ContactsFacetType["SUBSCRIPTION_STATUS"] = "SUBSCRIPTION_STATUS";
    ContactsFacetType["MEMBERSHIP_STATUS"] = "MEMBERSHIP_STATUS";
})(ContactsFacetType = exports.ContactsFacetType || (exports.ContactsFacetType = {}));
var EmailEmailTag;
(function (EmailEmailTag) {
    EmailEmailTag["UNTAGGED"] = "UNTAGGED";
    EmailEmailTag["MAIN"] = "MAIN";
    EmailEmailTag["HOME"] = "HOME";
    EmailEmailTag["WORK"] = "WORK";
    EmailEmailTag["OTHER"] = "OTHER";
})(EmailEmailTag = exports.EmailEmailTag || (exports.EmailEmailTag = {}));
var GetBy;
(function (GetBy) {
    GetBy["ID"] = "ID";
    GetBy["PATH"] = "PATH";
})(GetBy = exports.GetBy || (exports.GetBy = {}));
var GetContactResponseType;
(function (GetContactResponseType) {
    GetContactResponseType["REGULAR"] = "REGULAR";
    GetContactResponseType["IMPLICIT"] = "IMPLICIT";
    GetContactResponseType["MERGED"] = "MERGED";
})(GetContactResponseType = exports.GetContactResponseType || (exports.GetContactResponseType = {}));
var ImageProvider;
(function (ImageProvider) {
    ImageProvider["UNKNOWN"] = "UNKNOWN";
    ImageProvider["EXTERNAL"] = "EXTERNAL";
    ImageProvider["WIX_MEDIA"] = "WIX_MEDIA";
})(ImageProvider = exports.ImageProvider || (exports.ImageProvider = {}));
var JobStatus;
(function (JobStatus) {
    JobStatus["UNKNOWN_STATUS"] = "UNKNOWN_STATUS";
    JobStatus["NEW"] = "NEW";
    JobStatus["LOADING"] = "LOADING";
    JobStatus["RUNNING"] = "RUNNING";
    JobStatus["COMPLETED"] = "COMPLETED";
})(JobStatus = exports.JobStatus || (exports.JobStatus = {}));
var JobType;
(function (JobType) {
    JobType["UNKNOWN_TYPE"] = "UNKNOWN_TYPE";
    JobType["DELETE"] = "DELETE";
    JobType["UPDATE"] = "UPDATE";
    JobType["ADD_LABELS"] = "ADD_LABELS";
    JobType["REMOVE_LABELS"] = "REMOVE_LABELS";
    JobType["ADD_REMOVE_LABELS"] = "ADD_REMOVE_LABELS";
})(JobType = exports.JobType || (exports.JobType = {}));
var PhonePhoneTag;
(function (PhonePhoneTag) {
    PhonePhoneTag["UNTAGGED"] = "UNTAGGED";
    PhonePhoneTag["MAIN"] = "MAIN";
    PhonePhoneTag["HOME"] = "HOME";
    PhonePhoneTag["MOBILE"] = "MOBILE";
    PhonePhoneTag["WORK"] = "WORK";
    PhonePhoneTag["FAX"] = "FAX";
    PhonePhoneTag["OTHER"] = "OTHER";
})(PhonePhoneTag = exports.PhonePhoneTag || (exports.PhonePhoneTag = {}));
var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder = exports.SortOrder || (exports.SortOrder = {}));
var SourceType;
(function (SourceType) {
    SourceType["OTHER"] = "OTHER";
    SourceType["IMPORT"] = "IMPORT";
    SourceType["APP"] = "APP";
    SourceType["SITE_MEMBERS"] = "SITE_MEMBERS";
    SourceType["MANUAL"] = "MANUAL";
})(SourceType = exports.SourceType || (exports.SourceType = {}));
var SubdivisionType;
(function (SubdivisionType) {
    SubdivisionType["UNKNOWN_SUBDIVISION_TYPE"] = "UNKNOWN_SUBDIVISION_TYPE";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_1"] = "ADMINISTRATIVE_AREA_LEVEL_1";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_2"] = "ADMINISTRATIVE_AREA_LEVEL_2";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_3"] = "ADMINISTRATIVE_AREA_LEVEL_3";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_4"] = "ADMINISTRATIVE_AREA_LEVEL_4";
    SubdivisionType["ADMINISTRATIVE_AREA_LEVEL_5"] = "ADMINISTRATIVE_AREA_LEVEL_5";
    SubdivisionType["COUNTRY"] = "COUNTRY";
})(SubdivisionType = exports.SubdivisionType || (exports.SubdivisionType = {}));
var SubmitOperation;
(function (SubmitOperation) {
    SubmitOperation["UNKNOWN"] = "UNKNOWN";
    SubmitOperation["CREATE"] = "CREATE";
    SubmitOperation["UPDATE"] = "UPDATE";
})(SubmitOperation = exports.SubmitOperation || (exports.SubmitOperation = {}));
