// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/ViewSite/getViewSiteAction.js
import * as React from 'react';
import { ActionType } from '../../../../types';
import VisibleSmall from 'wix-ui-icons-common/VisibleSmall';
import { PermissionMethods } from '../../../../services/actionsPermissions';
import { sendSiteActionsBI } from '../../../../services/BIService';
import { BIEvents } from '../../../../consts/bi';
function onViewSiteClicked(params) {
    var origin = params.origin, _a = params.quickActionsData, metaSiteId = _a.metaSiteId, viewUrl = _a.viewUrl;
    sendSiteActionsBI({
        evid: BIEvents.quickActionClicked,
        name: 'quick_actions.view_live_site',
        msid: metaSiteId,
        origin: origin
    });
    window.open(viewUrl);
}
export var getViewSiteAction = function (quickActionsData) {
    return {
        type: ActionType.click,
        actionName: 'view-site',
        textKey: 'quickActions.view_site',
        visible: Boolean(quickActionsData.published),
        icon: React.createElement(VisibleSmall, null),
        isPermitted: PermissionMethods.always,
        onActionClick: onViewSiteClicked,
    };
};
