// @wix/ambassador-action-triggers-server/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Visibility = exports.VersionType = exports.TranslationState = exports.TemplateType = exports.TemplateState = exports.StructNullValue = exports.StrategyType = exports.SourceSourceType = exports.SortOrder = exports.SimpleType = exports.SelectionType = exports.SearchRulesRequestOperator = exports.SchemaFieldType = exports.ScheduleStatus = exports.RuleStatus = exports.RuleSource = exports.ReengagementMode = exports.PrimitiveType = exports.PreconditionType = exports.IntegratorSourceType = exports.Exposure = exports.EventType = exports.ErrorResult = exports.ConditionConfigurationOperator = exports.AuditStatus = exports.ActionVisibilityOverride = void 0;
var ActionVisibilityOverride;
(function (ActionVisibilityOverride) {
    ActionVisibilityOverride["UNKNOWN_ACTION_VISIBILITY_OVERRIDE"] = "UNKNOWN_ACTION_VISIBILITY_OVERRIDE";
    ActionVisibilityOverride["HIDE"] = "HIDE";
    ActionVisibilityOverride["SHOW"] = "SHOW";
})(ActionVisibilityOverride = exports.ActionVisibilityOverride || (exports.ActionVisibilityOverride = {}));
var AuditStatus;
(function (AuditStatus) {
    AuditStatus["Undefined"] = "Undefined";
    AuditStatus["Pending"] = "Pending";
    AuditStatus["Success"] = "Success";
    AuditStatus["Failure"] = "Failure";
})(AuditStatus = exports.AuditStatus || (exports.AuditStatus = {}));
var ConditionConfigurationOperator;
(function (ConditionConfigurationOperator) {
    ConditionConfigurationOperator["UNDEFINED"] = "UNDEFINED";
    ConditionConfigurationOperator["IN"] = "IN";
})(ConditionConfigurationOperator = exports.ConditionConfigurationOperator || (exports.ConditionConfigurationOperator = {}));
var ErrorResult;
(function (ErrorResult) {
    ErrorResult["NoError"] = "NoError";
    ErrorResult["InvalidTpaInstance"] = "InvalidTpaInstance";
    ErrorResult["IllegalParameter"] = "IllegalParameter";
    ErrorResult["NotSupported"] = "NotSupported";
    ErrorResult["EventNotFound"] = "EventNotFound";
})(ErrorResult = exports.ErrorResult || (exports.ErrorResult = {}));
var EventType;
(function (EventType) {
    EventType["UNKNOWN_EVENT_TYPE"] = "UNKNOWN_EVENT_TYPE";
    EventType["LEGACY"] = "LEGACY";
    EventType["DOMAIN_EVENT"] = "DOMAIN_EVENT";
})(EventType = exports.EventType || (exports.EventType = {}));
var Exposure;
(function (Exposure) {
    Exposure["UNKNOWN_EXPOSURE"] = "UNKNOWN_EXPOSURE";
    Exposure["SETUP"] = "SETUP";
    Exposure["HIDDEN"] = "HIDDEN";
    Exposure["EXPOSED"] = "EXPOSED";
})(Exposure = exports.Exposure || (exports.Exposure = {}));
var IntegratorSourceType;
(function (IntegratorSourceType) {
    IntegratorSourceType["User"] = "User";
    IntegratorSourceType["Service"] = "Service";
})(IntegratorSourceType = exports.IntegratorSourceType || (exports.IntegratorSourceType = {}));
var PreconditionType;
(function (PreconditionType) {
    PreconditionType["UNKNOWN_PRECONDITION_TYPE"] = "UNKNOWN_PRECONDITION_TYPE";
    PreconditionType["PETRI"] = "PETRI";
    PreconditionType["STATE"] = "STATE";
    PreconditionType["APP_DEPENDENCY"] = "APP_DEPENDENCY";
    PreconditionType["DEALER_OFFERING"] = "DEALER_OFFERING";
})(PreconditionType = exports.PreconditionType || (exports.PreconditionType = {}));
var PrimitiveType;
(function (PrimitiveType) {
    PrimitiveType["UNKNOWN_PRIMITIVE_TYPE"] = "UNKNOWN_PRIMITIVE_TYPE";
    PrimitiveType["TEXT"] = "TEXT";
    PrimitiveType["BOOLEAN"] = "BOOLEAN";
    PrimitiveType["NUMBER"] = "NUMBER";
})(PrimitiveType = exports.PrimitiveType || (exports.PrimitiveType = {}));
var ReengagementMode;
(function (ReengagementMode) {
    ReengagementMode["UNKNOWN_REENGAGEMENT_MODE"] = "UNKNOWN_REENGAGEMENT_MODE";
    ReengagementMode["DISABLED"] = "DISABLED";
    ReengagementMode["ENABLED"] = "ENABLED";
})(ReengagementMode = exports.ReengagementMode || (exports.ReengagementMode = {}));
var RuleSource;
(function (RuleSource) {
    RuleSource["UNKNOWN"] = "UNKNOWN";
    RuleSource["USER"] = "USER";
    RuleSource["SERVICE"] = "SERVICE";
})(RuleSource = exports.RuleSource || (exports.RuleSource = {}));
var RuleStatus;
(function (RuleStatus) {
    RuleStatus["UnknownStatus"] = "UnknownStatus";
    RuleStatus["Active"] = "Active";
    RuleStatus["Suspended"] = "Suspended";
})(RuleStatus = exports.RuleStatus || (exports.RuleStatus = {}));
var ScheduleStatus;
(function (ScheduleStatus) {
    ScheduleStatus["Undefined"] = "Undefined";
    ScheduleStatus["Waiting"] = "Waiting";
    ScheduleStatus["Cancelled"] = "Cancelled";
    ScheduleStatus["Done"] = "Done";
    ScheduleStatus["Error"] = "Error";
})(ScheduleStatus = exports.ScheduleStatus || (exports.ScheduleStatus = {}));
var SchemaFieldType;
(function (SchemaFieldType) {
    SchemaFieldType["UNKNOWN_SCHEMA_FIELD_TYPE"] = "UNKNOWN_SCHEMA_FIELD_TYPE";
    SchemaFieldType["TEXT"] = "TEXT";
    SchemaFieldType["BOOLEAN"] = "BOOLEAN";
    SchemaFieldType["NUMBER"] = "NUMBER";
    SchemaFieldType["DATETIME"] = "DATETIME";
    SchemaFieldType["MONEY"] = "MONEY";
    SchemaFieldType["LINK"] = "LINK";
    SchemaFieldType["BACKOFFICE_LINK"] = "BACKOFFICE_LINK";
    SchemaFieldType["LIVESITE_LINK"] = "LIVESITE_LINK";
    SchemaFieldType["MULTILINGUAL"] = "MULTILINGUAL";
    SchemaFieldType["IMAGE_LINK"] = "IMAGE_LINK";
    SchemaFieldType["ENUM"] = "ENUM";
})(SchemaFieldType = exports.SchemaFieldType || (exports.SchemaFieldType = {}));
var SearchRulesRequestOperator;
(function (SearchRulesRequestOperator) {
    SearchRulesRequestOperator["UNKNOWN"] = "UNKNOWN";
    SearchRulesRequestOperator["AND"] = "AND";
    SearchRulesRequestOperator["OR"] = "OR";
})(SearchRulesRequestOperator = exports.SearchRulesRequestOperator || (exports.SearchRulesRequestOperator = {}));
var SelectionType;
(function (SelectionType) {
    SelectionType["UNKNOWN_SELECTION_TYPE"] = "UNKNOWN_SELECTION_TYPE";
    SelectionType["SINGLE"] = "SINGLE";
    SelectionType["MULTI"] = "MULTI";
})(SelectionType = exports.SelectionType || (exports.SelectionType = {}));
var SimpleType;
(function (SimpleType) {
    SimpleType["UNKNOWN_SIMPLE_TYPE"] = "UNKNOWN_SIMPLE_TYPE";
    SimpleType["MONEY"] = "MONEY";
    SimpleType["LINK"] = "LINK";
    SimpleType["BACKOFFICE_LINK"] = "BACKOFFICE_LINK";
    SimpleType["LIVESITE_LINK"] = "LIVESITE_LINK";
    SimpleType["MULTILINGUAL"] = "MULTILINGUAL";
    SimpleType["IMAGE_LINK"] = "IMAGE_LINK";
    SimpleType["GUID"] = "GUID";
    SimpleType["EMAIL"] = "EMAIL";
    SimpleType["PHONE"] = "PHONE";
    SimpleType["CONTACT_ID"] = "CONTACT_ID";
})(SimpleType = exports.SimpleType || (exports.SimpleType = {}));
var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder = exports.SortOrder || (exports.SortOrder = {}));
var SourceSourceType;
(function (SourceSourceType) {
    SourceSourceType["UNKNOWN_SOURCE_TYPE"] = "UNKNOWN_SOURCE_TYPE";
    SourceSourceType["WIX_APP"] = "WIX_APP";
    SourceSourceType["EXTERNAL"] = "EXTERNAL";
    SourceSourceType["ADMIN"] = "ADMIN";
    SourceSourceType["OTHER"] = "OTHER";
})(SourceSourceType = exports.SourceSourceType || (exports.SourceSourceType = {}));
var StrategyType;
(function (StrategyType) {
    StrategyType["NotSet"] = "NotSet";
    StrategyType["Simple"] = "Simple";
    StrategyType["Debounce"] = "Debounce";
})(StrategyType = exports.StrategyType || (exports.StrategyType = {}));
var StructNullValue;
(function (StructNullValue) {
    StructNullValue["NULL_VALUE"] = "NULL_VALUE";
})(StructNullValue = exports.StructNullValue || (exports.StructNullValue = {}));
var TemplateState;
(function (TemplateState) {
    TemplateState["UNKNOWN_TEMPLATE_STATUS"] = "UNKNOWN_TEMPLATE_STATUS";
    TemplateState["INTERNAL"] = "INTERNAL";
    TemplateState["PUBLIC"] = "PUBLIC";
    TemplateState["ARCHIVED"] = "ARCHIVED";
    TemplateState["LIVE"] = "LIVE";
})(TemplateState = exports.TemplateState || (exports.TemplateState = {}));
var TemplateType;
(function (TemplateType) {
    TemplateType["UNKNOWN_TEMPLATE_TYPE"] = "UNKNOWN_TEMPLATE_TYPE";
    TemplateType["PREINSTALLED"] = "PREINSTALLED";
    TemplateType["RECIPE"] = "RECIPE";
})(TemplateType = exports.TemplateType || (exports.TemplateType = {}));
var TranslationState;
(function (TranslationState) {
    TranslationState["UNKNOWN_TRANSLATION_STATE"] = "UNKNOWN_TRANSLATION_STATE";
    TranslationState["NOT_REQUESTED"] = "NOT_REQUESTED";
    TranslationState["PENDING"] = "PENDING";
    TranslationState["PARTIAL"] = "PARTIAL";
    TranslationState["TRANSLATED"] = "TRANSLATED";
})(TranslationState = exports.TranslationState || (exports.TranslationState = {}));
var VersionType;
(function (VersionType) {
    VersionType["UNKNOWN_VERSION_TYPE"] = "UNKNOWN_VERSION_TYPE";
    VersionType["MINOR"] = "MINOR";
    VersionType["MAJOR"] = "MAJOR";
})(VersionType = exports.VersionType || (exports.VersionType = {}));
var Visibility;
(function (Visibility) {
    Visibility["UNKNOWN_VISIBILITY"] = "UNKNOWN_VISIBILITY";
    Visibility["VISIBLE"] = "VISIBLE";
    Visibility["READONLY"] = "READONLY";
    Visibility["HIDDEN"] = "HIDDEN";
})(Visibility = exports.Visibility || (exports.Visibility = {}));
