// @wix/dealer-common/dist/src/converters/toDomain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertOfferToBannerData = exports.isNil = void 0;
function isNil(x) {
    return x === null || x === undefined;
}
exports.isNil = isNil;
function getDynamicParams(newDynamicParams) {
    var dynamicParams = {};
    for (var propName in newDynamicParams) {
        if (!isNil(newDynamicParams[propName])) {
            dynamicParams[propName] = newDynamicParams[propName];
        }
    }
    return dynamicParams;
}
function convertOfferToBannerData(offer) {
    var offerDynamicParams = offer.asset.dynamicParams || {};
    var dynamicParams = getDynamicParams(offerDynamicParams);
    var payload = JSON.parse(offer.asset.payloadJson);
    var bannerData = {
        bannerPosition: offer.realEstateId,
        bannerId: payload.templateId,
        bannerTemplate: payload.template,
        lightboxTriggers: payload.triggers,
        lightboxGuid: offer.offerGuid,
        dynamicParams: dynamicParams,
        saleType: offerDynamicParams.SALE_TYPE,
        isPremium: undefined,
        daysLeft: !isNil(offerDynamicParams.DAYS_LEFT)
            ? parseInt(offerDynamicParams.DAYS_LEFT, 10)
            : undefined,
    };
    return bannerData;
}
exports.convertOfferToBannerData = convertOfferToBannerData;
