// @wix/business-manager-api/dist/src/pageLinkBuilderFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pageLinkBuilderFactory = void 0;
var URI = require("urijs");
var wix_experiments_1 = require("@wix/wix-experiments");
var PageComponentsUrlTemplates_1 = require("./url-templates/PageComponentsUrlTemplates");
var MyAccountUrlTemplates_1 = require("./url-templates/MyAccountUrlTemplates");
var ExternalPageUrlTemplates_1 = require("./url-templates/ExternalPageUrlTemplates");
var MyAccountAppsUrlTemplates_1 = require("./url-templates/MyAccountAppsUrlTemplates");
var isGuid = function (str) { return str.match(/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i) !== null; };
function pageLinkBuilderFactory(metaSiteId, isDebug, pageComponentsInfo, experiments) {
    if (isDebug === void 0) { isDebug = false; }
    if (experiments === void 0) { experiments = {}; }
    var wixExperiments = new wix_experiments_1.default({ experiments: experiments });
    return function (targetId, contextData) {
        if (contextData === void 0) { contextData = {}; }
        var templateUrl = getTemplateUrl(metaSiteId, pageComponentsInfo, targetId, wixExperiments);
        if (!templateUrl) {
            return;
        }
        return processUrlTemplate(templateUrl, contextData, isDebug, wixExperiments);
    };
}
exports.pageLinkBuilderFactory = pageLinkBuilderFactory;
function getTemplateUrl(metaSiteId, pageComponentsInfo, targetId, experiments) {
    var pageComponentInfo = getPageComponentInfoByPageComponentId(pageComponentsInfo, targetId) ||
        getPageComponentInfoByAppDefId(pageComponentsInfo, targetId);
    if (pageComponentInfo) {
        return PageComponentsUrlTemplates_1.getBusinessManagerPageComponentTemplate(metaSiteId, pageComponentInfo.route);
    }
    if (experiments.enabled('specs.wos.DeprecateRedirectToMyAccount')) {
        var externalPageUrlTemplate = ExternalPageUrlTemplates_1.getExternalPageUrlTemplate(metaSiteId, targetId);
        if (externalPageUrlTemplate) {
            return externalPageUrlTemplate;
        }
        else if (isGuid(targetId)) {
            return PageComponentsUrlTemplates_1.getBusinessManagerPageComponentTemplate(metaSiteId, "app/" + targetId);
        }
        return;
    }
    var appDefId = pageComponentInfo ? pageComponentInfo.appDefId : targetId;
    return MyAccountUrlTemplates_1.getMyAccountUrlTemplate(metaSiteId, targetId) ||
        ExternalPageUrlTemplates_1.getExternalPageUrlTemplate(metaSiteId, targetId) ||
        MyAccountAppsUrlTemplates_1.getMyAccountAppsUrlTemplates(metaSiteId, appDefId);
}
function getPageComponentInfoByAppDefId(pageComponentsInfo, appDefId) {
    return Object
        .keys(pageComponentsInfo)
        .map(function (pageComponentId) { return pageComponentsInfo[pageComponentId]; })
        .find(function (pageComponentInfo) { return pageComponentInfo.appDefId === appDefId && pageComponentInfo.isMain === true; });
}
function getPageComponentInfoByPageComponentId(pageComponentsInfo, pageComponentId) {
    return pageComponentsInfo[pageComponentId];
}
function processUrlTemplate(urlTemplate, context, isDebug, experiments) {
    var uri = new URI(urlTemplate.baseUrl(context));
    if (urlTemplate.query) {
        uri.setQuery(interpolateOptions(urlTemplate.query, context));
    }
    if (!experiments.enabled('specs.wos.DeprecateRedirectToMyAccount') && isDebug && (uri.path().indexOf('my-account') !== -1 || !uri.host())) {
        uri.addQuery('debug');
    }
    return uri.toString();
}
function interpolateOptions(options, context) {
    return Object.keys(options || {}).reduce(function (result, key) {
        var interpolated;
        if (options[key] === '=') {
            interpolated = context[key];
        }
        else if (typeof options[key] === 'function') {
            interpolated = options[key](context);
        }
        else {
            interpolated = options[key];
        }
        if (interpolated) {
            result[key] = interpolated;
        }
        return result;
    }, {});
}
