// @wix/quick-actions-dropdown/dist/es/src/types/index.js
export * from './site-actions';
export var ContributorRoles;
(function (ContributorRoles) {
    ContributorRoles["owner"] = "owner";
    ContributorRoles["admin"] = "contributor";
    ContributorRoles["limitedAdmin"] = "limitedAdmin";
    ContributorRoles["websiteManager"] = "websiteManager";
    ContributorRoles["limitedWebsiteManager"] = "limitedWebsiteManager";
    ContributorRoles["backOfficeManager"] = "backOfficeManager";
    ContributorRoles["blogEditor"] = "blogEditor";
    ContributorRoles["blogWriter"] = "blogWriter";
    ContributorRoles["eventManager"] = "eventCheckInAssistant";
    ContributorRoles["bookingsAdmin"] = "bookingsAdmin";
    ContributorRoles["bookingsStaffMember"] = "bookingsStaffMember";
    ContributorRoles["analyticsViewer"] = "siteAnalyticsViewer";
    ContributorRoles["storeManager"] = "storeManager";
    ContributorRoles["cashierManager"] = "cashierManager";
    ContributorRoles["customerManager"] = "customerManager";
    ContributorRoles["customerFinanceManager"] = "customerFinanceManager";
})(ContributorRoles || (ContributorRoles = {}));
export var TOrigin;
(function (TOrigin) {
    TOrigin["Dashboard"] = "dashboard";
    TOrigin["mySites"] = "my sites";
    TOrigin["trash"] = "trash";
    TOrigin["siteList"] = "site_list.quick-actions";
    TOrigin["SiteSettings"] = "site_settings.quick-actions";
    TOrigin["SiteSettingsOverview"] = "site_settings.overview-tab";
    TOrigin["SiteSettingsDomain"] = "site_settings.domain-tab";
})(TOrigin || (TOrigin = {}));
export var PrefetchBundle;
(function (PrefetchBundle) {
    PrefetchBundle[PrefetchBundle["AccountLevel"] = 0] = "AccountLevel";
})(PrefetchBundle || (PrefetchBundle = {}));
