// @wix/quick-actions-dropdown/dist/es/src/types/site-actions.js
// Todo: deprecate
export var DialogNames;
(function (DialogNames) {
    DialogNames["rename"] = "rename";
    DialogNames["transfer"] = "transfer";
    DialogNames["deleteSite"] = "delete";
    DialogNames["moveToTrash"] = "move-to-trash";
    DialogNames["restore"] = "restore";
    DialogNames["removeFromTrash"] = "remove-from-trash";
    DialogNames["duplicate"] = "duplicate";
    DialogNames["getFeedback"] = "get-feedback";
    DialogNames["addContributor"] = "add-contributor";
    DialogNames["goToFolder"] = "go-to-Folder";
    DialogNames["moveSiteToFolder"] = "move-site-to-folder";
    DialogNames["migrateToEditorX"] = "migrate-to-editorx";
})(DialogNames || (DialogNames = {}));
export var ActionType;
(function (ActionType) {
    ActionType[ActionType["separator"] = 0] = "separator";
    ActionType[ActionType["loader"] = 1] = "loader";
    ActionType[ActionType["dialog"] = 2] = "dialog";
    ActionType[ActionType["click"] = 3] = "click";
})(ActionType || (ActionType = {}));
/**
 *  Actions - Type Checkers
 */
export var isActionWithDialog = function (sa) { return sa.type === ActionType.dialog; };
