// @wix/quick-actions-dropdown/dist/es/src/components/lazyActionsConnector/SiteActions/GetFeedback/getFeedbackSiteAction.js
import * as React from 'react';
import { DialogNames, ActionType } from '../../../../types';
import { FeedbackSiteAction } from './FeedbackSiteAction';
import ChatSmall from 'wix-ui-icons-common/ChatSmall';
import { PermissionMethods } from '../../../../services/actionsPermissions';
export var getFeedbackSiteAction = function (experiments) {
    var isPermitted = experiments['specs.wos1.EnableGetFeedbackonEditorX'] === 'true' ?
        PermissionMethods.htmlSiteAndHasGetFeedbackPermission :
        PermissionMethods.htmlSiteAndNotEditorXAndHasGetFeedbackPermission;
    var isPermittedDeprecated = experiments['specs.wos1.EnableGetFeedbackonEditorX'] === 'true' ?
        PermissionMethods.htmlSiteAndHasGetFeedbackPermissionDeprecated :
        PermissionMethods.htmlSiteAndNotEditorXAndHasGetFeedbackPermissionDeprecated;
    return {
        type: ActionType.dialog,
        dialog: {
            dialogName: DialogNames.getFeedback,
            component: FeedbackSiteAction,
            biName: 'quick_actions.get_feedback'
        },
        actionName: 'get-feedback',
        textKey: 'quickActions.get_feedback',
        icon: React.createElement(ChatSmall, null),
        isPermitted: experiments['specs.experts.FeedBackNg2'] === 'true' ? isPermitted : isPermittedDeprecated,
        visible: true,
    };
};
