// @wix/ambassador-dealer-v1-serving-offer/build/es/http.js
import { serializer } from '@wix/metro-runtime/ambassador';
import { resolveUrl } from '@wix/metro-runtime';
var _listOffersBulkRequest = {};
var _listOffersBulkResponse = { realEstates: 'Map#_listOffersResponse' };
var _listOffersRequest = {};
var _listOffersResponse = { offers: '_offer' };
var _offer = { startDate: 'google.protobuf.Timestamp' };
function resolveWixDealerOffersServingApiV1DealerOffersServingUrl(opts) {
    var domainToMappings = {
        'api._api_base_domain_': [
            {
                srcPath: '/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'www._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'support._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
            {
                srcPath: '/answers-routing/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'scheduler._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'premium._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'events._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'editor._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'ecom._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'apps._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'bo._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
        'manage._base_domain_': [
            {
                srcPath: '/_api/dealer-offers-serving-service',
                destPath: '',
            },
        ],
    };
    return resolveUrl(Object.assign(opts, { domainToMappings: domainToMappings }));
}
export function listOffers(payload) {
    return function _listOffers(_a) {
        var host = _a.host;
        var toJSON = serializer(_listOffersRequest, {}).toJSON;
        var fromJSON = serializer(_listOffersResponse, { _offer: _offer }).fromJSON;
        var serializedData = toJSON(payload);
        var metadata = {
            method: 'POST',
            url: resolveWixDealerOffersServingApiV1DealerOffersServingUrl({
                protoPath: '/v1/dealer/serving/offers/list',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromJSON,
        };
        return metadata;
    };
}
export function listOffersBulk(payload) {
    return function _listOffersBulk(_a) {
        var host = _a.host;
        var toJSON = serializer(_listOffersBulkRequest, {}).toJSON;
        var fromJSON = serializer(_listOffersBulkResponse, {
            _listOffersResponse: _listOffersResponse,
            _offer: _offer,
        }).fromJSON;
        var serializedData = toJSON(payload);
        var metadata = {
            method: 'POST',
            url: resolveWixDealerOffersServingApiV1DealerOffersServingUrl({
                protoPath: '/v1/dealer/serving/offers/bulk',
                data: serializedData,
                host: host,
            }),
            data: serializedData,
            transformResponse: fromJSON,
        };
        return metadata;
    };
}
export * from './types';
