// @wix/ambassador-premium-user-domains/types.js
/* eslint-disable */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1ManagementType = exports.ManagementTypeEnumManagementType = exports.DnsRecordType = exports.ContributorRole = exports.ConnectivityStatus = exports.BindType = exports.AssignState = void 0;
var AssignState;
(function (AssignState) {
    AssignState["ASSIGNED"] = "ASSIGNED";
    AssignState["UNASSIGNED"] = "UNASSIGNED";
})(AssignState = exports.AssignState || (exports.AssignState = {}));
var BindType;
(function (BindType) {
    BindType["PRIMARY"] = "PRIMARY";
    BindType["REDIRECT"] = "REDIRECT";
})(BindType = exports.BindType || (exports.BindType = {}));
var ConnectivityStatus;
(function (ConnectivityStatus) {
    ConnectivityStatus["UNKNOWN"] = "UNKNOWN";
    ConnectivityStatus["OK"] = "OK";
    ConnectivityStatus["MISSING_RECORDS"] = "MISSING_RECORDS";
    ConnectivityStatus["INVALID_RECORDS"] = "INVALID_RECORDS";
    ConnectivityStatus["PROPAGATION"] = "PROPAGATION";
    ConnectivityStatus["REDUNDANT_NS"] = "REDUNDANT_NS";
})(ConnectivityStatus = exports.ConnectivityStatus || (exports.ConnectivityStatus = {}));
var ContributorRole;
(function (ContributorRole) {
    ContributorRole["DOMAIN_MANAGER"] = "DOMAIN_MANAGER";
    ContributorRole["BILLING_MANAGER"] = "BILLING_MANAGER";
    ContributorRole["WEBSITE_MANAGER"] = "WEBSITE_MANAGER";
})(ContributorRole = exports.ContributorRole || (exports.ContributorRole = {}));
var DnsRecordType;
(function (DnsRecordType) {
    DnsRecordType["A"] = "A";
    DnsRecordType["CNAME"] = "CNAME";
    DnsRecordType["NS"] = "NS";
})(DnsRecordType = exports.DnsRecordType || (exports.DnsRecordType = {}));
var ManagementTypeEnumManagementType;
(function (ManagementTypeEnumManagementType) {
    ManagementTypeEnumManagementType["UNKNOWN"] = "UNKNOWN";
    ManagementTypeEnumManagementType["POINTING"] = "POINTING";
    ManagementTypeEnumManagementType["DNS"] = "DNS";
    ManagementTypeEnumManagementType["REGISTRATION"] = "REGISTRATION";
    ManagementTypeEnumManagementType["POINTING_GOOGLE"] = "POINTING_GOOGLE";
    ManagementTypeEnumManagementType["POINTING_REG_RU"] = "POINTING_REG_RU";
    ManagementTypeEnumManagementType["POINTING_MSO_365"] = "POINTING_MSO_365";
    ManagementTypeEnumManagementType["PHOTOGRAPHERS"] = "PHOTOGRAPHERS";
})(ManagementTypeEnumManagementType = exports.ManagementTypeEnumManagementType || (exports.ManagementTypeEnumManagementType = {}));
var V1ManagementType;
(function (V1ManagementType) {
    V1ManagementType["POINTING"] = "POINTING";
    V1ManagementType["DNS"] = "DNS";
})(V1ManagementType = exports.V1ManagementType || (exports.V1ManagementType = {}));
